{
 "metadata": {
  "name": "",
  "signature": "sha256:d7be862d5347b436fcfb20dfb9ea1e397a11b8a8b349fd7142f9242fdd3542c8"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "<center><h1> Functional Impact SNP Set Enrichment (fiSSEA) </h1></center>\n",
      "<center><h4>Authors: Erick R. Scott, Adam Mark, Ali Torkamani, Chunlei Wu & Andrew Su</h4></center>\n",
      "<center><h4>November 2014</h4></center>"
     ]
    },
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "Libraries"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import os, sys, time, gc, argparse\n",
      "import json, httplib2\n",
      "import pandas as pd\n",
      "from pandas.io.json import json_normalize\n",
      "import numpy as np\n",
      "import subprocess"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1
    },
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "PandasVCF, MyVariant.info, fiSSEA Libraries"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import warnings\n",
      "warnings.simplefilter(action = \"ignore\", category = FutureWarning)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 2
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "sys.path.append('../pandasVCF/src/multi_sample/')\n",
      "from pandasVCFmulti import *\n",
      "\n",
      "sys.path.append('../myvariant/src/')\n",
      "import myvariant\n",
      "mv = myvariant.MyVariantInfo()\n",
      "\n",
      "sys.path.append('../src/')\n",
      "from fiSSEA_multi import *"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 3
    },
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "VCF CDS Variants"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def intersect_cds_vcf(vcf_path, cds_bed, write_path):\n",
      "    '''\n",
      "    Convenience function to create coding snp vcf file\n",
      "    '''\n",
      "    os.system('bedtools intersect -a ' + vcf_path + ' -b ' + cds_bed + ' -wa -header |uniq > ' + write_path )\n",
      "    return write_path"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 4
    },
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "Running fiSSEA "
     ]
    },
    {
     "cell_type": "heading",
     "level": 4,
     "metadata": {},
     "source": [
      "<h1>GSEA Command Line Run</h1>\n",
      "\n",
      "<ol> Please download the gsea.jar source from http://www.broadinstitute.org/gsea/downloads.jsp</ol>\n",
      "\n",
      "<ol>We have included the tissue-specific gene set .gmt data derived from the Xavier Lab's Enrichment Profiler dataset (http://xavierlab2.mgh.harvard.edu/EnrichmentProfiler/download.html) in the /fiSSEA/gsea/genesets/ directory</ol>\n",
      "\n"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "!tabix -p vcf ALL.chr1.phase3_shapeit2_mvncall_integrated_v4.20130502.genotypes_CDS_10000.vcf.gz"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "open: No such file or directory\r\n",
        "[tabix] was bgzip used to compress this file? ALL.chr1.phase3_shapeit2_mvncall_integrated_v4.20130502.genotypes_CDS_10000.vcf.gz\r\n"
       ]
      }
     ],
     "prompt_number": 39
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "!ls /Users/erickscott/Dropbox/TSRI/fiSSEA_data/"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "ALL.chr1.phase3_shapeit2_mvncall_integrated_v4.20130502.genotypes_CDS.vcf.gz\r\n",
        "ALL.chr1.phase3_shapeit2_mvncall_integrated_v4.20130502.genotypes_CDS.vcf.gz.tbi\r\n",
        "ALL.chr1.phase3_shapeit2_mvncall_integrated_v4.20130502.genotypes_CDS_1000.vcf\r\n",
        "ALL.chr1.phase3_shapeit2_mvncall_integrated_v4.20130502.genotypes_CDS_10000.vcf.gz\r\n"
       ]
      }
     ],
     "prompt_number": 38
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "example_vcf = '../data//ALL.chr1.phase3_shapeit2_mvncall_integrated_v4.20130502.genotypes_CDS_10000.vcf'\n",
      "sample_id = 'all'\n",
      "chunk = 1000\n",
      "fi_score = 'dbnsfp.cadd.phred'\n",
      "\n",
      "#Set the path to the gsea .jar file\n",
      "gsea_jar = '../gsea/gsea2-2.1.0.jar'\n",
      "\n",
      "#Set the path to the tissue specific gene set\n",
      "tissue_genesets_path = '../gsea/genesets/EnrichmentValues_Primary_U133A_ave_percent_over70percent_no_brain.gmt'\n",
      "\n",
      "#Set the path to write GSEA results\n",
      "output_dir_path = '../gsea/output/'\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 29
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "vcf_df_obj.add_variant_annotations()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "End of File Reached\n"
       ]
      },
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 24,
       "text": [
        "1"
       ]
      }
     ],
     "prompt_number": 24
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "pandasVCFmulti Iteration Example"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "\n",
      "vcf_df_obj = Vcf(example_vcf, 'all', cols=['#CHROM', 'POS', 'REF', 'ALT', 'FORMAT'], chunksize=1000, n_cores=6)\n",
      "stopIteration = False\n",
      "data = []\n",
      "while stopIteration == False:\n",
      "    \n",
      "    vcf_df_obj.get_vcf_df_chunk()\n",
      "    if vcf_df_obj.stopIteration == True: break  #prevents duplicating data\n",
      "    vcf_df_obj.add_variant_annotations()\n",
      "    data.append(vcf_df_obj.df_annot)  #aggregating annotation data\n",
      "    \n",
      "pd.concat(data)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 28
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "fiSSEA_multi Example"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "fiSSEA_results = fiSSEA(example_vcf, sample_id=sample_id, chunksize=chunk, fiSSEA_score=fi_score, n_cores=4)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#Call GSEA jar PreRanked using several default parameters and specifying a specific sample ('HG00099')\n",
      "#NOTE THIS WILL NOTE PRODUCE A VALID GSEA RESULTS, THERE ARE TOO FEW GENES\n",
      "fiSSEA_results.run_GSEA_preranked(gsea_jar, tissue_genesets_path, output_dir_path, 'HG00099')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 31,
       "text": [
        "0"
       ]
      }
     ],
     "prompt_number": 31
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "Opens GSEA Summary Html (working with Chrome browser)"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "fiSSEA_results.open_GSEA_html()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 32,
       "text": [
        "0"
       ]
      }
     ],
     "prompt_number": 32
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "Example code for iterating across all chromosomes"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#Generate Functional Impact Scores for all Nonsynon DNA variants in Coding VCF\n",
      "chrom_range = map(str,range(1,23))\n",
      "chrom_range.extend(['X'])\n",
      "\n",
      "fi_scores_all_chroms = []\n",
      "for c in chrom_range:\n",
      "    example_vcf = file_path + c + suffix\n",
      "    fiSSEA_results = fiSSEA(example_vcf, sample_id=sample_id, chunksize=chunk, fiSSEA_score=fi_score, n_cores=4)\n",
      "    fi_scores_all_chroms.append(fiSSEA_results.fi_scores)\n",
      "\n",
      "\n",
      "fi_scores_all_chroms = pd.concat(fi_scores_all_chroms)\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "Functional Impact Annotation Results"
     ]
    },
    {
     "cell_type": "heading",
     "level": 4,
     "metadata": {},
     "source": [
      "fiSSEA_results object has the following attributes"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "dir(fiSSEA_results)[-8:]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 10,
       "text": [
        "['fiSSEA_score',\n",
        " 'fiSSEA_status',\n",
        " 'fi_scores',\n",
        " 'gene_stat',\n",
        " 'get_fi_scores',\n",
        " 'get_multigene_var_records',\n",
        " 'mv_annot',\n",
        " 'vcf_df']"
       ]
      }
     ],
     "prompt_number": 10
    },
    {
     "cell_type": "heading",
     "level": 4,
     "metadata": {},
     "source": [
      "Original vcf parsed with pandasVCFmulti, saving fiSSEA required columns"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "fiSSEA_results.vcf_df.head()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>CHROM</th>\n",
        "      <th>POS</th>\n",
        "      <th>REF</th>\n",
        "      <th>ALT</th>\n",
        "      <th>sample_ids</th>\n",
        "      <th>a1</th>\n",
        "      <th>a2</th>\n",
        "      <th>vartype1</th>\n",
        "      <th>vartype2</th>\n",
        "      <th>hgvs_a1</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>0</th>\n",
        "      <td> 1</td>\n",
        "      <td> 10363371</td>\n",
        "      <td> G</td>\n",
        "      <td> C</td>\n",
        "      <td> NA18983</td>\n",
        "      <td> G</td>\n",
        "      <td> C</td>\n",
        "      <td> ref</td>\n",
        "      <td> snp</td>\n",
        "      <td> chr1:g.10363371G&gt;G</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>1</th>\n",
        "      <td> 1</td>\n",
        "      <td> 10363371</td>\n",
        "      <td> G</td>\n",
        "      <td> C</td>\n",
        "      <td> NA19000</td>\n",
        "      <td> G</td>\n",
        "      <td> C</td>\n",
        "      <td> ref</td>\n",
        "      <td> snp</td>\n",
        "      <td> chr1:g.10363371G&gt;G</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>2</th>\n",
        "      <td> 1</td>\n",
        "      <td> 10363433</td>\n",
        "      <td> C</td>\n",
        "      <td> T</td>\n",
        "      <td> NA18632</td>\n",
        "      <td> T</td>\n",
        "      <td> C</td>\n",
        "      <td> snp</td>\n",
        "      <td> ref</td>\n",
        "      <td> chr1:g.10363433C&gt;T</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>3</th>\n",
        "      <td> 1</td>\n",
        "      <td> 10363487</td>\n",
        "      <td> T</td>\n",
        "      <td> C</td>\n",
        "      <td> HG00126</td>\n",
        "      <td> T</td>\n",
        "      <td> C</td>\n",
        "      <td> ref</td>\n",
        "      <td> snp</td>\n",
        "      <td> chr1:g.10363487T&gt;T</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>4</th>\n",
        "      <td> 1</td>\n",
        "      <td> 10363487</td>\n",
        "      <td> T</td>\n",
        "      <td> C</td>\n",
        "      <td> HG01789</td>\n",
        "      <td> T</td>\n",
        "      <td> C</td>\n",
        "      <td> ref</td>\n",
        "      <td> snp</td>\n",
        "      <td> chr1:g.10363487T&gt;T</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 42,
       "text": [
        "   CHROM       POS REF ALT sample_ids a1 a2 vartype1 vartype2  \\\n",
        "0      1  10363371   G   C    NA18983  G  C      ref      snp   \n",
        "1      1  10363371   G   C    NA19000  G  C      ref      snp   \n",
        "2      1  10363433   C   T    NA18632  T  C      snp      ref   \n",
        "3      1  10363487   T   C    HG00126  T  C      ref      snp   \n",
        "4      1  10363487   T   C    HG01789  T  C      ref      snp   \n",
        "\n",
        "              hgvs_a1  \n",
        "0  chr1:g.10363371G>G  \n",
        "1  chr1:g.10363371G>G  \n",
        "2  chr1:g.10363433C>T  \n",
        "3  chr1:g.10363487T>T  \n",
        "4  chr1:g.10363487T>T  "
       ]
      }
     ],
     "prompt_number": 42
    },
    {
     "cell_type": "heading",
     "level": 4,
     "metadata": {},
     "source": [
      "Mean functional impact score for each Gene - Input into GSEA"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "fiSSEA_results.fi_scores.head()\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th>sample_ids</th>\n",
        "      <th>HG00096</th>\n",
        "      <th>HG00097</th>\n",
        "      <th>HG00099</th>\n",
        "      <th>HG00100</th>\n",
        "      <th>HG00101</th>\n",
        "      <th>HG00102</th>\n",
        "      <th>HG00103</th>\n",
        "      <th>HG00105</th>\n",
        "      <th>HG00106</th>\n",
        "      <th>HG00107</th>\n",
        "      <th>...</th>\n",
        "      <th>NA21128</th>\n",
        "      <th>NA21129</th>\n",
        "      <th>NA21130</th>\n",
        "      <th>NA21133</th>\n",
        "      <th>NA21135</th>\n",
        "      <th>NA21137</th>\n",
        "      <th>NA21141</th>\n",
        "      <th>NA21142</th>\n",
        "      <th>NA21143</th>\n",
        "      <th>NA21144</th>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.genename</th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>ACAP3</th>\n",
        "      <td>  0.00</td>\n",
        "      <td>  4.97</td>\n",
        "      <td>  0.00</td>\n",
        "      <td>  0.00</td>\n",
        "      <td>  0.000</td>\n",
        "      <td>  0.00</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>  0.00</td>\n",
        "      <td>  0.00</td>\n",
        "      <td>...</td>\n",
        "      <td>  0.00</td>\n",
        "      <td>  0.00</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>  0.00</td>\n",
        "      <td>  0.00</td>\n",
        "      <td> 0.000</td>\n",
        "      <td>  0.00</td>\n",
        "      <td>  0.00</td>\n",
        "      <td> 0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>ACOT7</th>\n",
        "      <td>  0.00</td>\n",
        "      <td>  0.00</td>\n",
        "      <td>  0.00</td>\n",
        "      <td>  0.00</td>\n",
        "      <td>  0.000</td>\n",
        "      <td>  0.00</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>  0.00</td>\n",
        "      <td>  0.00</td>\n",
        "      <td>...</td>\n",
        "      <td>  0.00</td>\n",
        "      <td>  0.00</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>  0.00</td>\n",
        "      <td>  0.00</td>\n",
        "      <td> 0.000</td>\n",
        "      <td>  0.00</td>\n",
        "      <td>  0.00</td>\n",
        "      <td> 0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>ACTRT2</th>\n",
        "      <td>  0.00</td>\n",
        "      <td>  0.00</td>\n",
        "      <td>  0.00</td>\n",
        "      <td> 16.11</td>\n",
        "      <td> 32.220</td>\n",
        "      <td> 32.22</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> 16.11</td>\n",
        "      <td>  0.00</td>\n",
        "      <td>...</td>\n",
        "      <td> 16.11</td>\n",
        "      <td>  0.00</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> 11.45</td>\n",
        "      <td>  0.00</td>\n",
        "      <td> 0.000</td>\n",
        "      <td>  0.00</td>\n",
        "      <td>  0.00</td>\n",
        "      <td> 0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>AGRN</th>\n",
        "      <td>  0.00</td>\n",
        "      <td>  0.00</td>\n",
        "      <td>  0.00</td>\n",
        "      <td>  0.00</td>\n",
        "      <td>  8.994</td>\n",
        "      <td>  0.00</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>  0.00</td>\n",
        "      <td>  0.00</td>\n",
        "      <td>...</td>\n",
        "      <td>  0.00</td>\n",
        "      <td>  0.00</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>  0.00</td>\n",
        "      <td>  0.00</td>\n",
        "      <td> 9.775</td>\n",
        "      <td>  0.00</td>\n",
        "      <td> 15.76</td>\n",
        "      <td> 0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>AJAP1</th>\n",
        "      <td> 20.28</td>\n",
        "      <td> 10.14</td>\n",
        "      <td> 10.14</td>\n",
        "      <td> 20.28</td>\n",
        "      <td> 20.280</td>\n",
        "      <td>  0.00</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> 20.28</td>\n",
        "      <td> 20.28</td>\n",
        "      <td>...</td>\n",
        "      <td>  0.00</td>\n",
        "      <td> 10.14</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>  0.00</td>\n",
        "      <td> 20.28</td>\n",
        "      <td> 0.000</td>\n",
        "      <td> 10.14</td>\n",
        "      <td>  0.00</td>\n",
        "      <td> 0</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "<p>5 rows \u00d7 2504 columns</p>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 43,
       "text": [
        "sample_ids       HG00096  HG00097  HG00099  HG00100  HG00101  HG00102  \\\n",
        "dbnsfp.genename                                                         \n",
        "ACAP3               0.00     4.97     0.00     0.00    0.000     0.00   \n",
        "ACOT7               0.00     0.00     0.00     0.00    0.000     0.00   \n",
        "ACTRT2              0.00     0.00     0.00    16.11   32.220    32.22   \n",
        "AGRN                0.00     0.00     0.00     0.00    8.994     0.00   \n",
        "AJAP1              20.28    10.14    10.14    20.28   20.280     0.00   \n",
        "\n",
        "sample_ids       HG00103  HG00105  HG00106  HG00107   ...     NA21128  \\\n",
        "dbnsfp.genename                                       ...               \n",
        "ACAP3                  0        0     0.00     0.00   ...        0.00   \n",
        "ACOT7                  0        0     0.00     0.00   ...        0.00   \n",
        "ACTRT2                 0        0    16.11     0.00   ...       16.11   \n",
        "AGRN                   0        0     0.00     0.00   ...        0.00   \n",
        "AJAP1                  0        0    20.28    20.28   ...        0.00   \n",
        "\n",
        "sample_ids       NA21129  NA21130  NA21133  NA21135  NA21137  NA21141  \\\n",
        "dbnsfp.genename                                                         \n",
        "ACAP3               0.00        0        0     0.00     0.00    0.000   \n",
        "ACOT7               0.00        0        0     0.00     0.00    0.000   \n",
        "ACTRT2              0.00        0        0    11.45     0.00    0.000   \n",
        "AGRN                0.00        0        0     0.00     0.00    9.775   \n",
        "AJAP1              10.14        0        0     0.00    20.28    0.000   \n",
        "\n",
        "sample_ids       NA21142  NA21143  NA21144  \n",
        "dbnsfp.genename                             \n",
        "ACAP3               0.00     0.00        0  \n",
        "ACOT7               0.00     0.00        0  \n",
        "ACTRT2              0.00     0.00        0  \n",
        "AGRN                0.00    15.76        0  \n",
        "AJAP1              10.14     0.00        0  \n",
        "\n",
        "[5 rows x 2504 columns]"
       ]
      }
     ],
     "prompt_number": 43
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "print fiSSEA_results.fi_scores['HG00097'].describe()\n",
      "fiSSEA_results.fi_scores['HG00097'].hist(bins=20)\n",
      "plt.title('Distribution of Max CADD Phred Score by Gene (Snyderome, Chr6)')\n",
      "plt.xlabel('Mean CADD Phred Score')\n",
      "plt.ylabel('Gene Count')\n",
      "plt.tight_layout()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "count    129.000000\n",
        "mean       2.333132\n",
        "std        6.370349\n",
        "min        0.000000\n",
        "25%        0.000000\n",
        "50%        0.000000\n",
        "75%        0.000000\n",
        "max       36.160000\n",
        "Name: HG00097, dtype: float64\n"
       ]
      },
      {
       "metadata": {},
       "output_type": "display_data",
       "png": "iVBORw0KGgoAAAANSUhEUgAAAaUAAAEaCAYAAACxTkgOAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzt3XdYFOfaBvB7UQEVUNRYIlJiAQGlKEVFWTCWaAzEEktE\nFMwxaDTR6DmJJWqKJoIaKxoMohL1xGgUMca+lliwYIkNMSiWGFE4slJE4fn+IMzHwi6gMrsz2ed3\nXVwwM7sz97w77LPzvrO7CiIiMMYYYxJgYugAjDHGWAkuSowxxiSDixJjjDHJ4KLEGGNMMrgoMcYY\nkwwuSowxxiRDL0UpIiICX375ZbWsKz09HZaWlii5kl2pVOL777+vlnUDQJ8+fbBu3bpqW19VLVu2\nDE5OTnj11Vf1vu1/kri4OHTt2lW26xeDvb099u3bZ+gYoli3bh3ef/99vW935MiRmDFjht63q2/V\n/fwaGRmJ2bNnV3ibly5K9vb2qFOnDqysrGBnZ4fAwED89NNPGreJjo7G9OnTq7Su/fv3V3gbW1tb\nqNVqKBQKAIBCoRD+fl6zZs1CSEiIxrxffvml3DyxFRQU4NNPP8XOnTtx9+7dcstVKhVMTEzQv39/\njfnnzp2DiYkJAgICRMl19+5dhISEoFWrVqhXrx7c3NywcOFCFBUVCbd5/PgxLCws0KdPn3L3r8qx\nMXLkSJiZmcHKygpNmzaFj48P5s2bh/z8fJ25Zs2ahVq1asHS0hK2trYICwvDuXPnqm/HX8LPP/+M\n/v37o2HDhmjatCn69u2Lx48fGzTTy/yPlPXkyRPMnDkTbm5usLKygq2tLQYNGoSkpKRqWf/zICLM\nmzcPEyZMEObpq/2rs00NqaioCAsWLECHDh1Qr1492NvbIzw8HDdv3gTw4vs5b948ODs7w8rKCp6e\nnnj06BEAIDw8HKtWrUJubq7O+750UVIoFEhMTER2djY2b94MX19ffPTRR5g8efILraui9/I+e/bs\nZaJK1pkzZ2BpaQkHBwedt3nllVdw/PhxZGZmCvPWrFmDNm3aiPLPce/ePbRv3x5qtRpLlizBgwcP\nsHbtWhw4cABqtVq43ebNm2FrawuVSoW//vpLYx1VOTYUCgX+85//IDs7GydOnMCQIUOwceNG+Pv7\naxS/susdOnQo1Go1kpOThXU8LyKq8Hh7Xn/99RfCwsIwfvx43Lt3D5cvX8aQIUOqbf2AYf8HiAg9\nevTAjh07MHbsWNy+fRunTp1CcHAwtm7dqvc827dvR8OGDeHs7AxAP+1f2oscO1J7DgsJCcGKFSsw\nYsQI3Lx5E7/99hucnJywZ8+e515Xyb4tWLAAmzZtQlRUFB49eoT4+HiYm5sDABo0aIDu3btXfPZF\nL8ne3p727dunMW/VqlVUo0YNunbtGhERhYaG0vTp04mIKCcnh8LDw8nOzo4aNGhA3bp1o6KiIho+\nfDiZmJhQ7dq1ycLCgiIjIyktLY0UCgX9+OOP5OLiQkqlkm7cuEEKhYIKCwuJiEipVNKXX35JAQEB\n1Lx5c5o7dy49fvyYiIgOHDhANjY2Gtns7Oxo7969tHPnTjI1NaVatWqRhYUFubu7ExGRv78/rVq1\nSrj9tm3b6PXXXydXV1eKjo6mnJwcIiIh2+bNm8nJyYnatWtH69at09lOOTk5tHz5cnJ1daUePXpQ\nQkICERHt2bOHateuTSYmJmRhYUGjRo0qd9+S/YiIiKBly5YREdGzZ8+oefPm9Pnnn5NSqRRuO2HC\nBGrRogU1btyY/vWvf9HZs2eFZX369KGPP/5YmB48eDCFhYVpzTt69Ghq06aNzv0pERAQQPPnz6fX\nX3+doqKiNJZV5dgYOXKkcGyUSE1NJYVCofE4lDZz5kwaPny4MH306FGqVasW5efn0+rVq8nPz49m\nz55Nr776KvXq1YuOHj0q3Nbf35+++uor6tGjB1laWtL169fp7t279Pnnn1PLli3pnXfeoePHjwu3\nV6vV9NVXX1Hz5s0pMDCQ5s6dS35+flpzrVy5kgICAnS21ZMnT2j9+vXUvXt3qlevHvn5+VFeXh4R\nEZ0+fZpCQ0PJzs6Opk+fTnfu3BHuZ2dnR8uWLaNOnTpR3bp1qbCwkFJTU2nKlClka2tLo0ePposX\nL+rcrr29PS1ZsoS8vLzotddeo+joaCooKCAiIhcXF9q+fbtw24KCAmrYsKHGcVNi3bp1ZGJiQnfv\n3tW5LSKqsD1DQ0Np4sSJ9M4771Djxo0pPDycbt++XaX7ljVhwgT65JNPhOnK2r+ibY8dO1bjf4OI\nqF+/frRw4UIiIrpx4wa999571KRJExo9ejQNHz5c47g9d+4cjRkzhlq0aEGTJk2imzdvCsu0PX6V\nPd7Lly8nX19faty4MU2ZMoUeP35MgwYNoqZNm9KHH35I2dnZwu2f51go7ciRI6RQKDT+P8pSKpU0\nd+5c6tmzJzVt2pQmTpxIWVlZRERan5+JiDp16kS7du3Suc7o6GgKDg7WuVyUopSRkUE1a9akjRs3\nElHxE8+MGTOIiGjp0qU0bNgwevToET179oyOHDmic10lOx0cHEzXr1+n/Px8YV5JUfL396emTZtS\nQkICpaamUvfu3YUDVVtRKr2NWbNmUUhIiMZypVJJ33//PRER7d+/n2xtbWnPnj2UkpJC3bt3p5kz\nZ2pkGzx4MKWnp9OuXbvIzMxMeJIp67PPPqOAgAC6evUq7du3j+zt7enAgQNERKRSqcrlLK1kP44e\nPUo+Pj5ERLRjxw7q1asXrVq1SqMoxcfHU2ZmJmVlZdGUKVOoS5cuwrJ79+5R48aNaf/+/RQfH08t\nW7YUCnhZbdq0ocmTJ+vMRFT8j1qjRg26desWfffdd9S+fXuN5VU9NsoWJSIiNzc3GjNmjNbtli5K\nf/31F40cOZL69etHRESrV68mU1NTmj17NmVmZtLMmTM1ioi/vz+9+uqrlJiYSE+fPqWCggLy8PCg\nOXPmUFZWFiUmJpK1tbXQLh9//DH16tWL/vjjD9q2bRs1adKEunbtqrM9zMzM6OOPP6YjR44IT/wl\nFixYQN7e3nTw4EEqLCykY8eO0ZMnTygnJ4csLCwoJiaG7t+/TxMmTCB/f3+NdnR2dqZDhw5Rfn4+\nPXv2jBo3bkyrV6+m7OxsWrNmTYXHj52dHTk6OtLhw4fp7Nmz5OHhQStWrCAionnz5tHgwYOF227d\nurXc41jiX//6F3l5eencTomK2jM0NJSsrKxoy5YtlJGRQW+++abG46/tvmq1Wut2evXqRTExMcJ0\nZe1f0baTkpLo1VdfpaKiIiIqPk7r1KlD9+/fJyKiDh060Mcff0wZGRkUGRlJpqamwnPagwcPyNra\nmrZu3UqPHj2iOXPmUOfOnYXtln38qvJ4e3h4UHJyMp07d46srKzIy8uLEhIS6O7du+Tj40Nr164l\nInruY6G0OXPmUJMmTSq8jb+/P7Vo0YL27t1Lt2/fJi8vL+HForbn5/v375O5uTl99tln5ODgQH37\n9qXExESNde7evbvCF7yiFCUiIldXV4qMjCQizaK0ePFi6tmzJ126dKnSdZXs9KFDh8rNK32mVLqw\n7Nq1i1xdXYmo8qJU9hV3yfpKitKECRPo008/FZbt2bNH+IctyXH69GlhuaOjI+3cuVNrO7m5uWm8\nepg2bRpNmDBBZ87SSi9v3bo1Xb16lQYPHkzr168vV5RKy87Oprp161JGRoYwb/PmzWRjY0ONGjWi\n3377Tec269SpQ1u2bNG5nIjoiy++oE6dOhFR8T9mzZo1KTk5WVhe1WNDW1EaN24c9enTR+t2Z86c\nSaamplS/fn1q1aoVjRs3ji5fvkxExUXJ2tpaOD7u3r1LtWrVEp4UlUqlxtlhSkoKOTo6aqw/ODiY\nfvzxRyIicnZ21tiHkJAQnWdKREQnT56kkJAQql+/PrVo0ULj7NHb25t+/vnncvfZsmWL0I5ExWfV\nderUoQcPHhBRcTt+/vnnwvLdu3dTjx49NNbh7u5OSUlJWjPZ29sL/39ExWcUb775JhER3blzhyws\nLIQn/gEDBgiPTVl9+vQRjlkiouTkZKpfvz5ZWVkJbVhZe4aGhgovIIiINmzYILzQquy+ZTk7O9OO\nHTs05lXU/hVtm4iobdu2tGfPHiIiWrJkCfXt25eIil/MmZuba7zgbNGihdCm3333Hb333nvCspJC\n8ddffxFR+cevKo/3ggULhOU9evSg/v37C9NfffUVhYaGEtHzHwulRUREaLSHNkqlksaPHy9Mz507\nV3gRo+35+aeffiKFQkEDBw6k27dv086dO6levXqUnp4u3ObSpUtkZmamc5uiXH2XkZGBK1euoEWL\nFqW7CQEUD3QplUq8+eabaNeuXZWu7PDx8alwubu7u/C3h4cHLl68iJycnBdM//+OHj2KDh06CNMd\nOnTAhQsXNMZUSm+7WbNmWi9UUKvVOH/+fLl1HT58+LkzhYSEYMmSJVCpVHj77bfL9WvHxcWhb9++\neOWVV2Bra4u8vDxcuHBBWP7mm2+isLAQTk5O6Ny5s87ttGjRAkePHq0wy9q1azFo0CAAQMOGDaFU\nKrFmzZoK76Pt2NDm0KFDsLW11bl88ODByMrKwrVr17B06VI4OTkJy1xcXGBiUnxoN2vWDM+ePdMY\n7yp9PO3duxdpaWmwtrYWfvbt24fDhw9DrVbj8uXL5Y6vinTs2BFr165FZmYmvv32W8yYMQP79+9H\nTk4OTp06hS5dupS7z9GjR+Hp6SlM16lTB61bt9Zo/7KZDx8+rJE5NTUVhw4d0pmr7D4cO3YMAPDq\nq6+iS5cu+Omnn/C///0Pv/76K959912t62jRogWOHDmisc6srCxs2bIFT548EbLpak+geDywdJam\nTZvizp07VbpvWXZ2dsJ9S+hq/8q2DQAjRoxAfHw8ACA+Pl644CkpKQmtWrUSxkUAaDxee/fuxQ8/\n/CBkbtSoEXJycjQej9KPX1Uebzc3N+HvJk2aaEw3btxYo82e91go0aJFC5w4caLS21XUZmX3zdLS\nEgDwySefoHnz5ujduzfeeOMNbNy4UbjN7du3YWdnp3N7ohSlhIQEEJFGw5eoU6cOPv30U1y/fh2x\nsbGYNGkSLl26BACoUaOG1sHDmjVrVri95ORk4e8zZ87AxcUFdevWRfPmzZGZmYnCwkIAwIMHD3D7\n9m2N9WrbXokuXbrg1KlTwvSpU6fQrl07oeGrytLSEu3bty+3rm7duj3XegBg+PDhiI6ORt++fTX+\nSQDg1q1bmDRpEqZOnYqbN28iPT0dtWvX1tjHadOmwdnZGX/++afGgVJWt27dsH37dp3Ljx49itTU\nVHz55Zdo1qwZmjVrhmPHjmH9+vU6L1AAtB8bZS/UuH79On7//Xd4eXlpXUdlF8RUpvTxFBgYiJYt\nWyIrK0v4yc7OxuLFi2FpaQknJ6dyx1dVLixRKBTo378/unTpApVKhbp168LLy0vjSb1Ely5dcPr0\naWE6JycH165d03jRUDazUqnUyKxWq/Hxxx/rzFN2H0qvOzQ0FPHx8di0aRM6d+6MZs2aaV2Hn58f\nkpOTyz0plX4sKmpPbbcvLSAgoNL7lta2bVtcv35d67Ky7V/ZtoHi/61t27bh3LlzuHLlCoKDgwEA\nXl5eSE1NRV5ennDbM2fOaOzziBEjNHI/fvwYAwcOFG5T+vGryuNdlq7cL3IslPD390dGRobwAuVF\nld43R0dHKBQK4UVhSfbS/zOpqanCxSnaVEtRKmmwM2fOYMaMGZg9ezYmTJiA1q1baywHgMTERKSm\npqKoqAh169aFqamp8OTaoUMHjQerqtvet28fduzYgT/++ANRUVHo168fAKB169Zo1KgRVq9ejYyM\nDMycOVOjcTp06IBLly4Jr/LKCgoKwoYNG7B//36kpqYiMjISb7/9dpXaQtu6IiMjkZKSApVKhQ0b\nNggH/fNwcHDAoUOH8NVXX5VblpGRASJC06ZNoVarMXXqVI19O3ToEOLi4rBu3TrExcVh/PjxWs/s\nAOCLL77Aw4cPMWDAAOzatQsFBQW4cOEC3n77bTx69Ahr1qxBz549cfnyZZw7dw7nzp3D77//jry8\nPPzyyy/l2qOiY6PkNjdv3sSiRYswePBgdOzYEaNGjdKa7WUKUtn7Ozo6wsLCAlFRUbh37x6ePn2K\nkydP4sqVKwCK37cWFRWFtLQ0JCYmVvh+n4SEBPz3v/9FVlYW8vLysHPnThw5ckR4xT1kyBDMmzcP\nR44cQWFhIY4dO4aCggL06NEDFy9eRGxsLO7fv4/p06fDy8sLDRs21Lqd119/HRcuXMDatWuRlZWF\n/Px8qFSqcsWi9P5u3rwZv/32G86fP4/vvvsOb775prD87bffxpkzZ7B48WKMGDFC5/69++676Ny5\nM4KCgrBy5Uph26dOnRL+ryprz4oeOycnpwrvW1ZgYKDGK/3K2r+y48bGxgYdO3bEiBEjMHDgQJiZ\nmQEoPjtwcXHBzJkzkZGRgQULFmiceb/zzjvYsmULtm7dipycHOTk5GDHjh06L0V/3se7IpUdC7Nm\nzdL5lpHOnTtj2LBhGDVqFJYsWYKsrCzcu3cP8+fPR0xMjHC75/l/s7OzQ69evfD111/j3r172Lt3\nL3bv3o3hw4cLtzlx4gQCAwN1rqNailK/fv1gZWWF4OBgHDlyBFFRUViwYIGwvPS17qmpqejRowfq\n1auH9957D19++SVee+01AMD777+PxMRENGjQQLi/tlelpecpFAp88MEHWLBgAbp27YqAgABMmzZN\nWB4dHY3Y2Fh4e3ujffv2sLGxEZb5+/ujTZs2cHBwQMeOHcttR6lUYuHChZgzZw6Cg4MRFBSEKVOm\naM1R0TwA+Pe//43g4GD0798fX331FRYsWAB/f/9K76dteefOndG0aVNhfskyT09PjB07FoGBgejW\nrRtcXV2FbrLs7GyEhoZi2bJlaNasGfz8/BAeHo6wsDCt22vSpAnOnz+PunXrYuzYsWjUqBFCQkIQ\nEBCAWrVqYdOmTRg/fjwaN24s/Njb2yMkJARr164V1lOVY2PevHmwsrKCt7c31q9fj4EDB+LQoUM6\n26Si905oW1bZ9NatW/H06VN0794dzZo1w6effoqCggIAwMyZM9GlSxf4+fnh22+/1XhPTFnW1taI\niYlBmzZtYGdnh5iYGKxYsUIowGPHjsW4ceMwbdo0NGzYEJ9++qnw4mz//v04ePAgvLy8ULt2bfzw\nww86t1OjRg2oVCpcvXoVHTp0gK2tLebPn1/hJfTjxo3DpEmTEBwcjPDwcIwcOVJYbm5ujv79++PG\njRvl3gtXdj379u1Dnz59sHz5ctjb28PJyQmnT5/Gjz/+WKX2rOzxqei+ZfXt2xcPHz4Ueloqa/+q\nHBuhoaG4cOFCufcqbtq0CZmZmXB1dcWVK1cwePBgYZm1tTV27dqFAwcOoE2bNmjdujXWrl2r8xh9\n3se7bM7S+6HrWCgpJLdu3YKfn5/O9a5btw5jxoxBXFwc7O3t4ePjg6tXr+KNN96odNtll5X44Ycf\n8OzZM7Rr1w6LFi3CunXrhOerzMxM7N+/H6NHj9a9sxWOcr2EUaNGUePGjYWLDoiIJk+eTE5OTuTh\n4UEffvgh5ebmCssWLVpErVq1orZt29Lhw4fFisUY02L27NnlrkSVg/j4eJ1Xab6IgwcPkp2dXbWt\nz9Dc3d0pMzPT0DEE8+bNo9mzZ1d4GwWRON88e/jwYVhYWGDEiBHCQPuePXvQvXt3AMCYMWPg6+uL\n8PBw3L9/H926dcPu3buRlpaGiRMnavTZMsbE8/DhQ7Rv3x47d+5E+/btDR3HYJ4+fYrg4GD07t0b\n48ePN3QcoyXaZ9917doV1tbWGvN69OgBExMTmJiYoFevXjh48CCA4j7G3r17w9bWFv7+/iAijSvc\nGGPiiImJgaOjI8LCwoy6IF2+fBkNGzZEjRo1EBERYeg4Rq3iy9pEFBMTI/QrJiUloW3btsIyR0dH\nJCUlCWdVjDFxvPfee3jvvfcMHcPg2rZti+zsbEPHYDBQUfr8889haWkpvMdFWw+itgG0Ro0a4eHD\nh6LnY4wxY9KyZUukpqYaOgYAA3yfUlxcHHbt2iW8SQ0ofvNVyRU0AHDlyhWt71F5+PChcAmx1H9C\nQ0MNnoGzctZ/Uk7OKt6Prvd7GYJei9Kvv/6KyMhIJCQkaLzx09vbG7t27UJ6errwNQ3P+wZVqbG3\ntzd0hCrjrOKQS1a55AQ4qzEQrftu6NChOHjwIB48eIAWLVpg9uzZmDt3LgoKCvD6668DADp16oTl\ny5ejSZMmiIiIQGBgIExNTbFy5UqxYjHGGJMw0YrShg0bys3T9UZNAPjwww/x4YcfihVH7+rXr2/o\nCFXGWcUhl6xyyQlwVmOg9zElY1H6QwyljrOKQy5Z5ZIT4KzGQLQ3z4rhZT+IkzHGWHlSem7lMyXG\nGGOSwUVJJKU/Ll/qOKs45JJVLjkBzmoMuCgxxhiTDB5TYowxIyel51Y+U2KMMSYZXJREIqf+ZM4q\nDrlklUtOgLMaAy5KjDHGJIPHlBhjzMhJ6bmVz5QYY4xJBhclkcipP5mzikMuWeWSE+CsxoCLEmOM\nMcngMSXGGDNyUnpu5TMlxhhjksFFSSRy6k/mrOKQS1a55AQ4qzHgosQYY0wyZDem5Obm/0L3NTEB\nFi6cDX//F7s/Y4z9U0lpTEm0r0MXy7lzs17ofqam3+L333/nosQYYxImw+475Qv9mJg011dAAPLq\nT+as4pBLVrnkBDirMZBhUWKMMfZPJbsxJeDF4pqbj0NUlDPGjRtXvaEYY0zmpDSmxGdKjDHGJIOL\nkkjk1J/MWcUhl6xyyQlwVmPARYkxxphk8JgSY4wZOR5TYowxxrTgoiQSOfUnc1ZxyCWrXHICnNUY\ncFFijDEmGaIVpbCwMDRp0gTt2rUT5qnVagQFBcHW1hbBwcF4/PixsGzx4sVo3bo1nJ2dceTIEbFi\n6Y1SqTR0hCrjrOKQS1a55AQ4qzEQrSiNGjUKv/76q8a86Oho2Nra4tq1a7CxscGKFSsAAPfv38fy\n5cuxb98+REdHY8KECWLFYowxJmGiFaWuXbvC2tpaY15SUhLCw8NhZmaGsLAwnDhxAgBw4sQJ9O7d\nG7a2tvD39wcRQa1WixVNL+TUn8xZxSGXrHLJCXBWY6DXMaWTJ0/CyckJAODk5ISkpCQAxUWpbdu2\nwu0cHR2FZYwxxoyHXr+64nmugy9+T5I2IwHY//13fQDuKP4kcABQ/f1b+3RKSgpUKpXQ11vySkaM\naaVSKer6jXm6hFTy6JoumSeVPP+U6RJSySPHx1+lUiEuLg4AYG9vD0khEaWlpZGrq6sw3b9/fzpz\n5gwREZ06dYoGDBhAREQJCQk0YcIE4XZubm6UnZ1dbn0ACKAX+jE3H0tLly4Vc3cZY0yWRC4Fz0Wv\n3Xc+Pj6IjY1FXl4eYmNj4evrCwDw9vbGrl27kJ6eDpVKBRMTE1haWuozWrUr+6pOyjirOOSSVS45\nAc5qDEQrSkOHDkXnzp2RkpKCFi1aYPXq1YiIiEB6ejocHR1x584dvP/++wCAJk2aICIiAoGBgRg7\ndiwWLVokVizGGGMSxp99xxhjRo4/+44xxhjTgouSSOTUn8xZxSGXrHLJCXBWY8BFiTHGmGTwmBJj\njBk5HlNijDHGtOCiJBI59SdzVnHIJatccgKc1RhwUWKMMSYZPKbEGGNGjseUGGOMMS24KIlETv3J\nnFUccskql5wAZzUGXJQYY4xJBo8pMcaYkeMxJcYYY0wLLkoikVN/MmcVh1yyyiUnwFmNARclxhhj\nksFjSowxZuR4TIkxxhjTgouSSOTUn8xZxSGXrHLJCXBWY8BFiTHGmGTwmBJjjBk5HlNijDHGtOCi\nJBI59SdzVnHIJatccgKc1RhwUWKMMSYZPKbEGGNGjseUGGOMMS24KIlETv3JnFUccskql5wAZzUG\nXJQYY4xJBo8pMcaYkTP6MaWYmBh07twZHTp0wEcffQQAUKvVCAoKgq2tLYKDg/H48WNDRGOMMWZA\nei9KmZmZmDNnDvbs2YOTJ08iJSUFu3btQnR0NGxtbXHt2jXY2NhgxYoV+o5WreTUn8xZxSGXrHLJ\nCXBWY6D3olS7dm0QER49eoS8vDzk5uaifv36SEpKQnh4OMzMzBAWFoYTJ07oOxpjjDEDM8iY0s6d\nOxEUFAQzMzNMmDABX331Fezs7HD16lWYm5sjNzcXbdu2xc2bNzXD8pgSY4xVOymNKdXU9wYzMjIQ\nERGBS5cuwdraGoMGDUJiYuJzNMhIAPZ//10fgDsA5d/Tqr9/a59OSUmBSqWCUlk8XXJ6zdM8zdM8\nbUzTKpUKcXFxAAB7e3tICulZYmIiDR48WJhevnw5/fvf/6b+/fvTmTNniIjo1KlTNGDAgHL3BUAA\nvdCPuflYWrp0qd7288CBA3rb1svirOKQS1a55CTirGIxQCnQSe9jSl27dsWpU6eQmZmJJ0+eYOfO\nnejZsyd8fHwQGxuLvLw8xMbGwtfXV9/RGGOMGZhBxpTi4uKwevVq5Obmonfv3pg9ezZycnIwfPhw\nJCcnw9PTE/Hx8bCwsNAMy2NKjDFW7aQ0psRvnmWMMSMnpaLEHzMkkpJBRTngrOKQS1a55AQ4qzHg\nosQYY0wyuPuOMcaMHHffMcYYY1pwURKJnPqTOas45JJVLjkBzmoMKi1KISEhVZrHGGOMvaxKx5Q8\nPDyQnJwsTOfm5qJr1644ffq06OHK4jElxhirfrIYU5ozZw4sLS1x4cIFWFpaCj+urq4YPny4PjMy\nxhgzEjqL0tSpU6FWqzF58mSo1Wrh548//sDEiRP1mVGW5NSfzFnFIZescskJcFZjUOmnhH/99dd4\n+vQpkpOTkZ+fL8zv1q2bqMEYY4wZn0rHlBYvXozIyEg4OzvD1NRUmL99+3bRw5XFY0qMMVb9pDSm\nVOmZ0nfffYdLly7B0tJSH3kYY4wZsUovCbe1tcXjx4/1keUfRU79yZxVHHLJKpecAGc1BpWeKVlZ\nWcHd3R2B5pY2AAAcTklEQVQ9e/ZE/fr1ARSf6i1evFj0cIwxxoxLpWNKJV+Zq3EnhQKhoaFiZdKJ\nx5QYY6z6yWpMaeTIkXqIwRhjjFVhTMnBwaHcz2uvvaaPbLImp/5kzioOuWSVS06AsxqDSs+UTp48\nKfydmZmJuLg4vhKPMcaYKJ77+5SKiorg5uaGCxcuiJVJJx5TYoyx6ierMaXTp0//XQyA/Px8HDx4\nEHZ2dqIHY4wxZnwqHVP6+OOPhZ8vvvgCT58+xZIlS/SRTdbk1J/MWcUhl6xyyQlwVmNQ6ZkSNyxj\njDF9qXRMSa1WIz4+HgkJCQCAoKAgvPvuuwa52IHHlBhjrPpJaUyp0u67qKgoXLhwAZ9//jlmz56N\nCxcuICoqSh/ZGGOMGZlKi9K2bduwePFieHl5wdvbG4sWLcK2bdv0kU3W5NTtyVnFIZescskJcFZj\nUGlR8vT0xObNm0FEKCoqws8//wxPT099ZGOMMWZkKh1TunbtGv7zn/8Ib6L19vbG119/jdatW+sl\nYGk8psQYY9VPSmNKlV5917p1a2zZsgUFBQUAoPFFf4wxxlh10tl9N3/+fCxatEiYNjU1hampKZYs\nWYJvv/32pTaak5OD0NBQtGnTBs7Ozjhx4gTUajWCgoJga2uL4OBg2X+Hk5z6kzmrOOSSVS45Ac5q\nDHQWpdjYWIwZM6bc/H/961/4/vvvX2qjM2fOhK2tLc6fP4/z58/DyckJ0dHRsLW1xbVr12BjY4MV\nK1a81DYYY4zJj86i9PTpU5ibm5ebb2pqiidPnrzURvfu3YupU6fC3NwcNWvWRL169ZCUlITw8HCY\nmZkhLCwMJ06ceKltGJpSqTR0hCrjrOKQS1a55AQ4qzHQWZSsrKywa9eucvP37t0LCwuLF97g7du3\nkZ+fj4iICPj4+OCbb75BXl4eTp48CScnJwCAk5MTkpKSXngbjDHG5ElnUZo+fTomTZqE5cuX448/\n/sD169exbNkyTJw4EdOnT3/hDebn5yMlJQUDBgyASqXCxYsX8eOPP0rmyo/qIqf+ZM4qDrlklUtO\ngLMaA51X3wUHB+O1117Dpk2b0K9fPwDAgAED8MMPP8DNze2FN9iqVSs4OjoK6xw6dCjWrl0LLy8v\nXL58GR4eHrh8+TK8vLx0rGEkAPu//64PwB2A8u9p1d+/tU+npKRApVIJp9UlB42xT5eQSp6Kps+e\nPSupPBVNnz17VlJ5+PHX77SUH3+VSoW4uDgAgL29PaTkub9PqTq89dZbmDZtGry8vDBhwgR4eHjg\n4cOHuHXrFubNm4fJkyfDwcEBkydP1gzL71NijLFqJ6v3KYkhKioKI0aMQH5+Pl5//XUMGTIERUVF\nGD58OBwdHeHp6YlvvvnGENEYY4wZUKUfMySGNm3a4Pjx4zh79iyioqJQt25dWFpaYtu2bUhPT8fW\nrVtf6mIKKSjbNSJlnFUccskql5wAZzUGBilKjDHGmDaVjik9fPgQq1atwm+//YaEhARcunQJx44d\nQ3h4uL4yCnhMiTHGqp+UxpQqPVP67LPPYGlpiRs3bgAo/iy8hQsXip2LMcaYEaq0KJ06dQpjx45F\njRo1AAA1a9YU/ma6yak/mbOKQy5Z5ZIT4KzGoErfp3Tr1i1hesuWLejatauooRhjjBmnSseUrl69\niilTpuDw4cNo2LAhHBwcsHz5cv4+JcYY+4eQ0phSpe9TcnR0REJCAu7fv4+ioiI0bdpUH7kYY4wZ\noSpdEn779m0cOHAAu3fvxtq1a7F27Vqxc8menPqTOas45JJVLjkBzmoMKj1TmjZtGhISEtC5c2eN\nb50dMWKEqMEYY4wZn0rHlJydnZGcnAwzMzN9ZdKJx5QYY6z6SWlMqdLuu/bt2wvvUWKMMcbEVGlR\nysjIQLt27eDn54d+/fqhX79+eOutt/SRTdbk1J/MWcUhl6xyyQlwVmNQ6ZjSjBkzys0r7kZjjDHG\nqleVvk+poKAAx48fR7du3ZCbm4tnz57ByspKH/k08JgSY4xVP1mNKW3ZsgW+vr4YNWoUgOLLw99+\n+23RgzHGGDM+lRal5cuX4/Dhw8KZUZs2bXD//n3Rg8mdnPqTOas45JJVLjkBzmoMKi1KCoUCderU\nEaYzMjLQsGFDUUMxxhgzTpWOKcXExODKlStITEzE1KlTsXbtWgwbNoy/T4kxxv4hpDSmVOnVd6NH\nj8bBgwdRUFCApKQkfP755+jSpYs+sjHGGDMyVeq+UyqVWLJkCZYtW8YFqYrk1J/MWcUhl6xyyQlw\nVmOgsyht3boVUVFRwrS3tzccHBzg4OCATZs26SUcY4wx46JzTMnX1xfx8fFo1aoVAMDd3R379u1D\nTk4ORo4cif379+s1KMBjSowxJgZZjCllZ2cLBQkA/Pz80LBhQzRs2BCPHz/WSzjGGGPGRWf33ZMn\nT/Dnn38K00uXLgUA3L17FxkZGeInkzk59SdzVnHIJatccgKc1RjoLEoBAQFYsmSJxjwiwtKlSxEY\nGCh6MMYYY8ZH55hSVlYW3nvvPZw+fRp+fn4AgEOHDqFjx45YtWoVrK2t9RoU4DElxhgTgyzGlKyt\nrfHTTz8hJycHO3bsgEKhQHR0NCwsLPSZjzHGmBGp9H1KdevWxTvvvINBgwZxQXoOcupP5qzikEtW\nueQEOKsxqLQoMcYYY/pSpe9TEkNhYSE6duwIGxsbbN++HWq1GsOHD0dycjI8PT0RHx9f7syMx5QY\nY6z6SWlMyWBnSosWLYKzs7PwLbbR0dGwtbXFtWvXYGNjgxUrVhgqGmOMMQMxSFG6ffs2fvnlF4we\nPVqozklJSQgPD4eZmRnCwsJw4sQJQ0SrNnLqT+as4pBLVrnkBDirMTBIUZo4cSIiIyNhYvL/mz95\n8iScnJwAAE5OTkhKSjJENMYYYwZU6VdXVLfExEQ0btwYHh4eGq8kqt6fORKA/d9/1wfgDkD593TJ\n+rRPp6SkQKVSQaksni7ZvhjTSqVS1PUb83QJqeTRNV0yTyp5/inTJaSSR46Pv0qlQlxcHADA3t4e\nUqL3Cx2mTp2KdevWoWbNmsjPz0d2djb69++P3NxcTJ8+HR4eHjh9+jTmzp2Ln376STMsX+jAGGPV\nzqgvdJgzZw5u3bqFtLQ0bNy4EYGBgVi3bh18fHwQGxuLvLw8xMbGwtfXV9/RqlXZV3VSxlnFIZes\ncskJcFZjYPD3KZVcfRcREYH09HQ4Ojrizp07eP/99w2cjDHGmL4Z7H1KL4K77xhjrPoZdfcdY4wx\npgsXJZHIqT+Zs4pDLlnlkhPgrMaAixJjjDHJ4DElxhgzcjymxBhjjGnBRUkkcupP5qzikEtWueQE\nOKsx4KLEGGNMMnhMiTHGjByPKTHGGGNacFESiZz6kzmrOOSSVS45Ac5qDLgoMcYYkwweU2KMMSPH\nY0qMMcaYFlyURCKn/mTOKg65ZJVLToCzGgMuSowxxiSDx5QYY8zI8ZgSY4wxpgUXJZHIqT+Zs4pD\nLlnlkhPgrMaAixJjjDHJ4DElxhgzcjymxBhjjGnBRUkkcupP5qzikEtWueQEOKsx4KLEGGNMMnhM\niTHGjByPKTHGGGNacFESiZz6kzmrOOSSVS45Ac5qDLgoMcYYkwweU2KMMSNn1GNKt27dQkBAAFxc\nXKBUKrF+/XoAgFqtRlBQEGxtbREcHIzHjx/rOxpjjDED03tRqlWrFhYuXIiLFy/ip59+wvTp06FW\nqxEdHQ1bW1tcu3YNNjY2WLFihb6jVSs59SdzVnHIJatccgKc1RjovSg1bdoU7u7uAIBGjRrBxcUF\nJ0+eRFJSEsLDw2FmZoawsDCcOHFC39EYY4wZmEHHlFJTU9GzZ0+cP38eLi4uuHr1KszNzZGbm4u2\nbdvi5s2bGrfnMSXGGKt+UhpTqmmoDavVagwePBgLFy6EhYXFczTISAD2f/9dH4A7AOXf06q/f2uf\nTklJgUqlglJZPF1yes3TPM3TPG1M0yqVCnFxcQAAe3t7SAoZQEFBAfXo0YMWLlwozOvfvz+dOXOG\niIhOnTpFAwYMKHc/AATQC/2Ym4+lpUuX6m0fDxw4oLdtvSzOKg65ZJVLTiLOKhYDlQKt9D6mREQI\nDw+Hq6srPvroI2G+j48PYmNjkZeXh9jYWPj6+uo7GmOMMQPT+5jSkSNH0K1bN7Rv3/7vMSJg7ty5\n6NKlC4YPH47k5GR4enoiPj4eFhYWmmF5TIkxxqqdUY8p+fn5oaioSOuybdu26TkNY4wxKeGPGRJJ\nyaCiHHBWccglq1xyApzVGHBRYowxJhn82XeMMWbkpDSmxGdKjDHGJIOLkkjk1J/MWcUhl6xyyQlw\nVmPARYkxxphk8JgSY4wZOR5TYowxxrTgoiQSOfUnc1ZxyCWrXHICnNUYcFFijDEmGTymxBhjRo7H\nlBhjjDEtuCiJRE79yZxVHHLJKpecAGc1BlyUGGOMSQaPKTHGmJHjMSXGGGNMCy5KIpFTfzJnFYdc\nssolJ8BZjQEXJcYYY5LBY0qMMWbkeEyJMcYY04KLkkjk1J/MWcUhl6xyyQlwVmPARYkxxphk8JgS\nY4wZOR5TYowxxrTgoiQSOfUnc1ZxyCWrXHICnNUYcFFijDEmGTymxBhjRo7HlIyQlVUDKBSKF/qx\nsmpg6PgvjPfbuPabsZfFRUkkZfuT1eosFJ/lPf9P8X31l7U6Vfd+y6Wfvjj7AUjx8S5LLm0KcFZj\nIKmidOjQIbRt2xatW7fGkiVLqn39H3/8yQu/elUoTJ/r9gEBARrTUnb27FlDR6gyOWUF5JFVTm3K\nWf/5JFWUPvzwQ6xcuRJ79+7FsmXL8ODBg2pd/5Mnj/Gir9qBp895+5llpqXrf//7n6EjVJmcsgLy\nyCqnNuWs/3ySKUqPHj0CAHTr1g12dnbo2bMnTpw4YeBUUlHzJc7wKj/Lmz17tkTHN8rvd0VZX/bs\ntvruK+0z44q8zFjYy7aZXMfSdLVZVY9Vue63WGoaOkCJkydPwsnJSZh2dnbG8ePH0bdvX43bWVn1\ne6H1FxRceKl8z+9GNa7rGV7ubKuyqxZHAojTukStNuQTrLb9HgldWct78as1X+6+Jfe/8RL3158b\nN24If///GOCLeLk2q8qxVjqrVOhus5GoyrFq2P8x6ZFMUaqKli1b4vr1xJdcy8scAM973zUG3Pbz\n3r9s1lL3fOlX/tW937qzir/t57EGz5e11Jb1fLa1Zk3pnIZrs6rst2ZWqdCVu2pZDX123bJlS4Nu\nvzTJFCUvLy9MmTJFmL548SJ69+6tcZvU1FR9x2KMMaZHkhlTqlevHoDiK/Bu3LiBPXv2wMfHx8Cp\nGGOM6ZNkzpQA4Ntvv8WYMWPw9OlTTJgwAY0aNTJ0JMYYY3okq48ZYowx9s8mme67yoj9xtrqZG9v\nj/bt28PDwwPe3t6GjqMhLCwMTZo0Qbt27YR5arUaQUFBsLW1RXBwMB4/fmzAhMW05Zw1axZsbGzg\n4eEBDw8P/PrrrwZM+P9u3bqFgIAAuLi4QKlUYv369QCk2a66skqtbfPz8+Hj4wN3d3f4+vpi4cKF\nAKTZprqySq1NSyssLISHhwf69Su+mllS7Uoy4e7uTgcPHqQbN26Qo6MjZWRkGDqSTvb29vTw4UND\nx9Dq0KFDdObMGXJ1dRXmffPNN/TBBx9Qfn4+jRs3jiIjIw2YsJi2nLNmzaL58+cbMJV2f/75JyUn\nJxMRUUZGBjk4OFB2drYk21VXVim2bU5ODhER5efnk4uLC6WkpEiyTYm0Z5Vim5aYP38+DRs2jPr1\n60dE0noOkMWZkhzfWEsS7RXt2rUrrK2tNeYlJSUhPDwcZmZmCAsLk0TbassJSLNdmzZtCnd3dwBA\no0aN4OLigpMnT0qyXXVlBaTXtnXq1AEAPH78GM+ePYOZmZkk2xTQnhWQXpsCwO3bt/HLL79g9OjR\nQj4ptassipKuN9ZKlUKhQGBgIIKDg5GQkGDoOJUq3b5OTk5ISkoycCLdlixZAl9fX3zzzTdQq9WG\njlNOamoqLl68CG9vb8m3a0nWkqtcpda2RUVFcHNzQ5MmTfDBBx/A1tZWsm2qLSsgvTYFgIkTJyIy\nMhImJv//9C+ldpVFUZKb3377DefOncPcuXMxadIk3Lt3z9CRKiTFV3PaREREIC0tDbt27cL169ex\ncuVKQ0fSoFarMXjwYCxcuBAWFhaSbtfSWevWrSvJtjUxMcG5c+eQmpqK5cuXIzk5WbJtqi2rFNs0\nMTERjRs3hoeHh0ZbSqldZVGUvLy8cOXKFWH64sWL8PX1NWCiijVr1gwA0LZtW7z11lvYvn27gRNV\nzMvLC5cvXwYAXL58GV5eXgZOpF3jxo2hUChQr149jBs3Dj///LOhIwmePn2KAQMGICQkBEFBQQCk\n267askq5be3t7dGnTx+cOHFCsm1aonRWKbbp0aNHkZCQAAcHBwwdOhT79+9HSEiIpNpVFkVJTm+s\nzc3NFU7TMzIysGvXrnKfTCE1Pj4+iI2NRV5eHmJjYyVb8P/8808AwLNnz7B+/Xr06dPHwImKERHC\nw8Ph6uqKjz76SJgvxXbVlVVqbfvgwQPhU7YfPnyI3bt3IygoSJJtqiur1NoUAObMmYNbt24hLS0N\nGzduRGBgINatWyetdjXM9RXPT6VSkZOTE7Vs2ZIWLVpk6Dg6/fHHH+Tm5kZubm4UGBhI33//vaEj\naRgyZAg1a9aMTE1NycbGhmJjYyk7O5veeustatGiBQUFBZFarTZ0TCFnrVq1yMbGhr7//nsKCQmh\ndu3aUYcOHWjixImSucLx8OHDpFAoyM3Njdzd3cnd3Z127twpyXbVlvWXX36RXNueP3+ePDw8qH37\n9tSzZ09as2YNEZEk21RXVqm1aVkqlUq4+k5K7cpvnmWMMSYZsui+Y4wxZhy4KDHGGJMMLkqMMcYk\ng4sSY4wxyeCixBhjTDK4KDHGGJMMLkpMb0xMTBASEiJMP3v2DK+88orw8fliWrlyJdq1awcnJye4\nu7tj3bp1wrIHDx6gVq1a5T4GpuQrSJycnODm5oaoqCg8ffoUAHDjxg3Url0bnp6eaN26Nfz8/LB1\n61at21apVKhXrx48PDzg5+eHmJgYAEBcXBzGjx9fbfuoa31ZWVkYO3YsPD094eLigr59+1bbNhmr\nbpL65ln2z1a3bl1cvHgR+fn5MDc3x549e2BjYwOFQiHqdhMSErB27Vp899136NSpE/73v//hv//9\nr7B806ZN6N27NzZs2IAxY8YI8xUKBVQqFRo0aIDffvsNc+fOxe3bt/Htt98CAFq1aoUzZ87gyZMn\n2LFjBz755BMAQHBwcLkM3bp1w/bt25GZmQk3NzcMHDiwyvtdVFSk8eGZuuha36pVq/DKK6/gzJkz\nAIDff/+9StutSGFhIWrUqPHS62GsLD5TYnrVp08f7NixAwCwYcMGDB06VPgwyLy8PCxYsAD+/v7o\n27cvVCoVgOKzkm7dusHT0xMDBw7EuXPnABSfgXTv3h1DhgyBs7Mzpk2bpnWbmzZtwrRp09CpUycA\nQP369TWKz8aNG/Hll1/i/v37uHPnjtZ1dOnSBV988QWWLVuGvLw8jWVmZmbo378/xowZg3nz5lW4\n/w0aNIC7u7vw1QAPHjxAnz594OrqisWLFwu3s7CwwIwZM+Du7o5jx45h7969GDRoEDp16oQ5c+YI\nt9u9ezc6deoEb29vnD17Vus2L126hJYtWwrTrq6uGvveo0cPuLm54dNPPwUAXLt2DWFhYXB3d8fM\nmTOFj81SKpWYNm0aOnbsiMWLF+Pq1auIiIiAj48Pxo0bh4cPH1a474xVicE+S4IZHQsLCzp//jwN\nHDiQ8vPzyd3dnVQqFb355ptERLR69WrhI6Tu3btH3t7eRESUm5tL+fn5RER0/PhxGjp0KBERHThw\ngGrVqkVXrlyh/Px8cnV1pfT0dI1tPnnyhJo0aUK5ublaM6Wnp5OTkxMREc2YMUPjS9m0fVmjjY0N\nXbhwgdLS0jS+gJCIKCUlhRo0aFBuGwcOHBD28fr169S8eXPKycmh1atXU+PGjenu3buUnZ1NNjY2\nVFBQQERECoVCaIucnBx6/fXXKS8vjwoLC2no0KF0/PhxevbsGdnb21Nqaio9ePCA3N3dafz48eW2\nf/ToUWrWrBn5+fnR4sWLhbZIS0sjNzc3unPnDhERZWVlERHR22+/TRs3bqSnT59SREQELV++nIiI\nlEolDR06lJ48eUJERP369RPae9myZfT1119rbWPGngd33zG9ateuHW7cuIENGzaUG9vYvHkzbty4\ngdWrVwMoHgtJS0tD06ZN8dlnn2Hfvn0oLCzErVu3hPt4e3vD0dERANC5c2ccPXoUgwcPrnKe//73\nvxg4cCAAYNCgQQgLC8OkSZO03paIUFRUpLObrKJlhw8fhqenJ5o3b46lS5cKXwrXs2dP4VPlnZ2d\nkZycDG9vb5iYmGDkyJEAgJ07d+LSpUvCmV5+fj4OHDiAwsJCuLq6CmdBQUFBWs9WOnXqhBs3bmDr\n1q3YsGEDoqKicP36dWzatAlDhgzBq6++CqD4DLKgoAAnT57E5s2boVAoMGrUKHz22WeIiIgAAAwb\nNgympqa4f/8+jhw5grfeegtAcXeevb19hW3NWFVwUWJ699Zbb2Hy5Mk4ePAgMjIyhPlFRUVYtmwZ\nunXrpnH7uLg4PHjwAEeOHEFOTg6aNGkiLCv97bSmpqbIz8/XuK+pqSl69uyJ/fv3ax3g37BhA/76\n6y/Ex8cDKP607OvXr2t0d5U4c+YM7t+/j5YtW2r9jqzExESNL6MsrWvXruW+wkShUKB+/fpa89eu\nXRtWVlZCu/Ts2VMo1iWOHj2qdVvamJqa4p133sE777wDpVKJgwcPAtD9PTpEBIVCUW55SQErLCxE\ngwYNkJycXOUMjFUFjykxvQsLC8OsWbPg4uKiMX/YsGFYuXKlMIZR8oR3584d2NnZwczMDDExMSgq\nKnqu7Q0cOBDffPONMI6TnZ2NmJgYpKSkICcnB7dv30ZaWhrS0tLwySefYP369cJ9iQhEhGPHjmH2\n7Nn44IMPYG5urrH+J0+eYNu2bfj+++8xZcqUKufSVRDK6tevHw4fPix8301mZibS09Ph6+uLixcv\n4o8//sDDhw+RkJCg9UztwIEDyM3NBVA8hpWeni6Mz23cuFEYR8vKyoKpqSm8vb2xefNmPHv2DGvW\nrBG+c6l05mbNmsHBwQGbN28GEeHp06e4dOlSlfedMV24KDG9KXnCbN68OT744ANhXsn8gQMHwtvb\nG7169YKrqytmzpwJAAgNDcWRI0fQrl07FBQUwMLCotw6dU0DxWdmw4YNQ3h4ONq2bQulUom6deti\n48aN6N+/v8ZtBwwYgI0bNwrTAQEBaNu2Ld5//31069YNkZGRwrLr168Ll1lHRkZi7ty5Gk/gpTNp\ny6Vrftn9MDc3R0xMDGbMmIH27dujZ8+euHfvHkxMTLBy5Uq8++67eOONN+Dv7691XadPn4aXlxfc\n3d0xbNgwzJ49G9bW1nBwcMDUqVMxfPhwuLu7Y/78+QCAr7/+Gjt37kTHjh3RqFEjDB8+XGuu5cuX\n48CBA3B3d4eHhweOHTumdfuMPQ/+6grGGGOSwWdKjDHGJIOLEmOMMcngosQYY0wyuCgxxhiTDC5K\njDHGJIOLEmOMMcngosQYY0wy/g8cN4uhh3wLigAAAABJRU5ErkJggg==\n",
       "text": [
        "<matplotlib.figure.Figure at 0x1159158d0>"
       ]
      }
     ],
     "prompt_number": 45
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "fiSSEA_results.hgvs_queries[:10]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 48,
       "text": [
        "[u'chr1:g.866422C>T',\n",
        " u'chr1:g.876664G>A',\n",
        " u'chr1:g.914263G>T',\n",
        " u'chr1:g.897062C>T',\n",
        " u'chr1:g.865625G>A',\n",
        " u'chr1:g.879368C>A',\n",
        " u'chr1:g.879492C>G',\n",
        " u'chr1:g.901994G>A',\n",
        " u'chr1:g.881602C>T',\n",
        " u'chr1:g.977356C>T']"
       ]
      }
     ],
     "prompt_number": 48
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "<center><h1>fiSSEA Functions</h1><center>"
     ]
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "MyVariant.info Query and Parsing"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "\n",
      "def myvariant_post(hgvs_list):\n",
      "    '''\n",
      "    Query and Parser for myvariant.info\n",
      "    Parses Raw Elastic Search results into\n",
      "    pandas dataframe\n",
      "    \n",
      "    Parameters\n",
      "    -------------\n",
      "    hgvs_list: list, required\n",
      "        \n",
      "        \n",
      "    Output\n",
      "    -------------\n",
      "    pandas df:\n",
      "        normalized json of myvariant results\n",
      "    '''\n",
      "\n",
      "    \n",
      "    if type(hgvs_list) == list:\n",
      "        hgvs_list = str(','.join(hgvs_list))\n",
      "        \n",
      "    assert type(hgvs_list) == str\n",
      "    \n",
      "    con = mv.getvariants(hgvs_list, fields='dbnsfp.cadd.phred,dbnsfp.genename')    \n",
      "\n",
      "    \n",
      "#     a = json_normalize(con).dropna(subset=['dbnsfp'])\n",
      "#     mv_df = json_normalize(a['dbnsfp'].values)\n",
      "#     mv_df.index = a['_id']\n",
      "#     mv_df.columns = ['dbnsfp.cadd.phred', 'dbnsfp.genename']\n",
      "    \n",
      "    mv_df = pd.concat(map(json_normalize, con))\n",
      "    mv_df.index = mv_df['_id']\n",
      "    \n",
      "    return mv_df\n",
      "    \n",
      "    "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 7
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def get_hgvs_id(non_ref_df, allele_call='a1'):\n",
      "    '''\n",
      "    This function returns hgvs formated strings for each variant\n",
      "    given a dataframe\n",
      "    '''\n",
      "    hgvs_a1 = 'chr' + non_ref_df['CHROM'].astype(str) + \":g.\" + non_ref_df['POS'].astype(str)  \\\n",
      "                + non_ref_df['REF'] + '>' +  non_ref_df[allele_call]\n",
      "    return hgvs_a1"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 8
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def get_myvariant_snp_annot(vcf_df_mv):\n",
      "    '''\n",
      "    Returns myvariant.info annotations for all snps for\n",
      "    input vcf pandas dataframe. Submits myvariant.info\n",
      "    POST request in snp batches of 1000.\n",
      "    \n",
      "    Parameters\n",
      "    -------------\n",
      "    vcf_df_mv: pandas df; required\n",
      "    \n",
      "        vcf_df_mv must contain the following columns:\n",
      "            vartype1, vartype2, CHROM, REF, POS, a1, a2\n",
      "            *please see pandasVCF docs for column definitions\n",
      "            \n",
      "\n",
      "    Output\n",
      "    -------------\n",
      "    myvariant.info results: pandas df\n",
      "    \n",
      "    '''\n",
      "    \n",
      "\n",
      " \n",
      "    if 'CHROM' not in vcf_df_mv.columns:\n",
      "        vcf_df_mv.reset_index(inplace=True)\n",
      "        \n",
      "    \n",
      "    vcf_df_mv = vcf_df_mv[(vcf_df_mv['vartype1'] == 'snp') | (vcf_df_mv['vartype2'] == 'snp')] #restrict to snps\n",
      "\n",
      "    non_ref_a1 = vcf_df_mv[vcf_df_mv['REF'] != vcf_df_mv['a1']][['CHROM', 'POS', 'REF', 'a1']]  #restrict a1 to non-reference alleles\n",
      "    hgvs_a1 = get_hgvs_id(non_ref_a1, 'a1')\n",
      "    \n",
      "    non_ref_a2 = vcf_df_mv[vcf_df_mv['REF'] != vcf_df_mv['a2']][['CHROM', 'POS', 'REF', 'a2']]  #restrict a2 to non-reference alleles\n",
      "    hgvs_a2 = get_hgvs_id(non_ref_a2, 'a2')\n",
      "\n",
      "    hgvs_ids = set(hgvs_a1.values) | set(hgvs_a2.values)  #set of non-reference a1 and a2 snps\n",
      "\n",
      "    myvariant_results = myvariant_post(list(hgvs_ids))\n",
      "        \n",
      "    return myvariant_results"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 9
    },
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "fiSSEA Function"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "class fiSSEA(object):\n",
      "\n",
      "    \n",
      "    def __init__(self, vcf_file_path, sample_id='', chunksize=1000, fiSSEA_score='dbnsfp.cadd.phred', gene_stat=np.max, n_cores=1):\n",
      "        \n",
      "        \n",
      "        #setting sample_id\n",
      "        self.sample_id = sample_id\n",
      "        \n",
      "        #setting fiSSEA_score\n",
      "        self.fiSSEA_score = fiSSEA_score\n",
      "        \n",
      "        #setting gene statistic\n",
      "        self.gene_stat = gene_stat\n",
      "        \n",
      "        #setting n_cores\n",
      "        self.n_cores = n_cores\n",
      "        \n",
      "        #open Vcf object\n",
      "        vcf_df_obj = Vcf(vcf_file_path, sample_id, cols=['#CHROM', 'POS', 'REF', 'ALT', 'FORMAT'], chunksize=chunksize, n_cores=n_cores)\n",
      "\n",
      "        #setting hgvs_list\n",
      "        self.hgvs_queries = []\n",
      "        \n",
      "        res = []\n",
      "        \n",
      "        stopiterating = False\n",
      "        while stopiterating == False:\n",
      "            \n",
      "            vcf_df_obj.get_vcf_df_chunk()\n",
      "            if vcf_df_obj.stopIteration == True: break\n",
      "            \n",
      "            vcf_df_obj.add_variant_annotations(inplace=True, split_columns='' )\n",
      "            \n",
      "            \n",
      "            vcf_df = vcf_df_obj.df[['a1', 'a2', 'vartype1', 'vartype2']]\n",
      "            self.vcf_df = vcf_df\n",
      "            \n",
      "            #vcf_df = vcf_df.append(vcf_df_obj.df[['a1', 'a2', 'vartype1', 'vartype2']])\n",
      "            #remove empty variant lines\n",
      "            #vcf_df_obj.df = vcf_df_obj.df[vcf_df_obj.df.ALT!='.']\n",
      "    \n",
      "            #replace chr if necessary, depends on format of annotations\n",
      "            #vcf_df_obj.df['CHROM'] = vcf_df_obj.df['CHROM'].str.replace('chr', '')\n",
      "            \n",
      "            \n",
      "            #CREATE df to pass to myvariant.info\n",
      "            vcf_df.reset_index(inplace=True)\n",
      "            mv_df = vcf_df[['CHROM', 'POS', 'REF','a1', 'a2','vartype1', 'vartype2']]\n",
      "            mv_df.drop_duplicates(subset=['CHROM', 'POS', 'a1', 'a2'], inplace=True)\n",
      "            \n",
      "            \n",
      "            #Retrieving myvariant.info annotations for all snps\n",
      "            mv_df = get_myvariant_snp_annot(mv_df) #query and parse myvariant.info using all snps in self.vcf_df\n",
      "            \n",
      "            #mv_df = pd.concat(myvariant_results)  #NOT NEEDED IF ONLY 1 DATAFRAME  #setting dataframe of parsed myvariant.info results\n",
      "            \n",
      "            if 'dbnsfp.cadd.phred' not in mv_df.columns and 'dbnsfp.genename' not in mv_df.columns: break  #occurs when myvariant results empty\n",
      "            \n",
      "            mv_df = mv_df[['dbnsfp.cadd.phred','dbnsfp.genename']]  #removing unnecessary columns\n",
      "            mv_df.dropna(inplace=True)  #removing useless annotations\n",
      "            mv_df.index.name = 'hgvs_a1'  #renaming index, setting up for merging\n",
      "            mv_df.reset_index(inplace=True)  #setting up for merging on hgvs_a1\n",
      "            self.hgvs_queries.extend( list(mv_df['hgvs_a1'].values)  )  #saving hgvs queries\n",
      "            \n",
      "            \n",
      "            #Merging a1 cadd.phred scores and genename\n",
      "            vcf_df['hgvs_a1'] = get_hgvs_id(vcf_df, 'a1')  #creating hgvs ids for a1 allele calls\n",
      "            mv_df.columns = ['hgvs_a1','dbnsfp.cadd.phred_a1', 'dbnsfp.genename']  #renaming myvariant df columns for merging a1\n",
      "            vcf_df = vcf_df.merge(mv_df, how='left', on=['hgvs_a1'])  #merging cadd scores and genenames to master_df\n",
      "            vcf_df['dbnsfp.cadd.phred_a1'].fillna(value=0, inplace=True)  #filling empty columns with 0\n",
      "            del vcf_df['hgvs_a1']  #removing hgvs_a1 column to save space\n",
      "            \n",
      "            vcf_df['hgvs_a2'] = get_hgvs_id(vcf_df, 'a2')  #creating hgvs ids for a2 allele calls\n",
      "            mv_df.columns = ['hgvs_a2', 'dbnsfp.cadd.phred_a2', 'dbnsfp.genename']  #renaming myvariant df columns for merging a2\n",
      "            vcf_df = vcf_df.merge(mv_df[['hgvs_a2', 'dbnsfp.cadd.phred_a2']], how='left', on=['hgvs_a2'])  #merging cadd scores to master_df\n",
      "            vcf_df['dbnsfp.cadd.phred_a2'].fillna(value=0, inplace=True)  #filling empty columns with 0\n",
      "            del vcf_df['hgvs_a2']  #removing hgvs_a2 column to save space\n",
      "            \n",
      "            vcf_df.dropna(subset=['dbnsfp.genename'], inplace=True)  #dropping variants without gene annotation\n",
      "            \n",
      "            \n",
      "            #Splitting variants intersecting multiple genes into separate rows\n",
      "            vcf_df['dbnsfp.genename_type'] = vcf_df['dbnsfp.genename'].map(type)\n",
      "            vcf_df_multigene = vcf_df[vcf_df['dbnsfp.genename_type']==list]\n",
      "            vcf_df = vcf_df[~vcf_df.index.isin(vcf_df_multigene.index)]\n",
      "            \n",
      "            if len(vcf_df_multigene) > 0:\n",
      "                #vcf_df_multigene['index'] = vcf_df_multigene.index\n",
      "                #vcf_df_multigene.drop_duplicates(subset='index', inplace=True)\n",
      "                vcf_df_multigene = vcf_df_multigene.apply(self.get_multigene_var_records, axis=1)\n",
      "                vcf_df_multigene = [n for i in vcf_df_multigene for n in i]\n",
      "                vcf_df = vcf_df.append(pd.DataFrame(vcf_df_multigene))\n",
      "                del vcf_df_multigene\n",
      "            res.append(vcf_df)\n",
      "            stopiterating = vcf_df_obj.stopIteration\n",
      "        \n",
      "        \n",
      "        vcf_df = pd.concat(res)\n",
      "        \n",
      "        \n",
      "        #CADD SCORES\n",
      "        vcf_df['cadd_score_sum'] = vcf_df['dbnsfp.cadd.phred_a1'] + vcf_df['dbnsfp.cadd.phred_a2']\n",
      "        fi_scores = vcf_df.pivot_table(index=['dbnsfp.genename'], columns='sample_ids', values='cadd_score_sum', aggfunc=self.gene_stat)\n",
      "        fi_scores.fillna(value=0, inplace=True)\n",
      "        \n",
      "        \n",
      "        #settting vcf_df as class object\n",
      "        self.fi_scores = fi_scores\n",
      "    \n",
      "        \n",
      "        \n",
      "        \n",
      "        \n",
      "        #get myvariant.info annotations, create fiSSEA dataframe\n",
      "        #self.fiSSEA_status = self.get_fi_scores()\n",
      "  \n",
      "    \n",
      "    def get_multigene_var_records(self, df_line):\n",
      "        '''\n",
      "        Splits multigene variants into separate\n",
      "        rows\n",
      "        '''\n",
      "        genes = df_line['dbnsfp.genename']\n",
      "        gene_series = [] #list of gene variants\n",
      "        for gene in genes: #iterate through all genes intersecting this variant\n",
      "            df_line_copy = df_line.copy()  #avoids aliasing\n",
      "            del df_line_copy['dbnsfp.genename']  #removing genename to allow gene insertion\n",
      "            df_line_copy['dbnsfp.genename'] = gene\n",
      "            gene_series.append(df_line_copy)\n",
      "\n",
      "        return gene_series\n",
      "    \n",
      "    \n",
      "    \n",
      "    def get_fi_scores(self):\n",
      "        '''\n",
      "        Creates fi_scores series, index on gene symbol, value fi_score.\n",
      "        '''\n",
      "        \n",
      "        #reducing size of dataframe for fiSSEA\n",
      "        fiSSEA_cols = ['dbnsfp.genename', self.fiSSEA_score]\n",
      "        mv_annot_fiSSEA = self.mv_annot[fiSSEA_cols] \n",
      "        \n",
      "        #dropping variants without a genename & filling fi_score NaN with 0\n",
      "        mv_annot_fiSSEA.dropna(subset=['dbnsfp.genename'], inplace=True)  \n",
      "        mv_annot_fiSSEA[self.fiSSEA_score].fillna(value=0)  \n",
      "\n",
      "        \n",
      "        #Splitting variants intersecting multiple genes into separate rows\n",
      "        mv_annot_fiSSEA['dbnsfp.genename_type'] = mv_annot_fiSSEA['dbnsfp.genename'].map(type)\n",
      "        mv_annot_fiSSEA_multigene = mv_annot_fiSSEA[mv_annot_fiSSEA['dbnsfp.genename_type']==list]\n",
      "        mv_annot_fiSSEA = mv_annot_fiSSEA[~mv_annot_fiSSEA.index.isin(mv_annot_fiSSEA_multigene.index)]\n",
      "        \n",
      "        if len(mv_annot_fiSSEA_multigene) > 0:\n",
      "            mv_annot_fiSSEA_multigene['index'] = mv_annot_fiSSEA_multigene.index\n",
      "            mv_annot_fiSSEA_multigene.drop_duplicates(subset='index', inplace=True)\n",
      "            mv_annot_fiSSEA_multigene = mv_annot_fiSSEA_multigene.apply(self.get_multigene_var_records, axis=1)\n",
      "            mv_annot_fiSSEA_multigene = [n for i in mv_annot_fiSSEA_multigene for n in i]\n",
      "            mv_annot_fiSSEA = mv_annot_fiSSEA.append(pd.DataFrame(mv_annot_fiSSEA_multigene))\n",
      "            del mv_annot_fiSSEA_multigene\n",
      "        \n",
      "        gc.collect()\n",
      "\n",
      "        \n",
      "        #Setting gene specific functional impact score\n",
      "        self.fi_scores = mv_annot_fiSSEA.groupby('dbnsfp.genename')[self.fiSSEA_score].aggregate(self.gene_stat)\n",
      "        \n",
      "        return 0\n",
      "\n",
      "\n",
      "    def write_rnk_file(self, single_sample_id):\n",
      "            '''\n",
      "            Writes functional impact scores to .rnk file in\n",
      "            the fiSSEA/gsea/rnk/ directory.\n",
      "    \n",
      "            Returns path to rnk file\n",
      "            '''\n",
      "    \n",
      "            fiSSEA_results_write = self.fi_scores[single_sample_id]\n",
      "            fiSSEA_results_write.name = '#' + single_sample_id + ',file:' + fiSSEA_results_write.name\n",
      "    \n",
      "    \n",
      "            rnk_path = self.rnk_write_dir + single_sample_id + '.rnk'\n",
      "            self.rnk_path = rnk_path\n",
      "            \n",
      "            if not os.path.exists(self.rnk_write_dir):\n",
      "                    os.makedirs(self.rnk_write_dir)\n",
      "            \n",
      "            fiSSEA_results_write.to_csv(self.rnk_path, sep='\\t')\n",
      "            return True\n",
      "        \n",
      "        \n",
      "        \n",
      "    def call_preranked_gsea(self):\n",
      "        '''\n",
      "        Calls GSEA PreRanked Analysis using several default parameters\n",
      "        '''\n",
      "    \n",
      "        gsea_cmd = ['java', '-cp', self.gsea_jar_path, '-Xmx512m', 'xtools.gsea.GseaPreranked', \\\n",
      "                '-gmx', self.tissue_geneset_path, '-collapse false', '-mode Max_probe', \\\n",
      "                 '-nperm 1000', '-rnk', self.rnk_path, '-scoring_scheme classic', \\\n",
      "                '-rpt_label my_analysis', '-include_only_symbols true', \\\n",
      "                '-rpt_label', self.sample_id,  '-chip gseaftp.broadinstitute.org://pub/gsea/annotations/GENE_SYMBOL.chip', \\\n",
      "                '-include_only_symbols true', '-make_sets true', '-plot_top_x 20', \\\n",
      "                '-rnd_seed timestamp', '-set_max 15000', '-set_min 15', '-zip_report false', \\\n",
      "                '-out', self.output_dir_path, '-gui false']\n",
      "    \n",
      "    \n",
      "        p = subprocess.Popen(gsea_cmd,\n",
      "                         stdout=subprocess.PIPE,\n",
      "                         stderr=subprocess.STDOUT)\n",
      "        self.gsea_cmd = gsea_cmd\n",
      "        run_stdout = [\">>> \" + line.rstrip() for line in iter(p.stdout.readline, b'')]\n",
      "        return run_stdout\n",
      "    \n",
      "    \n",
      "    \n",
      "    def run_GSEA_preranked(self, gsea_jar_path, tissue_geneset_path, output_dir_path, single_sample_id):\n",
      "        '''\n",
      "        Calls GSEA jar for a PreRanked analysis.\n",
      "        \n",
      "        Parameters\n",
      "        -------------\n",
      "        gsea_jar_path: str, required\n",
      "            Specifies the path to GSEA jar, e.g. /Users/bin/gsea/gsea2-2.1.0.jar\n",
      "                Can be downloaded from http://www.broadinstitute.org/gsea/downloads.jsp\n",
      "                \n",
      "        rnk_write_dir: str, required\n",
      "            Specifies the directory path where the rnk file should be written\n",
      "                e.g. /Users/data/gsea/rnk/\n",
      "        \n",
      "        tissue_geneset_path: str, required\n",
      "            Specifies the path to the tissue genesets (.gmt format)\n",
      "                e.g.  /Users/data/gsea/genesets/EnrichmentValues_Primary_U133A_ave_percent_over70percent_no_brain.gmt\n",
      "        \n",
      "        output_dir_path: str, required\n",
      "            Specifies the directory path where the GSEA output should be written\n",
      "                e.g. /Users/data/gsea/results/\n",
      "        \n",
      "        Returns\n",
      "        -------------\n",
      "        Writes GSEA PreRanked Results to output_dir\n",
      "        \n",
      "        '''\n",
      "        \n",
      "        #Setting GSEA parameter paths\n",
      "        self.gsea_jar_path = gsea_jar_path\n",
      "        self.rnk_write_dir = output_dir_path\n",
      "        self.tissue_geneset_path = tissue_geneset_path\n",
      "        self.output_dir_path = output_dir_path\n",
      "        \n",
      "        #Writing the rnk file to disk\n",
      "        self.write_rnk_file(single_sample_id)\n",
      "        \n",
      "        #Run PreRanked Gene Set Enrichment Analysis, please see run_preranked_gsea function to change\n",
      "        #parameters\n",
      "        #Returns 0 if successful run\n",
      "        run_stdout = self.call_preranked_gsea()\n",
      "        self.run_stdout = run_stdout\n",
      "\n",
      "        \n",
      "        last_written_dir = os.popen('ls -lht ' + self.output_dir_path).readlines()\n",
      "        last_written_dir = last_written_dir[1].split()[-1]\n",
      "        self.last_written_dir = last_written_dir  #likely the path to the output files\n",
      "        \n",
      "        return 0\n",
      "\n",
      "    \n",
      "    def open_GSEA_html(self):\n",
      "        '''\n",
      "        Polls the user-specified output directory for the last written\n",
      "        file and opens the index.html file in that directory\n",
      "        '''\n",
      "        #Opening GSEA HTML Results Report\n",
      "        os.system('open ' + self.output_dir_path + self.last_written_dir + '/index.html')\n",
      "        return 0\n",
      "\n",
      "    "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 24
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 16
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}