{
 "metadata": {
  "name": "",
  "signature": "sha256:96f68df4b8f8756ba099694b52000fe510a0003ecb4924790e0ae8a65446f72d"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "<center><h1> Functional Impact SNP Set Enrichment (fiSSEA) </h1></center>\n",
      "<center><h4>Authors: Erick R. Scott, Adam Mark, Ali Torkamani, Chunlei Wu & Andrew Su</h4></center>\n",
      "<center><h4>November 2014</h4></center>"
     ]
    },
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "Libraries"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import os, sys, time, gc, argparse\n",
      "import json, httplib2\n",
      "import pandas as pd\n",
      "from pandas.io.json import json_normalize\n",
      "import numpy as np\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1
    },
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "PandasVCF, MyGene.info, MyVariant.info Libraries"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "sys.path.append('../pandasVCF/src/single_sample/')\n",
      "from pdVCFsingle import *\n",
      "\n",
      "sys.path.append('../myvariant/src/')\n",
      "import myvariant\n",
      "mv = myvariant.MyVariantInfo()\n",
      "\n",
      "sys.path.append('../src/')\n",
      "from fiSSEA import *"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 2
    },
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "VCF CDS Variants"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def intersect_cds_vcf(vcf_path, cds_bed, write_path):\n",
      "    '''\n",
      "    Convenience function to create coding snp vcf file\n",
      "    '''\n",
      "    os.system('bedtools intersect -a ' + vcf_path + ' -b ' + cds_bed + ' -wa -header |uniq > ' + write_path )\n",
      "    return write_path"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 3
    },
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "Running fiSSEA "
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "example_vcf = '../data/snyderome_chr6_cds.vcf.gz'\n",
      "sample_id = 'TB0001907'\n",
      "chunk = 100000\n",
      "fi_score = 'dbnsfp.cadd.phred'\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 8
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "\n",
      "fiSSEA_results = fiSSEA(example_vcf, sample_id=sample_id, chunksize=chunk, fiSSEA_score=fi_score)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "input variant rows: 1377\n",
        "dropping 0 variants with genotype call == \".\" \n",
        "current variant rows: 1377\n"
       ]
      }
     ],
     "prompt_number": 9
    },
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "Results"
     ]
    },
    {
     "cell_type": "heading",
     "level": 4,
     "metadata": {},
     "source": [
      "fiSSEA_results object has the following attributes"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "dir(fiSSEA_results)[-8:]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 15,
       "text": [
        "['fiSSEA_score',\n",
        " 'fiSSEA_status',\n",
        " 'fi_scores',\n",
        " 'gene_stat',\n",
        " 'get_fi_scores',\n",
        " 'get_multigene_var_records',\n",
        " 'mv_annot',\n",
        " 'vcf_df']"
       ]
      }
     ],
     "prompt_number": 15
    },
    {
     "cell_type": "heading",
     "level": 4,
     "metadata": {},
     "source": [
      "Original vcf parsed with pandasVCF"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "fiSSEA_results.vcf_df.head()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>CHROM</th>\n",
        "      <th>POS</th>\n",
        "      <th>REF</th>\n",
        "      <th>ALT</th>\n",
        "      <th>ID</th>\n",
        "      <th>FILTER</th>\n",
        "      <th>multiallele</th>\n",
        "      <th>phase</th>\n",
        "      <th>GT1</th>\n",
        "      <th>GT2</th>\n",
        "      <th>a1</th>\n",
        "      <th>a2</th>\n",
        "      <th>zygosity</th>\n",
        "      <th>vartype1</th>\n",
        "      <th>vartype2</th>\n",
        "      <th>GT</th>\n",
        "      <th>DP</th>\n",
        "      <th>GL</th>\n",
        "      <th>GQ</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>0</th>\n",
        "      <td> 6</td>\n",
        "      <td> 348906</td>\n",
        "      <td> G</td>\n",
        "      <td> A</td>\n",
        "      <td> rs11242812</td>\n",
        "      <td>   PASS</td>\n",
        "      <td> 0</td>\n",
        "      <td> /</td>\n",
        "      <td> 0</td>\n",
        "      <td> 1</td>\n",
        "      <td> G</td>\n",
        "      <td> A</td>\n",
        "      <td> het-ref</td>\n",
        "      <td> ref</td>\n",
        "      <td> snp</td>\n",
        "      <td> 0/1</td>\n",
        "      <td> 46</td>\n",
        "      <td> -102.90,-13.86,-83.18</td>\n",
        "      <td> 99</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>1</th>\n",
        "      <td> 6</td>\n",
        "      <td> 350829</td>\n",
        "      <td> G</td>\n",
        "      <td> A</td>\n",
        "      <td>  rs1129085</td>\n",
        "      <td> filter</td>\n",
        "      <td> 0</td>\n",
        "      <td> /</td>\n",
        "      <td> 0</td>\n",
        "      <td> 1</td>\n",
        "      <td> G</td>\n",
        "      <td> A</td>\n",
        "      <td> het-ref</td>\n",
        "      <td> ref</td>\n",
        "      <td> snp</td>\n",
        "      <td> 0/1</td>\n",
        "      <td> 75</td>\n",
        "      <td> -68.57,-22.62,-216.17</td>\n",
        "      <td> 99</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>2</th>\n",
        "      <td> 6</td>\n",
        "      <td> 656143</td>\n",
        "      <td> C</td>\n",
        "      <td> A</td>\n",
        "      <td>  rs1211554</td>\n",
        "      <td>   PASS</td>\n",
        "      <td> 0</td>\n",
        "      <td> /</td>\n",
        "      <td> 1</td>\n",
        "      <td> 1</td>\n",
        "      <td> A</td>\n",
        "      <td> A</td>\n",
        "      <td> hom-alt</td>\n",
        "      <td> snp</td>\n",
        "      <td> snp</td>\n",
        "      <td> 1/1</td>\n",
        "      <td> 40</td>\n",
        "      <td>  -164.63,-12.04,-0.00</td>\n",
        "      <td> 99</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>3</th>\n",
        "      <td> 6</td>\n",
        "      <td> 930792</td>\n",
        "      <td> T</td>\n",
        "      <td> G</td>\n",
        "      <td> rs12202545</td>\n",
        "      <td>   PASS</td>\n",
        "      <td> 0</td>\n",
        "      <td> /</td>\n",
        "      <td> 0</td>\n",
        "      <td> 1</td>\n",
        "      <td> T</td>\n",
        "      <td> G</td>\n",
        "      <td> het-ref</td>\n",
        "      <td> ref</td>\n",
        "      <td> snp</td>\n",
        "      <td> 0/1</td>\n",
        "      <td> 43</td>\n",
        "      <td>  -74.59,-12.96,-92.95</td>\n",
        "      <td> 99</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>4</th>\n",
        "      <td> 6</td>\n",
        "      <td> 965576</td>\n",
        "      <td> A</td>\n",
        "      <td> G</td>\n",
        "      <td>  rs2756293</td>\n",
        "      <td>   PASS</td>\n",
        "      <td> 0</td>\n",
        "      <td> /</td>\n",
        "      <td> 1</td>\n",
        "      <td> 1</td>\n",
        "      <td> G</td>\n",
        "      <td> G</td>\n",
        "      <td> hom-alt</td>\n",
        "      <td> snp</td>\n",
        "      <td> snp</td>\n",
        "      <td> 1/1</td>\n",
        "      <td> 38</td>\n",
        "      <td>  -155.40,-11.44,-0.01</td>\n",
        "      <td> 99</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 12,
       "text": [
        "  CHROM     POS REF ALT          ID  FILTER  multiallele phase  GT1  GT2 a1  \\\n",
        "0     6  348906   G   A  rs11242812    PASS            0     /    0    1  G   \n",
        "1     6  350829   G   A   rs1129085  filter            0     /    0    1  G   \n",
        "2     6  656143   C   A   rs1211554    PASS            0     /    1    1  A   \n",
        "3     6  930792   T   G  rs12202545    PASS            0     /    0    1  T   \n",
        "4     6  965576   A   G   rs2756293    PASS            0     /    1    1  G   \n",
        "\n",
        "  a2 zygosity vartype1 vartype2   GT  DP                     GL  GQ  \n",
        "0  A  het-ref      ref      snp  0/1  46  -102.90,-13.86,-83.18  99  \n",
        "1  A  het-ref      ref      snp  0/1  75  -68.57,-22.62,-216.17  99  \n",
        "2  A  hom-alt      snp      snp  1/1  40   -164.63,-12.04,-0.00  99  \n",
        "3  G  het-ref      ref      snp  0/1  43   -74.59,-12.96,-92.95  99  \n",
        "4  G  hom-alt      snp      snp  1/1  38   -155.40,-11.44,-0.01  99  "
       ]
      }
     ],
     "prompt_number": 12
    },
    {
     "cell_type": "heading",
     "level": 4,
     "metadata": {},
     "source": [
      "Mean functional impact score for each Gene - Input into GSEA"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "fiSSEA_results.fi_scores.head()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 22,
       "text": [
        "dbnsfp.genename\n",
        "AARS2               1.336\n",
        "ABCC10             12.762\n",
        "ACAT2              24.255\n",
        "AIM1                9.095\n",
        "AKAP12              2.952\n",
        "Name: dbnsfp.cadd.phred, dtype: float64"
       ]
      }
     ],
     "prompt_number": 22
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "print fiSSEA_results.fi_scores.describe()\n",
      "fiSSEA_results.fi_scores.hist(bins=20)\n",
      "plt.title('Distribution of Mean CADD Phred Score by Gene (Snyderome, Chr6)')\n",
      "plt.xlabel('Mean CADD Phred Score')\n",
      "plt.ylabel('Gene Count')\n",
      "plt.tight_layout()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "count    330.000000\n",
        "mean       9.893912\n",
        "std        7.623389\n",
        "min        0.000000\n",
        "25%        4.347875\n",
        "50%        9.104750\n",
        "75%       13.976250\n",
        "max       38.000000\n",
        "dtype: float64\n"
       ]
      },
      {
       "metadata": {},
       "output_type": "display_data",
       "png": "iVBORw0KGgoAAAANSUhEUgAAAaYAAAEbCAYAAACRJSCoAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzt3XucHFWZ//HPkAQhkCEJYBKugygCgqIIyqIyKCB4wbsI\ngkFdVtldNassF3U1yiIXdfG2XhA0XFxAxQtsRAI4hbpK/KEJIAh4yYgarkkg4Q6Z+f3xnLKra6p6\naqqq59RJf9+vV79mqrpP1dOnq+vUOU9VNYiIiIiIiIiIiIiIiIiIiIiIiIiIiIiI5PoK8NGalrUD\nsA7oc9MR8O6alg3wI+CYGpdX1H8C9wErPay7lywELgx4+d0wAjzDdxBdcjrwAQ/rjah3v9RUw8Ar\nalzeZ4D31rGgYeARYC2wBvg/4D20Go6JLuvlEywzBLyrxLqgOTuRHbA63DLn+UFs5/G91PznuflD\nXYuss32xhnwNsApYChybes1OWIxfzig/AjyEHWjcD1wDvDX1mgh4FNu+HgRuAE4CNu4Q1yLgcbfc\nVcAS4NnuuY/T3c98vOV/GPiTi+0vwCVdjKWoOhumecDXgb9h7/GPwDdp1f9k2hr4K/C0xLzJqv8q\n+6Um6Qc+B/wZq7M/AGfT2letYOL77OnY/uA+4AHgusRzc4E7gWmdFrBRgZWMAq/B3sAOwBnYjuO8\nCQYbL6tTgza1xDJDsAO2A13V4TX3AS8GZifmzQfuwOptsu0HXIt9AXfGNtTjgUNTr3sH8FvgCLIb\nk+cCM4BdsAblS8DHEs+PAv+CbV9zgQ8Bb8MaxDyjwJluudsB97plw8QOmOre3uYDR2NHmDOAF2KN\ncZ18fke2BH4BbAK8BHuPL8B2PAd7iOdYYDF2kAKTU/9VNWkftzH2Hd8NeCVWZ/thB5H7lFjeFPf3\nHGAmsCswC1iQeM3dwG3A4eVCbslqMfcB1gO7u+lFwKnu/62A/6V1lP1TbGdxoSvzCNYynwAMYEdz\n78Ja7AjY0c2LG80h4FPY0fqDwA+wNwvW0/hLKrZhbMM8FNtgn3DrW+aej2h1wfuwIchh4B7gfGwH\nSSK2d7jY7sOOxvJsAVyA7SSHgY+45R/k3vN6F8c3MsrG7+PLwD+7eVOwo8H/oL3HtCtwNVa3twFv\nSTz3avc+H8SOSj6eeG6i7+fnwBc7PA/2/v4AvBFrQN+Uej7rSP1NWA8p/gyHGDsksj3wsHs/Wb4J\nfDIx/WqsbsF6yZdin+VarNHcO/HaYeBE4CYXx0bYAcEvsG12OXBA4vU7YTvetVjP7Ivk95i+iB1t\n5pntYv8bsBr4fuK544DfY5/rD7GeSWwE2y5+j/VQwA4Wl9Maxdizw3pHgPe5svcBZ2Gf3cYujj0S\nr306VvdZvfv/pPU96qRTfUbYZ/dzrE6vSq2rU9m0a4GjEtPj1X+ndS8G/jX1+puA17n/D8a+bw+4\n9US0b7fvAm7F6vPH2MFoLOvzG+/zPt49v9bFvDPwS7f+S2jvcUxkW0j6R6yhmN7hNSuwg8UbE+uO\ne6iD2D7qROAu7Du3K7b/2bzDMj9M9n5wQvK6cn/GhvSgfUdxOpZzmuIe+3dY1gD2ISwCNsXecDwv\nbpgi7M3vjlXgd2ntGAYZ2zAl1/FxrLFISnbB34V9+APAZsBlidfHcXzNxfVc4DGs4rNcgO1oNsMa\n19sT6zkgI86k+H3sB1zv5r0K28DfTath2sy9bj5WP3thO5rdEut5jvt/T2yji79YE3k/04Gn6LxT\nAHgp1iBsApwGXJ56PqthmgY8iR2hQf6QyHVY7zzLN2kdCG0O/A+t4YKFWINzKLbz/RT2hY4NA78B\ntsXqYVvsCDHuCR7kpuMd1i+xcfFp7v2uZew2FXs7tqM5ATtan5J6fjFwMXYQM9UtD2x7vQ/7PDcG\nvkD78McIthOd6WJ+PnYgtY97j+/Atvu84c8RbCc+E2v0b6e1U/1v2uv5A9iOMsv1tPd2s4xXnxH2\nnXsmtt0MYfuMTmW3ylnXvbQfdIxX/53W/RZa3z2wYfT7sc9pK+xzf6Nb5gJsG46329e55T4b+15+\nBGsgYunPr8jn/X1s294dO8D+CfYd7gduwT5zmPi2kHQJ9l3qZBirl7nYweSttPb7g1g9nI59PzZx\n678J+C/3Hm/C6i3pjcCvC8TXUV7D9EvgFPd/smH6BNar2bnAsgawD2EgY166xxTbDfug+hi/YVrI\n2KPb5I7wWtoTcbtgPayNEnFsk3h+KTZklTbFxZTcyf8TrQYlK86k5PN3uDguAY6kvWE6AuuBJn2N\n/J3F57ANBCb2frZ1r92lQ8wA52KNAtiX7Als3D+Wl9u4C3tvkN8wXYy9tyyLsMZnjVvWD7CeDdhn\nviTx2t2xHmtsBe15spMY29D8GPuC7YB98TZNPPctOueYjsJ6tA9hO7YT3fx5WK95i4wy59HeOGyG\n1WV81D2CbSOxr9DeYwQ7mn9ZTkwjwCGJ6eNpDXG9CDvIjN0AvDlnOb/HtuvY4dhnEPc+oHN9gn3e\nyZ768cCVBcumPcHYbTSv/sdb9yZYbyfeb30GG3bGrf8XqfX8hdZ2eyXt2/BGWK9zezed/vyKfN77\nJZ6/Afj3xPRnaPUMJ7otJC2hfd+aZQXtvdIz3TrB3tPjtDeCH8bi/xjWqL8MO3hN7hsPptVzzFQk\nx5RnO+yDjMVj+5/GhneWuJWfVGBZnXba6efjxFneUdREzKP9S3knVplzEvPuTvz/CLYRpW3lYkov\na9sSMV2IDbsMYkdNyZzJjtiOZE3icVQi3hdhX757sW73exg7JFPk/azBNq55Gc/FNsV2YN9x08ux\no6uj8go407DGa/U4r0tvX0mj2HY2y8X4euwLFLsn8f8j2E4nua0nt6cdsaPlZJ3ujx0hbuOmH028\nPvkZZ/kf7Iu3BXbQcyrWKGzv3s+DGWXS2+HD2JF/cvtJx/yhVMzb0fnzSn+H4gOUpdj7G8R2Hjsz\ntucbW0X7gc3l2Gfwb7R2Tp3qM5bcBh+lNexTpGzSGiwvkpRV/8n8V966HwO+jZ2x24flOeMDkG2w\nUZuk9Ofx+UTMcS457/Mr8nknt+FHU9OP0freltkWYunPM09enYH1ip5IPf8kNuz7FHYgPUT7gdEM\nbP+Uq2zDtA/2hn6e8dxDWFd6Z+yI6oPAge65vCT+eMn9HVL/P4kdDT1M+/joFNqP2Mdb7krae2s7\nYJV5T+ar893vYkovK70xF3ERdiS3GNsAk+7EuvyzEo8Z2MkDYF/KH2Ab5kzgq5T7jB/BesR5R84A\nb8CGFb6G9Vruwna+88dZ9uuwOv5Vh9dsjyXVf9bhNXknORQ5UST5mjuxHVC6Ts/C3tMs2rexHQuu\nYz027HwTNrx6J5ZjyuoxpbfDzbADir91iPm0VMybY7m1POnvUHLZ52MnDRyDHWgkdzRJ12IHAem6\nT053qs/xTLTsTeSfDZis/z1yXpN2PjYcGOeFl7r5K2n1fsDeb3L6TqwnmYx7M9qHBpOfX5HPu5Oq\n20LsGmxIvVOOaSKxgNU3jN1Gkq/bDTuQzVV0pxWvpB9LtF2MbUC3ZATxGmwMtw/r4q/Hjr7BdvhZ\nQ3zjrfto7M1Mx7qt38He6B3Y0fCrsCPxj9J+6ujd2AaQtxO7GDvaG8A+zE9hQ2gjOa+P40lbjx1t\nneaWs6Nb7kWd31qmFVj39yMZzy3Ghi6Oxt7vNOwgIe4mb44dMT2Bnep9FOPvRPPq5kRsyOsEWr2u\n52F1BtYAnYd96Z/nHvu7v8kdQbz82diX/kvYMMaajNdMx/JaP8R2Cnln5nU6826ilzFcBLwWO6Kb\ngm1Pg9jR65+xYZRPYHX9Emz7zjMf2xZnYN+tw7BGaSm2LV6JneAy0y0vHm65GHgnVndPw7bD67Gd\nTpavY72BfbH3uxl2AkinhPMJtHJM76d9x3URNu7/dvLzZ2DDwrOw7/4z3LpnYMO48XbWqT5jeZ9R\nkbJJP6I9D9qp/sdbN9jB2Cg2VJash8VuOW/ARlTeT3sv7qvYEFZ8MtgWtJ+UlDbRzzsdd19ierxt\nYRH5eaQLsZ7cZbTyY1u693JYh1g6uQ57H6dgdbU/9hlelXjNAbSGUEtbQes6pgewpN7xtFdUMse0\nwJV5CHvTyR3s4diXfQ3WkxrAdurJBjI9bwjb4cdn5f2QsadUr8QavQ9h1zDEOabZ2FH3amwHEy8v\nHg/uw856uxMb/rqA1hFtVmydrl2YiX3Q97rlfZRWHQ3SeaPr9Py7scRnbBfsrMd7aV0b9Fz33Juw\n4bS1wBVYUjV5MsdE3g9Yo/cj7HNfhX15jsZ2FE/SOtEiaTGtI9zkdUyrsCPut6VeP0TrOqa12IkJ\np9A5eZs+Ky8pfcLLAO3vOytnui+WGF+F1esVtI6Id8KGI9Zhw9PJOk17AzaKEA/Z3Uh7fmQWtqO4\n273mu4nn3oMNga/ChsiSQyzrGZureyXW61yDbf+Xkt8wjWBnnP0R22Y+zdiD0muw78545mG5xZW0\nrntJX8eUVZ/buefS29x82vOmnT6LtC2xfcwmbnq8+h9v3WDf23TeG6y+b6d1Vl56WUdjvYX4jNhz\nE89lfX4T+bx/lnofp2KnZCdjS28L8VDftXS+ELgfy1fdSevz/Ayts2bT35fk92uQ7P3W7lhO7iHs\nrNjXJZ6bh31mTTpt/u+GsQ9xGa0hndlY0vIObAcw00tkIr3pPPIb+yY7jXrv/HAMYxurUG2MjWql\nz070qbY7P3TDCtp7PWBH2fEZNCeRf6qwiNRrADva3tFzHL5NpzUqID1oBWPPFruN1tllc920iHTX\nqdgQzinjvXAD90ps6On7VDtbWQL2J2wY7wbsCmgYmwxfky4kIiIbPl8JqP2xU3G3xvJK6d7RKDln\nk22zzTajK1fqBt0iIhP0R+yM6cbz1WW9y/29D+s674udVRefgjkPOyNnjJUrVzI6Otrox/z5873H\noDgVp+JUnMkHE79UxxsfDdN0Wldrb4Zds3AzdspkfHHmfOxCURER6TE+hvLm0Lqr8lTs3mNLsHzT\nt7Fz7ocZ+7s9wRgYGPAdQiGKs16Ks16Ks3f5aJhWYFeKp63GbgUSvMHBQd8hFKI466U466U4e5dO\nixQRkUZRwyQiIo0y0RteNsGoO8NEREQK6uvrg0D2+eoxiYhIo6hh6oIoinyHUIjirJfirJfi7F1q\nmEREpFGCGG9MUY5JRGSClGMSEREpqZG/Ijie4457X6lyW289m9NOWxgfOXRNFEVBXHSnOOulOOul\nOHtXkA3TuefuUrLk+znttIV1hiIiIjULYrwxZTTnFzEK6GNkZKTrPSYRkaZRjklERKQkNUxdEMp1\nDYqzXoqzXoqzd6lhEhGRRglivDFFOSYRkQlSjklERKQkNUxdEMqYs+Ksl+Ksl+LsXWqYRESkUYIY\nb0zxlmPq75/NunVrSq4bZsyYxdq1q0uXFxEpK6QcUxBBpnhrmKxclRvI9qEb0IqIDyE1TBrK64rI\ndwCFhDI2rjjrpTjrFUqcIVHDJCIijRJEty5FQ3kiIhOkoTwREZGS1DB1ReQ7gEJCGRtXnPVSnPUK\nJc6QqGESEZFGCWK8MUU5JhGRCVKOSUREpCQ1TF0R+Q6gkFDGxhVnvRRnvUKJMyRqmEREpFGCGG9M\nUY5JRGSClGOSRurvn01fX1/pR3//bN9vQUR6gBqmroh8B5DJ7ow+mngMpaY7P6rcWb2KUMbwFWe9\nFGfvUsMkIiKNEsR4Y4pyTGXXHnj8IlKeckwiIiIl+WyYpgDLgCvc9GzgauAOYAkw01NcNYh8B1BQ\n5DuAQkIZw1ec9VKcvctnw/QB4FZaY0snYw3TLsC1blpERHqMr/HG7YBFwGnAB4HXArcBBwD3AHOx\nw/ldM8oqx1R27YHHLyLlKcc0vrOBfwdGEvPmYI0S7u+cyQ5KRET8m+phna8B7sXyS4M5r4kvnslx\nLDDg/p8J7JVYVOT+Zk9HUURfXx+Dg61poPD0eMu36eXAgtz1T2R9dU+3xxP/P9778Rv/8uXLWbBg\ngbf1F51O5hqaEE/etOqzN+oziiIWLVoEwMDAANLZp4C/ACuAu4CHgQuxoby57jXz3HSWURgt+WB0\nZGRktKzi6x7KXb9PY+PPizO//nwYGhryst6JUpz1Upz1oto4/qTyPd54AHAClmM6C1gFnImd+DCT\n7BMglGMqu/bA4xeR8pRjmph4T3cGcDB2uvjL3bSIiPQY3w3TdcDh7v/VwEHY6eKHAA/4Cqq6yHcA\nBUW+AygkmWtoMsVZL8XZu3w3TCIiIm2CGG9MUY6p7NoDj19EylOOSUREpCQ1TF0R+Q6goMh3AIWE\nMoavOOulOHuXGiYREWmUIMYbU5RjKrv2wOMXkfKUYxIRESlJDVNXRL4DKCjyHUAhoYzhK856Kc7e\npYZJREQaJYjxxhTlmMquPfD4RaQ85ZhERERKUsPUFZHvAAqKfAdQSChj+IqzXoqzd6lhEhGRRgli\nvDFFOaayaw88fhEpTzkmERGRktQwdUXkO4CCIt8BFBLKGL7irJfi7F1qmEREpFGCGG9MUY6p7NoD\nj19EylOOSUREpCQ1TF0R+Q6goMh3AIWEMoavOOulOHuXGiYREWmUIMYbU5RjKrv2wOMXkfKUYxIR\nESlJDVNXRL4DKCjyHUAhoYzhK856Kc7epYZJREQaJYjxxhTlmMquPfD4RaQ85ZhERERKUsPUFZHv\nAAqKfAdQSChj+IqzXoqzd6lhEhGRRglivDFFOaayaw88fhEpTzkmERGRktQwdUXkO4CCIt8BFBLK\nGL7irJfi7F1qmEREpFGCGG9MUY6p7NoDj19EylOOSUREpKQqDdOFBef1oMh3AAVFvgMoJJQxfMVZ\nL8XZu6o0THukpqcCexcotwmwFFgO3Aqc7ubPBq4G7gCWADMrxCZdMZW+vr7Sj/7+2b7fgIgEoMx4\n44eBU4BNgUcT858EzgFOLrCM6cAjWGP2c+AE4HDgfuAs4CRgVs6ylGMqu/Ya4g/5/Yv0spByTFWC\nPINijVAn04HrgGOBy4ADgHuAudg4064ZZdQwlV27GiaRnhVSw1RlKO9kYFvgH4CXJR5F17sca4SG\ngFuAOW4a93dOhdg8i3wHUFDkO4BCQhnDV5z1Upy9a2qFsmcCR2B5ovWJ+T8tUHYE2AvYArgKODD1\n/CjVDs1FRCRQVRqmNwDPBh6vsIwHgcXYSRPxEN7dwDzg3vxixwID7v+ZWBs36KYj9zd7Oooi+vr6\nGBxsTQOFp8db/theyNj1T2R9yen+/tmsW7eGapLxDI6Jr9vTZd//36OvUH/dnh4cHGxUPJ2mY02J\nR/VZ/3QURSxatAiAgYEBQlJlvPFK4K3AugmW2wp4CngAO4HiKuATwCuBVVhP7GSsxdHJDzWv33d5\n5ZhE/OiVHNOjWJ7oHOCL7vGFAuXmAT9xZZcCVwDXYidTHIydLv5yNx2oyHcABUW+AygkfVTaVIqz\nXoqzd1UZyrvcPZKKHA7fDLwgY/5q4KAK8YiIyAYgiG5diobyKqzfd3kN5Yn4EdJQXpUe04qMeaPA\nMyosU0REelyVHNM+icdLgc8D36ojqPBFvgMoKPIdQCGhjOErznopzt5VpWG6P/H4K/A54NV1BCUi\nIr2rynjj3rQSDhsBLwSOB55XNahxKMdUYf2+yyvHJOJHr+SYPktrL/UUMIxd1yS5ppZuFEVEekWV\nobxB7FZCB2LXHx0H3F5DTBuAKGf+U7TutlTmMVlxNksoY/iKs16Ks3dVaZhmAmcDv3aPz2L3vhMR\nESmtyrjS97CLZc93yzkGeC7wxhri6iToHFOvl1eOScSPkHJMVYK8kbEnOmTNq5sapoDLq2ES8SOk\nhqnqvfJemph+CfartBJI7iaUOEMZw1ec9VKcvavKWXnvBS6glVdaA8yvHJGIiPS0Orp1ccP0YA3L\nKkJDeQGX11CeiB8b+lDeh4B/TEw/6B7vBhbUEZSIiPSuMg3T27EhvLQLscZJAsndhBJnKGP4irNe\nirN3lWmYpgJPZMx/gkC6iSIi0lxlGpKbsTs93J2aPwe4BtizalDjUI4p4PLKMYn4saHnmD4NLMZu\nSTTDPQ508z5bW2QiItKTyjRMFwAfBT6J3bh1GPgE8B/AopriClzkO4CCIt8BFBLKGL7irJfi7F1l\nr2O60j1ERERqFcR4Y4pyTAGXV45JxI8NPcckIiLSNWqYuiLyHUBBke8ACgllDF9x1ktx9q4qDdNc\n4Dzgx256d3SBrYiIVFRlvPHHwDeBj2C/wzQNWAbsUUNcnSjHFHB55ZhE/OiVHNNWwKXAejf9JPbb\n4SIiIqVVaZgeArZMTL+YybvDeMNFvgMoKPIdQCGhjOErznopzt5V5feYPgRcATwD+AWwNfDmOoIS\nydLfP5t169aULj9jxizWrl1dY0Qi0g1VxxunAc92/9+ODed1m3JMAZevkmOqo/6V45JeFVKOqWqQ\n+wMDWM8r/sZn/SRGndQwBVxeDZOIHyE1TFVyTBdhN3TdH3ghsI97SCC5m1DiDGUMX3HWS3H2rio5\npr2xa5d0CCoiIrWp0q37DvABYGVNsRSlobyAy2soT8SPkIbyqvSYtgZuBX4FPO7mjQKHVw1KRER6\nV5Uc00Lg9cBp2A8Exg8JJHcTSpyhjOErznopzt5VpccUYWfkPRP7SfXpFZcnIiJSabzxn4DjgNnA\nzsAuwFeAV4xTbnvslPKnY0N/5wBfcMu5FNgR+1XctwIPZJSvlGOaMWNWpYs0fedoQi+vHJOIHyHl\nmKoEeSOwL3A98Hw372Zgz3HKzXWP5cDmwK+xIcF3AvcDZwEnAbOAkzPKV2qY/r6I0uXDbhh8l1fD\nJOJHSA1TlRzT47ROeoD2i2w7uRtrlMDut/c7YFvspInz3fzzscYqUJHvAAqKfAdQSChj+IqzXoqz\nd1VpmK7DfvJiOnAwdvr4FRNcxgDW21oKzAHucfPvcdMiItJjqnTrpmA/DHiIm74KOJfiYy2bY43b\nqcAPgDXY8F1sNZZ3StNQXsDlNZQn4kdIQ3lVzqJbj524cE6JstOAy4ALsUYJrJc0Fxvqmwfcm1/8\nWKyzBTAT2AsYdNOR+5s3Hc8r+npN1zkdD3sMDpab9r1+TWs6lOkoili0aBEAAwMDbOheD/xrYvpX\nwAr3eEuB8n3YWXlnp+bHJz2AnfRwRk75URgt+WC0evkirxuqWL7q+ouWz4uze+svY2hoaHR01N/6\nJxpn0ynOeoUSJ9WGGyZVmRzTicDliemNsZu4HgAcX6D8/sDRwIHYT7EvAw7FGqKDgTuAl5PfMImI\nyAaszHjjDVhDFPsSrR7UUuBFVYMax2j5hl85Jt/lR5VjEvEipBxTmR7TrNR0clhv6wqxiIiIlGqY\nlmJ3fUh7r3tOArk+KJQ4Q7lORHHWS3H2rjJn5f0bdibdUcBv3LwXAJsQ9EWxIiLSBGXHG/uwExSe\ngw363wL8pK6gxqEcU8DllWMS8SOkHFMQQaaoYQq4vBomET9Capiq3JJIckW+Aygo8h1AIaGM4SvO\neinO3qWGSUREGiWIbl2KhvICLq+hPBE/NJQnIiJSkhqmroh8B1BQ5DuAQkIZw1ec9VKcvUsNk4iI\nNEoQ440pyjEFXF45JhE/lGMSEREpSQ1TV0S+Aygo8h1AIaGM4SvOeinO3lXlF2xFJmhqPJwgIpIr\nxL2EckwqX7q8ckzSq5RjEhERKUkNU1dEvgMoKPIdQEGR7wAKCSXXoDjrFUqcIVHDJCIijRLEeGOK\nckwqX7q8ckzSq5RjEhERKUkNU1dEvgMoKPIdQEGR7wAKCSXXoDjrFUqcIVHDJCIijRLEeGOKckwq\nX9I04KnSpWfMmMXatasrrF/En5ByTEEEmaKGSeW9ldfJExKqkBomDeV1ReQ7gIIi3wEUFPkOoJBQ\ncg2Ks16hxBkSNUwiItIoQXTrUjSUp/LeymsoT0KloTwREZGS1DB1ReQ7gIIi3wEUFPkOoJBQcg2K\ns16hxBkSNUwiItIoQYw3pijHpPLeyivHJKFSjklERKQkNUxdEfkOoKDIdwAFRb4DKCSUXIPirFco\ncYZEDZOIiDRKEOONKcoxqby38soxSaiUYxIRESnJV8P0DeAe4ObEvNnA1cAdwBJgpoe4ahL5DqCg\nyHcABUW+AygklFyD4qxXKHGGxFfD9E3g0NS8k7GGaRfgWjctIiI9xud44wBwBbCnm74NOADrSc3F\nDpN3zSinHJPKeyuvHJOESjmmcuZgjRLu7xyPsYiIiCdTfQeQY5SOh7bHYh0usFTUXsCgm47c37zp\neF7R15eZXg4s6OLy65qO/29KPHnTzanPOJ8wODh2OplryHq+KdPLly9nwYIFjYknb1r1Wb3+Fi1a\nBMDAwABSzADtJz/chg3hAcxz01lGYbTkI27wqpQv8rqhiuWrrr9o+bw4J2v9RR9DntffKt/J0NBQ\nx+ebQnHWK5Q43f4vCE3KMZ0FrALOxE58mEn2CRCj5etXOSaVr1bevt8i4Qkpx+QryIuxEx22wvJJ\nHwN+CHwb2AEYBt4KPJBRVg2Tynsrr4ZJQhVSw+Tr5IcjgW2AjYHtsdPHVwMHYaeLH0J2oxSIyHcA\nBUW+Aygo8h1AIaFcz6I46xVKnCFp0ll5IiIiYXTrUjSUp/LeymsoT0KloTwREZGS1DB1ReQ7gIIi\n3wEUFPkOoJBQcg2Ks16hxBmSpl5gK9JAU+PhkFJmzJjF2rWra4xHZMMUxHhjinJMKh9seeWoxBfl\nmEREREpSw9QVke8ACop8B1BQ5DuAgiLfARQSSk5EcfYuNUwiItIoQYw3pijHpPLBlleOSXxRjklE\nRKQkNUxdEfkOoKDIdwAFRb4DKCjyHUAhoeREFGfvUsMkIiKNEsR4Y4pyTCofbHnlmMQX5ZhERERK\nUsPUFZHvAAqKfAdQUOQ7gIIi3wEUEkpORHH2LjVMIiLSKEGMN6Yox6TywZZXjkl8UY5JRESkJDVM\nXRH5DqCgyHcABUW+Aygo8h1AIaHkRBRn71LDJCIijRLEeGOKckwqH2x55ZjEF+WYRCSD/QJu2Ud/\n/+xKa+/C+cH4AAAJ1klEQVTvn11h/Rt7jV16ixqmroh8B1BQ5DuAgiLfARQUjfP8U1iPq9xj3bo1\nlaKz8qPAUIn1PznpsYeSuwklzpCoYRIRkUYJYrwxRTkmle/Z8lVyVJZj8LftK7/ml3JMItIF1XJU\nIauWH/Odn1OObaLUMHVF5DuAgiLfARQU+Q6goKjLy6+Wo5q8OOuRzN208mO+83NZj/FzdlXX32vU\nMImISKOE2L9XjknlVX7Sy4ecHwt//XVQjklERKQkNUxdEfkOoKDIdwAFRb4DKCjyHUBBke8ACgnn\n+qCowGv8Xlwdmqm+AxAR2fDFJ66Us25dECNwtQnx3SrHpPIqP+nlezvHU8f6feeolGMSEREpSQ1T\nV0S+Aygo8h1AQZHvAAqKfAdQUORhnb4vDu7m+qMa4pOkJjZMhwK3Ab8HTvIcS0nLfQdQkOKsl+LM\nV+bi4LMT//tYf9bFyVlC+dzD0bSGaQrwJaxx2h04EtjNa0SlPOA7gIIUZ70UZ70UZ69qWsO0L/AH\nYBi7z/4lwOt8BiQiIpOraaeLbwv8JTH9V+BF6Rf197+21MLXrZusq6+HJ2EddRj2HUBBw74DKGjY\ndwAFDfsOoKBh3wEUNOw7gA1O004dfBM2jHecmz4aa5jel3jNH4CdJzkuEZHQ/RF4pu8gimhaj+lv\nwPaJ6e2xXlNSEBUrIiIbhqlYqz4AbIyd7hLgyQ8iIrIhOQy4HRuyO8VzLCIiIiIiImEI5cLbYeAm\nYBnwK7+htPkGcA9wc2LebOBq4A5gCTDTQ1xpWXEuxHKNy9zj0MkPa4ztsZ8uvQX4LfB+N79JdZoX\n40KaVZ+bAEuxoftbgdPd/CbVJeTHuZBm1WdsChbPFW66afUZvCnY0N4AMI1m555WYBtA07wUeD7t\nO/yzgBPd/ycBZ0x2UBmy4vw48EE/4eSaC+zl/t8cG37ejWbVaV6MTazP6e7vVOB64CU0qy5jWXE2\nsT7BYvoWcLmbbmJ9ZmraBbZ5Qrvwtmmn4QP8DFiTmnc4cL77/3zg9ZMaUbasOKF5dXo3rXvRPAT8\nDrsOr0l1mhcjNK8+H3F/N8YORNfQrLqMZcUJzavP7YBXAefSiq2J9ZkplIYp68LbbXNe69socA1w\nA63rsZpqDjZshvs7x2Ms43kfcCNwHs0bghjAenlLaW6dDmAxXu+mm1afG2GN6D20hh+bWJdZcULz\n6vNs4N+BkcS8JtZnplAapsm4XUNd9sd2AIcB/4INTYWgrrtldsNXgJ2wYam7gM/6DafN5sBlwAeA\ndannmlKnmwPfxWJ8iGbW5wgWz3bAy4ADU883pS7TcQ7SvPp8DXAvll/K68k1pT4zhdIwFbnwtinu\ncn/vA76PDUM21T1YHgJgHrYxN9G9tL5I59KcOp2GNUoXAj9w85pWp3GMF9GKsan1CfAgsBjYm+bV\nZVIc5wtpXn3+AzZstwK4GHg5to02uT7bhNIw3QA8i9aFt0fQSug1yXRghvt/M+AQ2pP4TXM5MN/9\nP5/Wjqtp5iX+fwPNqNM+bNjmVuBziflNqtO8GJtWn1vRGv7aFDgYO9pvUl1CfpxzE69pQn1+GDt4\n3wl4G/AT4BiaV58bhBAuvN0JG39ejp2e26Q4LwZWAk9g+bp3YmcPXkOzTh9Nx/ku4ALsFPwbsS9T\nE8bGX4IN6yyn/TThJtVpVoyH0bz63BP4DRbnTVhuBJpVl5AfZ9PqM+kAWgfxTatPERERERERERER\nERERERERERERERERERGRuoxgV5zHpmJ3xLgi++W1OgG7aWn88yPHJJ7bCrsZ8HtSZYax61Fuwu6B\ndirwNPfcAPAodh3Lrdh98eaTbRC7M8Ay99qPufnHAl8s82Zy5C1vDvC/2PU2t2B3KBAJxlTfAcgG\n7WHgOdjv2DyGXSn/V7p/j673Aq8A9sHuDTcDuyI/9hbgx8CRwNcS80exRmU1dueOc9zzx7rn/wC8\nwP2/E/A97O4KizJi+CnwWuxuIMuxxrjo+96I9ptvTtQngatoNVp7VFhWbAqwvobliIwrlFsSSbh+\nBLza/X8kdmeH+MaSm2E/DLgU64kc7uYPYDv2X7vHfm7+IBAB38F6QxflrPMU4HisUQK7ueoFieff\nBnwUeDr5d6l/GGvgXk/2FfIrsN+7eX/Gc0mPuPfwTDe9DXAldvX9mYnXPQR8BmvE9gOOxuplGfBV\nWt/Vd2J3QFmK3RMty1zs/pKx3yb+PwnrES6n9UN3e2F3Hb8Ra2zj9xthd6n+f+597u3m3YA17Mlb\n8YiIBGEddhuX72BDYsuwW6TEQ3mfAt7u/p+J7XCnY/chi4fQnoXtGMEapgewnXsf8Avsbu5J/ViP\nJ8/2WKMG1rNI/sBb1o88LsN6XgOMvQfaTFq/z5M0SOs9bumWuxvW8/oj1oN7GjZ0GDeMI8Cb3f+7\nYbeRmeKmv4wNRc4D/uyWOQ34OfCFjPUfgv1O0E+w+6bF98Y7DPg/rAcbxw/WUMV3wf8E1hiB/azD\nl9z/U7H63tJNH4Hdh0+kdhrKk267GdupH8nYXMch2HDXCW76aVjDcTe2Q3weNnz0rESZX2H30gM7\n6h/AdrZFHYH9BARYg/kN4L86vL7TD8B1eu6lWC9wBOuZ/A54EXAtrZ/HuBXYEevdrMfuAg42DLk3\n1jMBa0juxu5aHQGr3PxLgV0y1r0EeAZ2777DsMZ1D+Ag7P0+5l73ALCFe/zMzTsfq5fYpe7vrtiw\n7DVuegqtz0GkVmqYZDJcjg1THQBsnXrujcDvU/MWYj8fcgy2A3ws8dzjif/XM3YbXosNi+2E9VTS\njsRODjjaTc8DdsZ6MmkzsIbvDmBWxvPPxxqXLD/DGt2k0Q7xP0Z7Dup8rLeTlP7V5k4N4xps2PRi\nrPf2Mrf88X5pNf38w4n5t5A/fChSG+WYZDJ8A2tsbknNv4r2HM3z3d9+rIcA8A5aQ1pFnQ78N62f\nINkca+R2wfJa22EN107AGcBRibJ9iTJfxn5T68GMdQwAn2ZiZ9kV/fnta7FhvbgRnw3sgOWVDnDT\n07CTOLIciA2JgtXBztgQ4NVYjmpT99ws7L2twe5EDlZPUUbMt7t4XuympwG7F3w/IhOiHpN0U9wD\n+ButXEXylzNPxX4n6CbsIOlP2AkQX8aGtd6BJdnjkxiSy8ybBvtF0c2x3NST7vFZ7KSH76Veexlw\niYsFLK/S5+L5XmI+2A7+N9jQ2jrg87SfVJGMKSuuTr8ampz/O+zkjCUujieBf8aGMRcCv8SG4Zbl\nLG9vrL6fcuW/jp2AAXaiww3Yz4osduuZj51gMR3rOb4zI64nsMbyC9jQ31QsF5XXYxQRERERERER\nEREREREREREREREREREREREREWms/w+C3I6X2RE1/gAAAABJRU5ErkJggg==\n",
       "text": [
        "<matplotlib.figure.Figure at 0x115e82f50>"
       ]
      }
     ],
     "prompt_number": 32
    },
    {
     "cell_type": "heading",
     "level": 4,
     "metadata": {},
     "source": [
      "All MyVariant.info Annotations for SNPs flattened into a dataframe"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "fiSSEA_results.mv_annot.head(1).T"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th>_id</th>\n",
        "      <th>chr6:g.39041502A&gt;C</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>_id</th>\n",
        "      <td>    chr6:g.39041502A&gt;C</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>cosmic.allele1</th>\n",
        "      <td>                   NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>cosmic.allele2</th>\n",
        "      <td>                   NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>cosmic.chrom</th>\n",
        "      <td>                   NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>cosmic.chromEnd</th>\n",
        "      <td>                   NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>cosmic.chromStart</th>\n",
        "      <td>                   NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>cosmic.mut_freq</th>\n",
        "      <td>                   NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>cosmic.mut_nt</th>\n",
        "      <td>                   NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>cosmic.tumor_site</th>\n",
        "      <td>                   NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.1000gp1.ac</th>\n",
        "      <td>                  1246</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.1000gp1.af</th>\n",
        "      <td>             0.5705128</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.1000gp1.afr_ac</th>\n",
        "      <td>                   310</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.1000gp1.afr_af</th>\n",
        "      <td>             0.6300813</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.1000gp1.amr_ac</th>\n",
        "      <td>                   218</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.1000gp1.amr_af</th>\n",
        "      <td>             0.6022099</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.1000gp1.asn_ac</th>\n",
        "      <td>                   299</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.1000gp1.asn_af</th>\n",
        "      <td>             0.5227273</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.1000gp1.eur_ac</th>\n",
        "      <td>                   419</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.1000gp1.eur_af</th>\n",
        "      <td>             0.5527704</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.aa</th>\n",
        "      <td>                   NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.aa.aapos_fathmm</th>\n",
        "      <td> ENSP00000362353:L260F</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.aa.aapos_sift</th>\n",
        "      <td> ENSP00000362353:L260F</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.aa.alt</th>\n",
        "      <td>                     F</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.aa.codonpos</th>\n",
        "      <td>                     3</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.aa.pos</th>\n",
        "      <td>                   260</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.aa.ref</th>\n",
        "      <td>                     L</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.aa.refcodon</th>\n",
        "      <td>                   TTA</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.allele1</th>\n",
        "      <td>                     A</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.allele2</th>\n",
        "      <td>                     C</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.ancestral_allele</th>\n",
        "      <td>                     C</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>...</th>\n",
        "      <td>...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>evs.hgvs.coding</th>\n",
        "      <td>              c.780A&gt;C</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>evs.hgvs.protein</th>\n",
        "      <td>             p.(L260F)</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>evs.ma_fin_percent.african_american</th>\n",
        "      <td>               41.9655</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>evs.ma_fin_percent.all</th>\n",
        "      <td>               43.3877</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>evs.ma_fin_percent.european_american</th>\n",
        "      <td>               44.1163</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>evs.on_illumina_human_exome_chip</th>\n",
        "      <td>                   yes</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>evs.polyphen2</th>\n",
        "      <td>                   NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>evs.polyphen2.class</th>\n",
        "      <td>                     b</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>evs.polyphen2.score</th>\n",
        "      <td>                     e</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>evs.ref_base_ncbi</th>\n",
        "      <td>                     A</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>evs.rs_id</th>\n",
        "      <td>             rs1042044</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>gwassnps.allele1</th>\n",
        "      <td>                   NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>gwassnps.allele2</th>\n",
        "      <td>                   NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>gwassnps.chrom</th>\n",
        "      <td>                   NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>gwassnps.chromEnd</th>\n",
        "      <td>                   NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>gwassnps.chromStart</th>\n",
        "      <td>                   NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>gwassnps.pubmedID</th>\n",
        "      <td>                   NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>gwassnps.rsid</th>\n",
        "      <td>                   NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>gwassnps.trait</th>\n",
        "      <td>                   NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>mutdb.allele1</th>\n",
        "      <td>                     A</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>mutdb.allele2</th>\n",
        "      <td>                     C</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>mutdb.chrom</th>\n",
        "      <td>                     6</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>mutdb.chromEnd</th>\n",
        "      <td>           3.90415e+07</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>mutdb.chromStart</th>\n",
        "      <td>           3.90415e+07</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>mutdb.cosmic_id</th>\n",
        "      <td>                  None</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>mutdb.dbsnp_id</th>\n",
        "      <td>             rs1042044</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>mutdb.mutpred_score</th>\n",
        "      <td>                 0.552</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>mutdb.strand</th>\n",
        "      <td>                     p</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>mutdb.uniprot_id</th>\n",
        "      <td>            VAR_015098</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>snpedia.text</th>\n",
        "      <td>                   NaN</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "<p>242 rows \u00d7 1 columns</p>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 21,
       "text": [
        "_id                                      chr6:g.39041502A>C\n",
        "_id                                      chr6:g.39041502A>C\n",
        "cosmic.allele1                                          NaN\n",
        "cosmic.allele2                                          NaN\n",
        "cosmic.chrom                                            NaN\n",
        "cosmic.chromEnd                                         NaN\n",
        "cosmic.chromStart                                       NaN\n",
        "cosmic.mut_freq                                         NaN\n",
        "cosmic.mut_nt                                           NaN\n",
        "cosmic.tumor_site                                       NaN\n",
        "dbnsfp.1000gp1.ac                                      1246\n",
        "dbnsfp.1000gp1.af                                 0.5705128\n",
        "dbnsfp.1000gp1.afr_ac                                   310\n",
        "dbnsfp.1000gp1.afr_af                             0.6300813\n",
        "dbnsfp.1000gp1.amr_ac                                   218\n",
        "dbnsfp.1000gp1.amr_af                             0.6022099\n",
        "dbnsfp.1000gp1.asn_ac                                   299\n",
        "dbnsfp.1000gp1.asn_af                             0.5227273\n",
        "dbnsfp.1000gp1.eur_ac                                   419\n",
        "dbnsfp.1000gp1.eur_af                             0.5527704\n",
        "dbnsfp.aa                                               NaN\n",
        "dbnsfp.aa.aapos_fathmm                ENSP00000362353:L260F\n",
        "dbnsfp.aa.aapos_sift                  ENSP00000362353:L260F\n",
        "dbnsfp.aa.alt                                             F\n",
        "dbnsfp.aa.codonpos                                        3\n",
        "dbnsfp.aa.pos                                           260\n",
        "dbnsfp.aa.ref                                             L\n",
        "dbnsfp.aa.refcodon                                      TTA\n",
        "dbnsfp.allele1                                            A\n",
        "dbnsfp.allele2                                            C\n",
        "dbnsfp.ancestral_allele                                   C\n",
        "...                                                     ...\n",
        "evs.hgvs.coding                                    c.780A>C\n",
        "evs.hgvs.protein                                  p.(L260F)\n",
        "evs.ma_fin_percent.african_american                 41.9655\n",
        "evs.ma_fin_percent.all                              43.3877\n",
        "evs.ma_fin_percent.european_american                44.1163\n",
        "evs.on_illumina_human_exome_chip                        yes\n",
        "evs.polyphen2                                           NaN\n",
        "evs.polyphen2.class                                       b\n",
        "evs.polyphen2.score                                       e\n",
        "evs.ref_base_ncbi                                         A\n",
        "evs.rs_id                                         rs1042044\n",
        "gwassnps.allele1                                        NaN\n",
        "gwassnps.allele2                                        NaN\n",
        "gwassnps.chrom                                          NaN\n",
        "gwassnps.chromEnd                                       NaN\n",
        "gwassnps.chromStart                                     NaN\n",
        "gwassnps.pubmedID                                       NaN\n",
        "gwassnps.rsid                                           NaN\n",
        "gwassnps.trait                                          NaN\n",
        "mutdb.allele1                                             A\n",
        "mutdb.allele2                                             C\n",
        "mutdb.chrom                                               6\n",
        "mutdb.chromEnd                                  3.90415e+07\n",
        "mutdb.chromStart                                3.90415e+07\n",
        "mutdb.cosmic_id                                        None\n",
        "mutdb.dbsnp_id                                    rs1042044\n",
        "mutdb.mutpred_score                                   0.552\n",
        "mutdb.strand                                              p\n",
        "mutdb.uniprot_id                                 VAR_015098\n",
        "snpedia.text                                            NaN\n",
        "\n",
        "[242 rows x 1 columns]"
       ]
      }
     ],
     "prompt_number": 21
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "Query and Parsing - Annotations from MyVariant.info "
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "<center><h1>fiSSEA Functions</h1><center>"
     ]
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "MyVariant.info Query and Parsing"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "\n",
      "def myvariant_post(hgvs_list):\n",
      "    '''\n",
      "    Query and Parser for myvariant.info\n",
      "    Parses Raw Elastic Search results into\n",
      "    pandas dataframe\n",
      "    \n",
      "    Parameters\n",
      "    -------------\n",
      "    hgvs_list: list, required\n",
      "        \n",
      "        \n",
      "    Output\n",
      "    -------------\n",
      "    pandas df:\n",
      "        normalized json of myvariant results\n",
      "    '''\n",
      "    \n",
      "    def normalize(con):\n",
      "        '''\n",
      "        This function uses the json.loads and json_normalize\n",
      "        modules to flatten the myvariant.info results\n",
      "        '''\n",
      "        temp_df = json_normalize(json.loads(con), 'docs')\n",
      "        source_parsed = []\n",
      "        for source_json in temp_df['_source'].values:\n",
      "            try:\n",
      "                source_parsed.append(json_normalize(source_json))\n",
      "            except TypeError:  #occurs with empty results\n",
      "                pass\n",
      "        return pd.concat(source_parsed)\n",
      "    \n",
      "    \n",
      "    \n",
      "    if type(hgvs_list) == list:\n",
      "        hgvs_list = ','.join(hgvs_list)\n",
      "        \n",
      "    assert type(hgvs_list) == str\n",
      "    \n",
      "    h = httplib2.Http()\n",
      "    headers = {'content-type': 'application/x-www-form-urlencoded'}\n",
      "    params = 'ids=' + hgvs_list +\",size=1000\"  #can pass a dictionary\n",
      "    #print params\n",
      "    res, con = h.request('http://myvariant.info:8001/api/variant/', 'POST', params, headers=headers)\n",
      "    \n",
      "    \n",
      "    mv_df = normalize(con)\n",
      "    mv_df.index = mv_df['_id']\n",
      "\n",
      "    return mv_df\n",
      "    \n",
      "    "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 4
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def get_myvariant_snp_annot(vcf_df_mv):\n",
      "    '''\n",
      "    Returns myvariant.info annotations for all snps for\n",
      "    input vcf pandas dataframe. Submits myvariant.info\n",
      "    POST request in snp batches of 1000.\n",
      "    \n",
      "    Parameters\n",
      "    -------------\n",
      "    vcf_df_mv: pandas df; required\n",
      "    \n",
      "        vcf_df_mv must contain the following columns:\n",
      "            vartype1, vartype2, CHROM, REF, POS, a1, a2\n",
      "            *please see pandasVCF docs for column definitions\n",
      "            \n",
      "\n",
      "    Output\n",
      "    -------------\n",
      "    myvariant.info results: pandas df\n",
      "    \n",
      "    '''\n",
      "    \n",
      "    def chunks(l, n):\n",
      "        \"\"\" \n",
      "        Yield successive n-sized chunks from l.\n",
      "        \"\"\"\n",
      "        l = list(l)\n",
      "        chunk_list = []\n",
      "        for i in xrange(0, len(l), n):\n",
      "            chunk_list.append( l[i:i+n] )\n",
      "        return chunk_list\n",
      "    \n",
      "    \n",
      " \n",
      "    if 'CHROM' not in vcf_df_mv.columns:\n",
      "        vcf_df_mv.reset_index(inplace=True)\n",
      "        \n",
      "    \n",
      "    vcf_df_mv = vcf_df_mv[(vcf_df_mv['vartype1'] == 'snp') | (vcf_df_mv['vartype2'] == 'snp')] #restrict to snps\n",
      "\n",
      "    non_ref_a1 = vcf_df_mv[vcf_df_mv['REF'] != vcf_df_mv['a1']][['CHROM', 'POS', 'REF', 'a1']]  #restrict a1 to non-reference alleles\n",
      "    hgvs_a1 = 'chr' + non_ref_a1['CHROM'].astype(str) + \":g.\" + non_ref_a1['POS'].astype(str)  \\\n",
      "                                 + non_ref_a1['REF'] + '>' +  non_ref_a1['a1']\n",
      "\n",
      "    non_ref_a2 = vcf_df_mv[vcf_df_mv['REF'] != vcf_df_mv['a2']][['CHROM', 'POS', 'REF', 'a2']]  #restrict a2 to non-reference alleles\n",
      "    hgvs_a2 = 'chr' + non_ref_a2['CHROM'].astype(str) + \":g.\" + non_ref_a2['POS'].astype(str) \\\n",
      "                                 + non_ref_a2['REF'] + '>' +  non_ref_a2['a2']\n",
      "\n",
      "\n",
      "    hgvs_ids = set(hgvs_a1.values) | set(hgvs_a2.values)  #set of non-reference a1 and a2 snps\n",
      "\n",
      "    hgvs_ids_chunks = chunks(hgvs_ids, 1000)  #batch queries of hgvs a1 and a2 variants, n=1000\n",
      "\n",
      "    myvariant_results = map(myvariant_post, hgvs_ids_chunks)\n",
      "        \n",
      "    return myvariant_results"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "fiSSEA Function"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "class fiSSEA(object):\n",
      "\n",
      "    \n",
      "    def __init__(self, vcf_file_path, sample_id='', chunksize=100000, fiSSEA_score='dbnsfp.cadd.phred', gene_stat=np.mean):\n",
      "        \n",
      "        \n",
      "        #setting fiSSEA_score\n",
      "        self.fiSSEA_score = fiSSEA_score\n",
      "        \n",
      "        #setting gene statistic\n",
      "        self.gene_stat = gene_stat\n",
      "        \n",
      "        #open Vcf object\n",
      "        vcf_df_obj = Vcf(vcf_file_path, sample_id, ['#CHROM', 'POS', 'REF', 'ALT', 'ID', 'FILTER', 'FORMAT'], chunksize=chunksize)\n",
      "\n",
      "        #read in vcf chunk\n",
      "        vcf_df_obj.get_vcf_df_chunk()\n",
      "\n",
      "        #drop duplicate vcf lines, if necessary\n",
      "        vcf_df_obj.df = vcf_df_obj.df.drop_duplicates(subset=['CHROM', 'POS'])\n",
      "\n",
      "        #remove empty variant lines\n",
      "        vcf_df_obj.df = vcf_df_obj.df[vcf_df_obj.df.ALT!='.']\n",
      "\n",
      "        #replace chr if necessary, depends on format of annotations\n",
      "        vcf_df_obj.df['CHROM'] = vcf_df_obj.df['CHROM'].str.replace('chr', '')\n",
      "\n",
      "        #parse vcf lines\n",
      "        vcf_df_obj.add_variant_annotations( inplace=True, verbose=True)\n",
      "        \n",
      "        #settting vcf_df as class object\n",
      "        self.vcf_df = vcf_df_obj.df\n",
      "\n",
      "        #get myvariant.info annotations, create fiSSEA dataframe\n",
      "        self.fiSSEA_status = self.get_fi_scores()\n",
      "        \n",
      "\n",
      "    \n",
      "    def get_multigene_var_records(self, df_line):\n",
      "        '''\n",
      "        Splits multigene variants into separate\n",
      "        rows\n",
      "        '''\n",
      "        genes = df_line['dbnsfp.genename']\n",
      "        gene_series = [] #list of gene variants\n",
      "        for gene in genes: #iterate through all genes intersecting this variant\n",
      "            df_line_copy = df_line.copy()  #avoids aliasing\n",
      "            del df_line_copy['dbnsfp.genename']  #removing genename to allow gene insertion\n",
      "            df_line_copy['dbnsfp.genename'] = gene\n",
      "            gene_series.append(df_line_copy)\n",
      "\n",
      "        return gene_series\n",
      "    \n",
      "    \n",
      "    \n",
      "    def get_fi_scores(self):\n",
      "        '''\n",
      "        Creates fi_scores series, index on gene symbol, value fi_score.\n",
      "        '''\n",
      "        \n",
      "        #Retrieving myvariant.info annotations for all snps\n",
      "        myvariant_results = get_myvariant_snp_annot(self.vcf_df) #query and parse myvariant.info using all snps in self.vcf_df\n",
      "        self.mv_annot = pd.concat(myvariant_results)  #setting dataframe of parsed myvariant.info results\n",
      "\n",
      "        #reducing size of dataframe for fiSSEA\n",
      "        fiSSEA_cols = ['dbnsfp.genename', self.fiSSEA_score]\n",
      "        mv_annot_fiSSEA = self.mv_annot[fiSSEA_cols] \n",
      "        \n",
      "        #dropping variants without a genename & filling fi_score NaN with 0\n",
      "        mv_annot_fiSSEA.dropna(subset=['dbnsfp.genename'], inplace=True)  \n",
      "        mv_annot_fiSSEA[self.fiSSEA_score].fillna(value=0)  \n",
      "\n",
      "        \n",
      "        #Splitting variants intersecting multiple genes into separate rows\n",
      "        mv_annot_fiSSEA['dbnsfp.genename_type'] = mv_annot_fiSSEA['dbnsfp.genename'].map(type)\n",
      "        mv_annot_fiSSEA_multigene = mv_annot_fiSSEA[mv_annot_fiSSEA['dbnsfp.genename_type']==list]\n",
      "        mv_annot_fiSSEA = mv_annot_fiSSEA[~mv_annot_fiSSEA.index.isin(mv_annot_fiSSEA_multigene.index)]\n",
      "        \n",
      "        if len(mv_annot_fiSSEA_multigene) > 0:\n",
      "            mv_annot_fiSSEA_multigene['index'] = mv_annot_fiSSEA_multigene.index\n",
      "            mv_annot_fiSSEA_multigene.drop_duplicates(subset='index', inplace=True)\n",
      "            mv_annot_fiSSEA_multigene = mv_annot_fiSSEA_multigene.apply(self.get_multigene_var_records, axis=1)\n",
      "            mv_annot_fiSSEA_multigene = [n for i in mv_annot_fiSSEA_multigene for n in i]\n",
      "            mv_annot_fiSSEA = mv_annot_fiSSEA.append(pd.DataFrame(mv_annot_fiSSEA_multigene))\n",
      "            del mv_annot_fiSSEA_multigene\n",
      "        \n",
      "        gc.collect()\n",
      "\n",
      "        \n",
      "        #Setting gene specific functional impact score\n",
      "        self.fi_scores = mv_annot_fiSSEA.groupby('dbnsfp.genename')[self.fiSSEA_score].aggregate(self.gene_stat)\n",
      "        \n",
      "        return 0"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 11
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}