{
 "metadata": {
  "name": "",
  "signature": "sha256:915ee061dc5c7d79e2bf942e0e025769153064ee0ae21e23b22dcb374583add4"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "<center><h1> Functional Impact SNP Set Enrichment (fiSSEA) </h1></center>\n",
      "<center><h4>Authors: Erick R. Scott, Adam Mark, Ali Torkamani, Chunlei Wu & Andrew Su</h4></center>\n",
      "<center><h4>November 2014</h4></center>"
     ]
    },
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "Libraries"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import os, sys, time, gc, argparse\n",
      "import json, httplib2\n",
      "import pandas as pd\n",
      "from pandas.io.json import json_normalize\n",
      "import numpy as np\n",
      "import subprocess"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1
    },
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "PandasVCF, MyVariant.info, fiSSEA Libraries"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "sys.path.append('../pandasVCF/src/single_sample/')\n",
      "from pdVCFsingle import *\n",
      "\n",
      "sys.path.append('../myvariant/src/')\n",
      "import myvariant\n",
      "mv = myvariant.MyVariantInfo()\n",
      "\n",
      "sys.path.append('../src/')\n",
      "from fiSSEA import *"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 2
    },
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "VCF CDS Variants"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def intersect_cds_vcf(vcf_path, cds_bed, write_path):\n",
      "    '''\n",
      "    Convenience function to create coding snp vcf file\n",
      "    '''\n",
      "    os.system('bedtools intersect -a ' + vcf_path + ' -b ' + cds_bed + ' -wa -header |uniq > ' + write_path )\n",
      "    return write_path"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 3
    },
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "Running fiSSEA "
     ]
    },
    {
     "cell_type": "heading",
     "level": 4,
     "metadata": {},
     "source": [
      "<h1>GSEA Command Line Run</h1>\n",
      "\n",
      "<ol> Please download the gsea.jar source from http://www.broadinstitute.org/gsea/downloads.jsp</ol>\n",
      "\n",
      "<ol>We have included the tissue-specific gene set .gmt data derived from the Xavier Lab's Enrichment Profiler dataset (http://xavierlab2.mgh.harvard.edu/EnrichmentProfiler/download.html) in the /fiSSEA/gsea/genesets/ directory</ol>\n",
      "\n"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "example_vcf = '../data/snyderome_chr6_cds.vcf.gz'\n",
      "sample_id = 'TB0001907'\n",
      "chunk = 100000\n",
      "fi_score = 'dbnsfp.cadd.phred'\n",
      "\n",
      "#Set the path to the gsea .jar file\n",
      "gsea_jar = '../gsea/gsea2-2.1.0.jar'\n",
      "\n",
      "#Set the path to the tissue specific gene set\n",
      "tissue_genesets_path = '../gsea/genesets/EnrichmentValues_Primary_U133A_ave_percent_over70percent_no_brain.gmt'\n",
      "\n",
      "#Set the path to write GSEA results\n",
      "output_dir_path = '../gsea/output/'\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 4
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#Generate Functional Impact Scores for all Nonsynon DNA variants in Coding VCF\n",
      "fiSSEA_results = fiSSEA(example_vcf, sample_id=sample_id, chunksize=chunk, fiSSEA_score=fi_score)\n",
      "\n",
      "#Call GSEA jar PreRanked using several default parameters\n",
      "fiSSEA_results.run_GSEA_preranked(gsea_jar, tissue_genesets_path, output_dir_path)\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "input variant rows: 1377\n",
        "dropping 0 variants with genotype call == \".\" \n",
        "current variant rows: 1377\n"
       ]
      },
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 8,
       "text": [
        "0"
       ]
      }
     ],
     "prompt_number": 8
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "Opens GSEA Summary Html (working with Chrome browser)"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "fiSSEA_results.open_GSEA_html()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 9,
       "text": [
        "0"
       ]
      }
     ],
     "prompt_number": 9
    },
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "Functional Impact Annotation Results"
     ]
    },
    {
     "cell_type": "heading",
     "level": 4,
     "metadata": {},
     "source": [
      "fiSSEA_results object has the following attributes"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "dir(fiSSEA_results)[-8:]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 10,
       "text": [
        "['fiSSEA_score',\n",
        " 'fiSSEA_status',\n",
        " 'fi_scores',\n",
        " 'gene_stat',\n",
        " 'get_fi_scores',\n",
        " 'get_multigene_var_records',\n",
        " 'mv_annot',\n",
        " 'vcf_df']"
       ]
      }
     ],
     "prompt_number": 10
    },
    {
     "cell_type": "heading",
     "level": 4,
     "metadata": {},
     "source": [
      "Original vcf parsed with pandasVCF"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "fiSSEA_results.vcf_df.head()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>CHROM</th>\n",
        "      <th>POS</th>\n",
        "      <th>REF</th>\n",
        "      <th>ALT</th>\n",
        "      <th>ID</th>\n",
        "      <th>FILTER</th>\n",
        "      <th>multiallele</th>\n",
        "      <th>phase</th>\n",
        "      <th>GT1</th>\n",
        "      <th>GT2</th>\n",
        "      <th>...</th>\n",
        "      <th>vartype1</th>\n",
        "      <th>vartype2</th>\n",
        "      <th>GT</th>\n",
        "      <th>AD</th>\n",
        "      <th>DP</th>\n",
        "      <th>GQ</th>\n",
        "      <th>GQX</th>\n",
        "      <th>MQ</th>\n",
        "      <th>PL</th>\n",
        "      <th>VF</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>0</th>\n",
        "      <td> 1</td>\n",
        "      <td>  69428</td>\n",
        "      <td> T</td>\n",
        "      <td> G</td>\n",
        "      <td> .</td>\n",
        "      <td> TruthSensitivityTranche99.90to100.00;LowGQX;Lo...</td>\n",
        "      <td> 0</td>\n",
        "      <td> /</td>\n",
        "      <td> 1</td>\n",
        "      <td> 1</td>\n",
        "      <td>...</td>\n",
        "      <td> snp</td>\n",
        "      <td> snp</td>\n",
        "      <td> 1/1</td>\n",
        "      <td> 11,3</td>\n",
        "      <td> 14</td>\n",
        "      <td> 3.01</td>\n",
        "      <td>  3</td>\n",
        "      <td> 11</td>\n",
        "      <td>     39,3,0</td>\n",
        "      <td> 0.214</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>1</th>\n",
        "      <td> 1</td>\n",
        "      <td>  69497</td>\n",
        "      <td> G</td>\n",
        "      <td> A</td>\n",
        "      <td> .</td>\n",
        "      <td>       TruthSensitivityTranche99.90to100.00;LowGQX</td>\n",
        "      <td> 0</td>\n",
        "      <td> /</td>\n",
        "      <td> 1</td>\n",
        "      <td> 1</td>\n",
        "      <td>...</td>\n",
        "      <td> snp</td>\n",
        "      <td> snp</td>\n",
        "      <td> 1/1</td>\n",
        "      <td>  1,1</td>\n",
        "      <td>  2</td>\n",
        "      <td> 3.01</td>\n",
        "      <td>  3</td>\n",
        "      <td> 30</td>\n",
        "      <td>     36,3,0</td>\n",
        "      <td> 0.500</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>2</th>\n",
        "      <td> 1</td>\n",
        "      <td>  69508</td>\n",
        "      <td> G</td>\n",
        "      <td> A</td>\n",
        "      <td> .</td>\n",
        "      <td>       TruthSensitivityTranche99.90to100.00;LowGQX</td>\n",
        "      <td> 0</td>\n",
        "      <td> /</td>\n",
        "      <td> 1</td>\n",
        "      <td> 1</td>\n",
        "      <td>...</td>\n",
        "      <td> snp</td>\n",
        "      <td> snp</td>\n",
        "      <td> 1/1</td>\n",
        "      <td>  0,1</td>\n",
        "      <td>  1</td>\n",
        "      <td> 3.01</td>\n",
        "      <td>  3</td>\n",
        "      <td> 37</td>\n",
        "      <td>     40,3,0</td>\n",
        "      <td> 1.000</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>3</th>\n",
        "      <td> 1</td>\n",
        "      <td> 877831</td>\n",
        "      <td> T</td>\n",
        "      <td> C</td>\n",
        "      <td> .</td>\n",
        "      <td>                                              PASS</td>\n",
        "      <td> 0</td>\n",
        "      <td> /</td>\n",
        "      <td> 1</td>\n",
        "      <td> 1</td>\n",
        "      <td>...</td>\n",
        "      <td> snp</td>\n",
        "      <td> snp</td>\n",
        "      <td> 1/1</td>\n",
        "      <td> 0,42</td>\n",
        "      <td> 42</td>\n",
        "      <td>   99</td>\n",
        "      <td> 99</td>\n",
        "      <td> 59</td>\n",
        "      <td> 1541,126,0</td>\n",
        "      <td> 1.000</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>4</th>\n",
        "      <td> 1</td>\n",
        "      <td> 881627</td>\n",
        "      <td> G</td>\n",
        "      <td> A</td>\n",
        "      <td> .</td>\n",
        "      <td>                                              PASS</td>\n",
        "      <td> 0</td>\n",
        "      <td> /</td>\n",
        "      <td> 1</td>\n",
        "      <td> 1</td>\n",
        "      <td>...</td>\n",
        "      <td> snp</td>\n",
        "      <td> snp</td>\n",
        "      <td> 1/1</td>\n",
        "      <td> 0,51</td>\n",
        "      <td> 51</td>\n",
        "      <td>   99</td>\n",
        "      <td> 99</td>\n",
        "      <td> 60</td>\n",
        "      <td> 1963,153,0</td>\n",
        "      <td> 1.000</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "<p>5 rows \u00d7 23 columns</p>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 11,
       "text": [
        "  CHROM     POS REF ALT ID                                             FILTER  \\\n",
        "0     1   69428   T   G  .  TruthSensitivityTranche99.90to100.00;LowGQX;Lo...   \n",
        "1     1   69497   G   A  .        TruthSensitivityTranche99.90to100.00;LowGQX   \n",
        "2     1   69508   G   A  .        TruthSensitivityTranche99.90to100.00;LowGQX   \n",
        "3     1  877831   T   C  .                                               PASS   \n",
        "4     1  881627   G   A  .                                               PASS   \n",
        "\n",
        "   multiallele phase  GT1  GT2 ...  vartype1 vartype2   GT    AD  DP    GQ  \\\n",
        "0            0     /    1    1 ...       snp      snp  1/1  11,3  14  3.01   \n",
        "1            0     /    1    1 ...       snp      snp  1/1   1,1   2  3.01   \n",
        "2            0     /    1    1 ...       snp      snp  1/1   0,1   1  3.01   \n",
        "3            0     /    1    1 ...       snp      snp  1/1  0,42  42    99   \n",
        "4            0     /    1    1 ...       snp      snp  1/1  0,51  51    99   \n",
        "\n",
        "  GQX  MQ          PL     VF  \n",
        "0   3  11      39,3,0  0.214  \n",
        "1   3  30      36,3,0  0.500  \n",
        "2   3  37      40,3,0  1.000  \n",
        "3  99  59  1541,126,0  1.000  \n",
        "4  99  60  1963,153,0  1.000  \n",
        "\n",
        "[5 rows x 23 columns]"
       ]
      }
     ],
     "prompt_number": 11
    },
    {
     "cell_type": "heading",
     "level": 4,
     "metadata": {},
     "source": [
      "Mean functional impact score for each Gene - Input into GSEA"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "fiSSEA_results.fi_scores.head()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 12,
       "text": [
        "dbnsfp.genename\n",
        "AADACL3             2.394\n",
        "AAK1                3.307\n",
        "ABCA12              0.008\n",
        "ABCA4              11.010\n",
        "ABCB10              9.549\n",
        "Name: dbnsfp.cadd.phred, dtype: float64"
       ]
      }
     ],
     "prompt_number": 12
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "print fiSSEA_results.fi_scores.describe()\n",
      "fiSSEA_results.fi_scores.hist(bins=20)\n",
      "plt.title('Distribution of Mean CADD Phred Score by Gene (Snyderome, Chr6)')\n",
      "plt.xlabel('Mean CADD Phred Score')\n",
      "plt.ylabel('Gene Count')\n",
      "plt.tight_layout()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "count    1173.000000\n",
        "mean        8.079441\n",
        "std         6.125616\n",
        "min         0.000000\n",
        "25%         3.317000\n",
        "50%         7.351000\n",
        "75%        11.550000\n",
        "max        36.000000\n",
        "dtype: float64\n"
       ]
      },
      {
       "metadata": {},
       "output_type": "display_data",
       "png": "iVBORw0KGgoAAAANSUhEUgAAAaYAAAEbCAYAAACRJSCoAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzt3XmcHGWdx/HPkAMSyJAEkAQIDIIICBJEQBaUUQHBA/FE\nEAjKuoCuGhcEoq4EURAFRGHBAzAcCioqwiKH4JTHKmHRDFFiuDYjRzhiCCScgWT2j99TdnVNV3V1\nT9VUPdPf9+vVr5mq6qr69dPV9dTz/J6uBhERERERERERERERERERERERERERERGRRBcBX8hpW1sC\nq4AuNx0Ax+S0bYBfAkfmuL2svgwsA5aWsO9OMhe4wuPtF2Et8MqygyjImcCnS9hvQL7npaoaAN6a\n4/bOBo7LY0MDwHPASmAF8D/AsdQqjla39ZYW1+kDPtrGvqA6J5EtsTLcKGF5L3by+Fls/i5ufl9h\nkaXbA6vIVwDLgfnA0bHnbI3FeGGD9dcCz2AXGv8AbgU+GHtOADyPHV9PA3cCJwPjU+KaB7zotrsc\nuAV4tVt2KsW+5822/zng/1xsDwFXFxhLVnlWTNOB7wGPYK/xAeD71Mp/JG0CPAysG5k3UuU/nPNS\nlXQD5wF/x8rsfuAb1M5VS2j9nD0ROx8sA54CfhNZNg14EBiXtoF1MuxkEHgn9gK2BL6KnTguaTHY\ncFtpFdrYNrbpgy2xE+jylOcsA94ATI3MmwXci5XbSNsLuA37AG6DHajHAwfGnncU8FfgUBpXJq8F\nJgHbYRXKBcAXI8sHgU9gx9c04ATgQ1iFmGQQOMttdwvgCbdtaO2CKe/jbRZwBHaFOQl4PVYZ56nM\nz8hGwB+A9YB9sNf4OuzEs38J8RwN3IBdpMDIlP9wVekcNx77jO8AvA0rs72wi8jd29jeGPf3u8Bk\nYHtgCjA78pzHgMXAwe2FXNOoxtwdWAPs6KbnAae7/zcG/pvaVfZvsZPFFW6d57Ca+USgB7ua+yhW\nYwfAVm5eWGn2AWdgV+tPA9diLxaspfFQLLYB7MA8EDtgV7v9LXDLA2pN8C6sC3IAeBy4DDtBEont\nKBfbMuxqLMmGwOXYSXIA+Lzb/n7uNa9xcVzaYN3wdVwIfNzNG4NdDf4n9S2m7YFfYWW7GPhAZNk7\n3Ot8GrsqOTWyrNXX83vg/JTlYK/vfuC9WAX6vtjyRlfq78NaSOF72MfQLpEZwLPu9TTyfeBLkel3\nYGUL1kr+EfZersQqzd0izx0ATgIWujjWwS4I/oAds/3AvpHnb42deFdiLbPzSW4xnY9dbSaZ6mJ/\nBHgS+Hlk2ceA+7D39RdYyyS0Fjsu7sNaKGAXi/3UejF2TtnvWuCTbt1lwNew9268i2OnyHNfgZV9\no9b9l6l9jtKklWeAvXe/x8r05ti+0taNuw04PDLdrPzT9n0D8O+x5y8E3u3+3x/7vD3l9hNQf9x+\nFFiEledN2MVoqNH71+z9Pt4tX+li3gb4o9v/1dS3OFo5FqL+FasoJqY8Zwl2sXhXZN9hC7UXO0ed\nBDyKfea2x84/G6Rs83M0Pg+2JKkp93esSw/qTxRnYjmnMe6xd8q2erA3YR4wAXvB4bywYgqwF78j\nVoDXUDsx9DK0Yoru41SssoiKNsE/ir35PcD6wE8jzw/j+I6L67XAC1jBN3I5dqJZH6tc74nsZ98G\ncUaFr2Mv4HY37+3YAX4MtYppffe8WVj5zMRONDtE9vMa9//O2EEXfrBaeT0TgZdJPykAvBGrENYD\nvgJcF1veqGIaB7yEXaFBcpfIb7DWeSPfp3YhtAHwQ2rdBXOxCudA7OR7BvaBDg0AfwY2x8phc+wK\nMWwJ7uemwxPWH7F+8XHu9a5k6DEV+jB2ojkRu1ofE1t+A3AVdhEz1m0P7Hhdhr2f44FvUd/9sRY7\niU52Me+KXUjt7l7jUdhxn9T9uRY7iU/GKv17qJ1U/4v6cv40dqJs5HbqW7uNNCvPAPvMbYsdN33Y\nOSNt3Y0T9vUE9Rcdzco/bd8foPbZA+tG/wf2Pm2Mve/vdducjR3D4XH7brfdV2Ofy89jFUQo/v5l\neb9/jh3bO2IX2L/GPsPdwN3Yew6tHwtRV2OfpTQDWLlMwy4mF1E77/di5XAm9vlYz+1/IXCue40L\nsXKLei/wpwzxpUqqmP4IzHH/Ryum07BWzTYZttWDvQk9DebFW0yhHbA3qovmFdNchl7dRk+Et1Gf\niNsOa2GtE4ljs8jy+ViXVdwYF1P0JP9v1CqURnFGRZff6+K4GjiM+orpUKwFGvUdkk8W52EHCLT2\nejZ3z90uJWaAi7FKAexDthrr9w8l5TYexV4bJFdMV2GvrZF5WOWzwm3rWqxlA/ae3xJ57o5YizW0\nhPo82ckMrWhuwj5gW2IfvAmRZT8gPcd0ONaifQY7sZ3k5k/HWs0bNljnEuorh/Wxsgyvutdix0jo\nIupbjGBX829KiGktcEBk+nhqXVx7YheZoTuB9yds5z7suA4djL0HYesD0ssT7P2OttSPB27MuG7c\naoYeo0nl32zf62GtnfC8dTbW7Yzb/x9i+3mI2nF7I/XH8DpYq3OGm46/f1ne770iy+8EPhuZPpta\ny7DVYyHqFurPrY0sob5VepbbJ9hrepH6SvBzWPxfxCr1N2EXr9Fz4/7UWo4NZckxJdkCeyNDYd/+\n17HunVvczk/OsK20k3Z8eZg4S7qKasV06j+UD2KFuWlk3mOR/5/DDqK4jV1M8W1t3kZMV2DdLr3Y\nVVM0Z7IVdiJZEXkcHol3T+zD9wTW7D6WoV0yWV7PCuzgmt5gWWgCdgL7iZvux66uDk9awRmHVV5P\nNnle/PiKGsSOsykuxkOwD1Do8cj/z2EnneixHj2etsKulqNlujd2hbiZm34+8vzoe9zID7EP3obY\nRc/pWKUww72epxusEz8On8Wu/KPHTzzmE2Ixb0H6+xX/DIUXKPOx19eLnTy2YWjLN7Sc+gub67D3\n4DPUTk5p5RmKHoPPU+v2ybJu1AosLxLVqPyj+a+kfb8A/BgbsduF5TnDC5DNsF6bqPj78c1IzGEu\nOen9y/J+R4/h52PTL1D73LZzLITi72eSpDIDaxWtji1/Cev2fRm7kO6j/sJoEnZ+StRuxbQ79oJ+\n32DZM1hTehvsiuo/gDe7ZUlJ/GbJ/S1j/7+EXQ09S33/6Bjqr9ibbXcp9a21LbHCfLzhs5P9w8UU\n31b8YM7iSuxK7gbsAIx6EGvyT4k8JmGDB8A+lNdiB+Zk4Nu09x4/h7WIk66cAd6DdSt8B2u1PIqd\nfGc12fa7sTK+I+U5M7Ck+u9SnpM0yCHLQJHocx7ETkDxMv0a9pqmUH+MbZVxH2uwbueFWPfqg1iO\nqVGLKX4cro9dUDySEvNXYjFvgOXWksQ/Q9FtX4YNGjgSu9CInmiibsMuAuJlH51OK89mWl13Icmj\nAaPlv1PCc+Iuw7oDw7zwfDd/KbXWD9jrjU4/iLUko3GvT33XYPT9y/J+pxnusRC6FetST8sxtRIL\nWHnD0GMk+rwdsAvZRFlPWuFOurFE21XYAXR3gyDeifXhdmFN/DXY1TfYCb9RF1+zfR+BvZiJWLP1\nJ9gLvRe7Gn47diX+BeqHjj6GHQBJJ7GrsKu9HuzNPAPrQlub8Pwwnrg12NXWV9x2tnLbvTL9pTW0\nBGv+fr7BshuwrosjsNc7DrtICJvJG2BXTKuxod6H0/wkmlQ2J2FdXidSa3XtgpUZWAV0Cfah38U9\n9nZ/oyeCcPtTsQ/9BVg3xooGz5mI5bV+gZ0UkkbmpY28a/VrDFcC78Ku6MZgx1MvdvX6d6wb5TSs\nrPfBju8ks7BjcRL22ToIq5TmY8fijdgAl8lue2F3y1XAR7CyWxc7Dm/HTjqNfA9rDeyBvd71sQEg\naQnnE6nlmD5F/YnrSqzf/8Mk58/AuoWnYJ/9V7p9T8K6ccPjLK08Q0nvUZZ1o35JfR40rfyb7Rvs\nYmwQ6yqLlsMNbjvvwXpUPkV9K+7bWBdWOBhsQ+oHJcW1+n7H4+6KTDc7FuaRnEe6AmvJ/ZRafmwj\n91oOSoklzW+w1zEHK6u9sffw5shz9qXWhdq2JdS+x/QUltQ7nvqCiuaYZrt1nsFedPQEezD2YV+B\ntaR6sJN6tIKMz+vDTvjhqLxfMHRI9VKs0jsB+w5DmGOail11P4mdYMLthf3BXdiotwex7q/LqV3R\nNoot7bsLk7E3+gm3vS9QK6Ne0g+6tOXHYInP0HbYqMcnqH036LVu2fuw7rSVwPVYUjU6mKOV1wNW\n6f0Se9+XYx+eI7ATxUvUBlpE3UDtCjf6Pabl2BX3h2LP76P2PaaV2MCEOaQnb+Oj8qLiA156qH/d\njXKme2CJ8eVYuV5P7Yp4a6w7YhXWPR0t07j3YL0IYZfdXdTnR6ZgJ4rH3HOuiSw7FusCX451kUW7\nWNYwNFf3NqzVuQI7/n9EcsW0Fhtx9gB2zHydoRelt2KfnWamY7nFpdS+9xL/HlOj8tzCLYsfc7Oo\nz5umvRdxG2HnmPXcdLPyb7ZvsM9tPO8NVt73UBuVF9/WEVhrIRwRe3FkWaP3r5X3+3ex13E6NiQ7\nGlv8WAi7+m4j/YvA3Vi+6kFq7+fZ1EbNxj8v0c9XL43PWztiOblnsFGx744sm469Z6UNm5+BvXl3\nY8F9ys2fiiUn78U+6JMj68zBEqyLqe+TFJFiXUJyZV9lXyHfOz8cydDKylfjsfNvfHRimXK780O7\npmFNfLAruXuw7rivURspczK10Sk7Yv2O47CrlfsZ3uAMEcmmB7va3qrkOMo2kVqvgHSIa7Gk4mJq\no8imuWmw1lJ0BN9N2JftRKQ4p2NdOHOaPXGUexvW9fRzdEHcMXqw3NIkhia9w+nzseRr6GKG3klA\nRERGuZFIQG2Ajfr4NLXbxoQGSR81NmTZZpttNrh0qW7QLSLSogewEdOVV3STdRxWKV2BdeWBjZ4L\nh1pOx0begI3hj46+2YIG4/qXLl3KmDHj23pcc801DA4Ojshj1qxZI7av0R6nDzEqTsVZ9Qetf1Wn\nNEVWTF3YSJ9F2K1xQtdR+xLmLGoV1nXYUOLx2BDdV5HwJcw1a1a2/Bg//iiWLVtWwMsUEZE8FdmV\ntze1sf3hHYnnYKPwfoyNrR+g9vs8i9z8RdidAT5OYjffuo1npxrZEZM9PT0jur92+RCnDzGC4syb\n4uxcRVZMvye5RbZfwvwzaH5TQS/09vaWHUImPsTpQ4ygOPOmODuXhkWKiEilqGISEZFKafWGl1Uw\n2M4vjU+YcBznnjuT444r9W4YIiKl6OrqAk/O+WoxiYhIpahiKkgQBGWHkIkPcfoQIyjOvCnOzqWK\nSUREKsWL/sYY5ZhERFqkHJOIiEibVDEVxJd+Zx/i9CFGUJx5U5ydSxWTiIhUihf9jTHKMYmItEg5\nJhERkTapYiqIL/3OPsTpQ4ygOPOmODuXKiYREakUL/obY5RjEhFpkXJMIiIibVLFVBBf+p19iNOH\nGEFx5k1xdi5VTCIiUile9DfGKMckItIi5ZhERETapIqpIL70O/sQpw8xguLMm+LsXKqYRESkUrzo\nb4xRjklEpEXKMYmIiLRJFVNBfOl39iFOH2IExZk3xdm5VDGJiEileNHfGKMck4hIi5RjEhERaZMq\npoL40u/sQ5w+xAiKM2+Ks3OpYhIRkUrxor8xRjkmEZEWKcckIiLSJlVMBfGl39mHOH2IERRn3hRn\n51LFJCIileJFf2OMckwiIi1SjklERKRNqpgK4ku/sw9x+hAjKM68Kc7OpYpJREQqxYv+xhjlmERE\nWqQck4iISJtUMRXEl35nH+L0IUZQnHlTnJ1LFZOIiFSKF/2NMcoxiYi0SDkmERGRNqliKogv/c4+\nxOlDjKA486Y4O5cqJhERqRQv+htjlGMSEWmRckwiIiJtKrpiuhR4HPhLZN5c4GFggXscFFk2B7gP\nWAwcUHBshfKl39mHOH2IERRn3hRn5yq6Yvo+cGBs3iBwLrCre9zo5u8IHOr+HghcOALxiYhIxYxE\nf2MPcD2ws5s+FXgGOCf2vDnAWuAsN30T1rq6PfY85ZhERFqkHFNznwTuAi4BJrt5m2FdfKGHgc1H\nOC4RESnZ2BL2eRHwJff/6VjL6ZiE5yY0jY7GGmJg9dpMoNdNB+5vfNpNuf7g3t7eQqfDeSO1v3an\nzzvvPGbOnFmZeBpN9/f3M3v27MrEkzQdf+/LjidpWuXZGeUZBAHz5s0DoKenB6nXQ/3gh6Rlp7hH\n6CZgzwbrDMJgy48JE44dvOiiiwZHSl9f34jtazh8iNOHGAcHFWfeFGe+aCcHUpIyckzTgUfd/58B\ndgcOxwY9/BDYA+vCuxXYlqGFqRyTiEiLfMoxFd2VdxWwL7Ax8BA28KEX63sbBJYAx7rnLgJ+7P6+\nDHwcj2p4ERHJR9GDHw7DBjWMB2Zg32s6CngtsAtwCPY9p9AZWCtpe+DmgmMrVLR/vMp8iNOHGEFx\n5k1xdi59T0hERCrFi/7GGOWYRERa5FOOSS0mERGpFFVMBfGl39mHOH2IERRn3hRn51LFJCIileJF\nf2OMckwiIi1SjklERKRNqpgK4ku/sw9x+hAjKM68Kc7OpYpJREQqxYv+xphSckzd3VNZtWpFW+tO\nmjSFlSufbGtdEZE8+JRjKuNnL7xklVJ7t+5btcqLY0FEpBLUlVeYoOwAMvGhf9yHGEFx5k1xdi5V\nTCIiUik+9jGVkmOy/tl2f4WjC/udLhGRcviUY1KLSUREKkUVU2GCsgOgu3sqXV1dbT26u6eWHf4/\n+dKHrzjzpTg7lyqmUaw2kjDt0ddwfrtD40VEhsuL/sYY5Ziy7tXDmEWkGMoxiYiItEkVU2GCsgPI\nKCg7gKZ86cNXnPlSnJ1LFZOIiFSKF/2NMR7mmMYBL7e15nDus6cck4iEfMox6V55I+JldJ89EZFs\nsnTlXZFxntQJyg4go6DsAJrypQ9fceZLcXauLBXTTrHpscBuBcQiIiKS2t/4OWAOMAF4PjL/JeC7\nwCkFxpXGwxxTObke5ZhEJORTjimtxXQGMAk42/0NH1Mpr1ISEZFRLktX3inA5sC/AG+KPCRVkNN2\nxrZ9v7uRjbM4vvThK858Kc7OlWVU3lnAocAiYE1k/m8LiUhi2h/R50mrXUSkTpYz173AzsCLBceS\nVcflmHzLbYlI9YyWHFPoAWB80YGIiIhAtorpeaAfG4l3vnt8q8igRoeg7AAyCsoOoClf+vAVZ74U\nZ+fKkmO6zj2i1McjIiKF8KK/MUY5phFaVzkmkdHDpxxTlhbTkgbzBoFX5hyLiIhIphzT7pHHG4Fv\nAj8oMqjRISg7gIyCsgNoypc+fMWZL8XZubJUTP+IPB4GzgPeUWRQIiLSubL0N+5GLVGxDvB64Hhg\nl6KCakI5phFaVzkmkdFjtOWYzqF2dnsZGAA+WFRAIiLS2bJ05fUCb3aP/YGPAfcUGNMoEZQdQEZB\n2QE05UsfvuLMl+LsXFkqpsnAN4A/ucc5wIZFBlWU2bM/W/ANUUVEZLiynHF/BvwFuMw9/0jgtcB7\nC4wrTds5puef/w4+5nqUYxKR4RptOaZtqK+E5gJ3FRKNiIh0vKz3yntjZHof4LliwhlNgrIDyChI\nmN/+70B1d0/NN0JP+vAVZ74UZ+fK0mI6DricWl5pBTCrsIikItr/HahVq7zoLRCRimrlDBJWTE8X\nEUgLlGPyYF3lp0SqxaccU1pX3gnAv0amn3aPY4DZRQYlIiKdK61i+jDWhRd3BVY5ZXEp8Dg2qi80\nFfgV9su4t2DD0UNzgPuAxcABGfdRUUHZAWQUlB1AU7704SvOfCnOzpVWMY0FVjeYv5rszcHvAwfG\n5p2CVUzbAbe5aYAdgUPd3wOBC5vEJyIio1BaBfMX7E4Pj8XmbwrcCuyccR89wPWR5y8G9sVaUtOw\nS/btsdbSWuAs97ybsKHpt8e2pxyTB+sqxyRSLaMlx/R14AbslkST3OPNbt45w9jnplilhPu7qft/\nM+zu5aGHgc2HsR8REfFQ2nDxy4FlwJeA17h5dwP/CdyY0/4HSb8sT1h2NNYQA0tRzcTqT6jlTOLT\nNFme9/rhvOHuP5yX9fmtTp9H4/IjNt3q9t2U63/v7e1te7q/v5/Zs2fntr2ipqO5hirEkzSt8uyM\n8gyCgHnz5gHQ09OD1OuhfvDDYqwLD2C6mwbLNZ0Sed5NwJ4NtjcIgy0/Jkw4drDdde3R6rp9w1h3\nOPttdd2+Qvabp76+vly3VxTFmS/FmS/a75sfcSPR39hDfY7pa8ByLJd0CtbkOQUb9PBDYA+sC+9W\nYFuGFqZyTB6sO6gck0il+JRjynLnh+G4ChvosDHwEPBF4KvAj7Eh5wPUfttpkZu/CLvtwMfxqIYX\nEZF8FD0c+zBsUMN4YAY2fPxJYD9suPgBwFOR55+BtZK2B24uOLaCBWUHkFFQdgBN+fI9EcWZL8XZ\nubJUTNOAS7CcD1iXW9Yv2IqIiLQkS3/jTVhL5/PY7zCNAxYAOxUYVxrlmDxYVzkmkWrxKceUpcW0\nMfAjYI2bfgnLAYmIiOQuS8X0DLBRZPoNlH+HcQ8EZQeQUVB2AE350oevOPOlODtXllF5J2DDvV8J\n/AHYBHh/kUGJiEjnytrfOA54tfv/Hqw7ryzKMXmwrnJMItXiU44p6/eY9sC+KDsWeJ2b1+gnMURE\nRIYlS47pSuyGrnsDrwd2dw9JFZQdQEZB2QE05UsfvuLMl+LsXFlaTLth311S34yIiBQuS3/jT4BP\nA0sLjiUr5Zg8WFc5JpFqGW05pk2w+9fdAbzo5g0CBxcVlIiIdK4sOaa5wCHAV7AfCAwfkiooO4CM\ngrIDaMqXPnzFmS/F2bmytJgCbETetthPUUzMuJ6IiEjLsvQ3/hvwMWAqsA12V/CLgLcWGFca5Zg8\nWFc5JpFq8SnHlKUr7xPAPsBKN30v8IrCIhIRkY6WpWJ6kdqgB7BuPF0ONxWUHUBGQdkBNOVLH77i\nzJfi7FxZKqbfYD95MRHYHxs+fn2RQYmISOfK0t84BvthwAPc9M3AxZTXalKOyYN1lWMSqRafckxZ\nRtetAb7rHiIiIoVK68o7BPj3yPQdwBL3+ECRQY0OQdkBZBSUHUBTvvThK858Kc7OlVYxnQRcF5ke\nj93EdV/g+CKDEhGRzpXW33gnVhGFLqDWgpoP7FlUUE0ox+TBusoxiVSLTzmmtBbTlNh0tFtvkwJi\nERERSa2Y5mN3fYg7zi2TVEHZAWQUlB1AU7704SvOfCnOzpU2Ku8zwLXA4cCf3bzXAethAyNERERy\n16y/sQt4C/AaLOFwN/DrooNqQjkmD9ZVjkmkWnzKMTX7HtMgcJt7iIiIFC7LLYmkLUHZAWQUlB1A\nU7704SvOfCnOzqWKSUREKsWL/sYY5Zg8WFc5JpFq8SnHpBaTFGAsXV1dbT26u6eWHbyIlEwVU2GC\nsgPIKChgmy9jra3WH6tWrRgaoSd9+IozX4qzc6liEhGRSvGivzFGOaZRvq7yUyL5U45JRESkTaqY\nChOUHUBGQdkBNOVLH77izJfi7FyqmEREpFK86G+MUY5pVK87DhvV155Jk6awcuWTba8vMlr5lGNq\ndq88kREWDjVvz6pVXnzuRCSFuvIKE5QdQEZB2QFkEJQdQCa+5BoUZ758idMnqphERKRSfOz3UI5J\n66aur+9BiQzlU45JLSYREakUVUyFCcoOIKOg7AAyCMoOIBNfcg2KM1++xOkTVUwiIlIpXvQ3xijH\npHVT11eOSWQo5ZhERETaVGbFNAAsBBYAd7h5U4FfAfcCtwCTS4ksF0HZAWQUlB1ABkHZAWTiS65B\ncebLlzh9UmbFNAj0ArsCe7h5p2AV03bAbW5aREQ6SJn9jUuA1wPLI/MWA/sCjwPTsEvl7WPrKcek\ndVPXV45JZCjlmLIZBG4F7gQ+5uZtilVKuL+blhCXiIiUqMybuO4NPApsgnXfLY4tHyTx0vlooMf9\nPxmYifUKQi0fEZ+myfK81w/nDXf/4bysz291+jwalx+x6Va33+764bzo8n5gdub1gyCgt7f3n/8D\nIzIdzTWUsf+s0/39/cyePbsy8SRNqzyHX37z5s0DoKenB2ndqcAJWOU0zc2bztDKCqwrr+XHhAnH\nDra7rj1aXbdvGOsOZ7+trttX0n5bWTcpxsbrl6Wvr6+0fbdCcebLlzjts+WHsvobJwJjgFXA+tgI\nvNOA/bCc01nYwIfJDB0AMdhO+SrH1Anr2vqDyjGJDOFTjqmsrrxNgZ9HYvgBVjndCfwYOAYbTv7B\nMoITEZHylDX4YQmW2JgJ7ASc6eY/ibWatgMOAJ4qJbpcBGUHkFFQdgAZBGUHkIkv32dRnPnyJU6f\n6M4PIiJSKV70N8Yox6R1U9dXjklkKJ9yTGoxiYhIpahiKkxQdgAZBWUHkEFQdgCZ+JJrUJz58iVO\nn6hiEhGRSvGivzFGOSatm7q+ckwiQynHJCIi0iZVTIUJyg4go6DsADIIyg4gE19yDYozX77E6RNV\nTCIiUile9DfGKMekdVPXV45JZCjlmERERNqkiqkwQdkBZBSUHUAGQdkBZOJLrkFx5suXOH2iiklE\nRCrFi/7GGOWYtG7q+soxiQylHJOIiEibVDEVJig7gIyCsgPIICg7gEx8yTUoznz5EqdPVDHJKDOW\nrq6uth7d3VPLDl5E8KS/MUY5Jq1b2L6Vn5LRSjkmERGRNqliKkxQdgAZBWUHkEFQdgCZ+JJrUJz5\n8iVOn6hiEhGRSvGivzFGOSatW9i+lWOS0Uo5JhERkTapYipMUHYAGQVlB5BBUHYAmfiSa1Cc+fIl\nTp+oYhIRkUrxor8xRjkmrVvYvpVjktFKOSYREZE2qWIqTFB2ABkFZQeQQVB2AJn4kmtQnPnyJU6f\nqGISEZFK8aK/MUY5Jq1b2L6VY5LRSjkmERGRNqliKkxQdgAZBWUHkEFQdgCZ+JJrUJz58iVOn4wt\nOwCR0aC7eyqrVq1oc+1xwEttrTlp0hRWrnyyzf2KVJMX/Y0xyjFp3cL23W6Oyfrvyykv5cUkC59y\nTGoxifwsrvHlAAAJwUlEQVTT2PDDKyIlUo6pMEHZAWQUlB1ABsEI7edlrOXSzmMk4xweX3IiirNz\nqWISEZFK8bHfQjkmrVvBfZe17jispdc6DZzoLMoxicgICbsfW7dqlRfnKOlA6sorTFB2ABkFZQeQ\nQVB2ABkFZQeQUVB2AJn4krvxJU6fqGISEZFK8bEtrxyT1q3gvv1cV9+B6hw+5ZjUYhIRkUpRxVSY\noOwAMgrKDiCDoOwAMgrKDiCjoOwAMvEld+NLnD5RxSQiIpXiRX9jjHJMWreC+/ZxXX0HqpP4lGPS\n95hEOpa+AyXVpK68wgRlB5BRUHYAGQRlB5BRUHYAGQVlB5CJL7kbX+L0SRUrpgOBxcB9wMklxzIM\n/WUHkJEPcfoQIyjOfPX35xtnd/dUurq62np0d08dsTilel15Y4ALgP2AR4D/Ba4D/lZmUO15quwA\nMvIhTh9iBMWZr6eeyjdO+yHH/Lsu845Tqtdi2gO4HxjAftLzauDdZQYkIiIjq2otps2BhyLTDwN7\nxp/U3f2ulje8evXC9qNqy8AI769dA2UHkMFA2QFkNFB2ABkNlB1AJgMDA0PmDe8n7Icj/UckTzvt\ntMRlGsHYuqoNrXkflmP6mJs+AquYPhl5zv3ANiMcl4iI7x4Ati07iCyq1mJ6BJgRmZ6BtZqivChY\nEREZHcZitXoPMB4bPrRDmQGJiIgcBNyDddnNKTkWERERERERP/jyxdsBYCGwALij3FDqXAo8Dvwl\nMm8q8CvgXuAWYHIJccU1inMulmtc4B4HjnxYQ8wA+oC7gb8Cn3Lzq1amSXHOpVpluh4wH+u+XwSc\n6eZXqTyTYpxLtcoyNAaL53o3XaWyHBXGYF17PdidJ6uce1qCHQBV80ZgV+pP+F8DTnL/nwx8daSD\naqBRnKcC/1FOOImmATPd/xtg3c87UL0yTYqzimU60f0dC9wO7EP1yrNRjFUsS7CYfoDdpACqV5aJ\nqvYF2yS+ffG2asPwAX4HxL8AcjBwmfv/MuCQEY2osUZxQvXK9DFq9/Z5Brs7yeZUr0yT4oTqlelz\n7u947GJ0BdUrz0YxQvXKcgvg7cDF1GKrWlkm8qViavTF280Tnlu2QeBW4E5q38eqqk2xbjPc301L\njKWZTwJ3AZdQvS6IHqyVN59ql2kPFuftbrpqZboOVok+Tq37sWrl2ShGqF5ZfgP4LLA2Mq9qZZnI\nl4ppOD/QM9L2xj78BwGfwLqmfDBIdcv5ImBrrEvqUeCccsOpswHwU+DTwKrYsiqV6QbANVicz1DN\nMl2LxbMF8CbgzbHlVSjPeIy9VK8s3wk8geWXklpyVSjLRL5UTFm+eFsVj7q/y4CfY92QVfU4loMA\nmI4dzFX0BLUP0sVUp0zHYZXSFcC1bl4VyzSM80pqcVa1TAGeBm4AdqOa5Qm1GF9P9cryX7BuuyXA\nVcBbsGO0qmU5hC8V053Aq6h98fZQagm9KpkITHL/rw8cQH0Sv2quA2a5/2dRO2lVzfTI/++hGmXa\nhXXbLALOi8yvWpkmxVm1Mt2YWhfYBGB/7Iq/SuWZFOO0yHOqUJafwy7etwY+BPwaOJJqleWo4cMX\nb7fG+p/7saG5VYrzKmApsBrL130EGz14K9UaPhqP86PA5dgQ/LuwD1MV+sb3wbp1+qkfJly1Mm0U\n50FUr0x3Bv6MxbkQy49AtcozKcaqlWXUvtQu4qtUliIiIiIiIiIiIiIiIiIiIiIiIiIiIiIiInla\ni33jPDQWuyPG9Y2fnqsTsRuWhj8/cmRk2cbYzYCPja0zgH0fZSF2D7TTgXXdsh7geex7LIuw++LN\norFe7M4AC9xzv+jmHw2c386LSZC0vU2B/8a+b3M3docCEW+MLTsAGdWeBV6D/Y7NC9g35R+m+Ht0\nHQe8Fdgduy/cJOwb+aEPADcBhwHficwfxCqVJ7E7d3zXLT/aLb8feJ37f2vgZ9idFeY1iOG3wLuw\nu4H0Y5Vx1te9DvU332zVl4CbqVVaOw1jW6ExwJoctiPSlC+3JBJ//RJ4h/v/MOzODuGNJdfHfhhw\nPtYSOdjN78FO7H9yj73c/F4gAH6CtYauTNjnHOB4rFICu7nq5ZHlHwK+ALyC5LvUP4tVcIfQ+Bvy\nS7Dfu/lUg2VRz7nXsK2b3gy4Efv2/VmR5z0DnI1VYnsBR2DlsgD4NrXP6kewO6DMx+6J1sg07P6S\nob9G/j8ZaxH2U/uhu5nYHcfvwirb8PUG2F2q/9e9zt3cvDuxij16Kx4RES+swm7j8hOsS2wBdouU\nsCvvDODD7v/J2Al3InYfsrAL7VXYiRGsYnoKO7l3AX/A7uYe1Y21eJLMwCo1sJZF9AfeGv3I4wKs\n5dXD0HugTab2+zxRvdRe40ZuuztgLa8HsBbculjXYVgxrgXe7/7fAbuNzBg3fSHWFTkd+Lvb5jjg\n98C3Guz/AOx3gn6N3TctvC/eQcD/YC3YMH6wiiq8C/5pWGUE9rMOF7j/x2LlvZGbPhS7B59I7tSV\nJ0X7C3ZSP4yhuY4DsO6uE930uljF8Rh2QtwF6z56VWSdO7B76YFd9fdgJ9usDsV+/gGswrwUODfl\n+Wk/AJe27I1YK3At1jL5G7AncBu1n8dYBGyFtW7WYHcAB+uG3A1rmYBVJI9hd60OgOVu/o+A7Rrs\n+xbgldi9+w7CKtedgP2w1/uCe95TwIbu8Ts37zKsXEI/cn+3x7plb3XTY6i9DyK5UsUkI+E6rJtq\nX2CT2LL3AvfF5s3Ffj7kSOwE+EJk2YuR/9cw9BheiXWLbY21VOIOwwYHHOGmpwPbYC2ZuElYxXcv\nMKXB8l2xyqWR32GVbtRgSvwvUJ+Dugxr7UTFf7U5rWJcgXWbXoW13t7ktt/sl1bjy5+NzL+b5O5D\nkdwoxyQj4VKssrk7Nv9m6nM0u7q/3VgLAeAoal1aWZ0J/Be1nyDZAKvktsPyWltgFdfWwFeBwyPr\ndkXWuRD7Ta2nG+yjB/g6rY2yy/rz27dh3XphJT4V2BLLK+3rpsdhgzgaeTPWJQpWBttgXYC/wnJU\nE9yyKdhrW4HdhRysnIIGMd/j4nmDmx4H7Jjx9Yi0RC0mKVLYAniEWq4i+suZp2O/EbQQu0j6P2wA\nxIVYt9ZRWJI9HMQQ3WbSNNgvim6A5aZeco9zsEEPP4s996fA1S4WsLxKl4vnZ5H5YCf4P2Nda6uA\nb1I/qCIaU6O40n41NDr/b9jgjFtcHC8BH8e6MecCf8S64RYkbG83rLxfdut/DxuAATbQ4U7sZ0Vu\ncPuZhQ2wmIi1HD/SIK7VWGX5LazrbyyWi0pqMYqIiIiIiIiIiIiIiIiIiIiIiIiIiIiIiIiIiFTW\n/wMNZqbll8WilAAAAABJRU5ErkJggg==\n",
       "text": [
        "<matplotlib.figure.Figure at 0x1160b5210>"
       ]
      }
     ],
     "prompt_number": 13
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "fiSSEA_results.mv_annot[['dbnsfp.genename','dbnsfp.cadd.phred']].head(10)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>dbnsfp.genename</th>\n",
        "      <th>dbnsfp.cadd.phred</th>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>_id</th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>chr2:g.75907351T&gt;C</th>\n",
        "      <td>  C2orf3</td>\n",
        "      <td>  9.424</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr1:g.160395003G&gt;A</th>\n",
        "      <td>     NaN</td>\n",
        "      <td>    NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr1:g.202574784G&gt;A</th>\n",
        "      <td>     NaN</td>\n",
        "      <td>    NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr1:g.204924020C&gt;T</th>\n",
        "      <td>   NFASC</td>\n",
        "      <td> 16.530</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr2:g.29225504T&gt;C</th>\n",
        "      <td> FAM179A</td>\n",
        "      <td>  1.584</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr2:g.198950240G&gt;A</th>\n",
        "      <td>   PLCL1</td>\n",
        "      <td> 20.200</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr2:g.26399214G&gt;A</th>\n",
        "      <td>     NaN</td>\n",
        "      <td>    NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr1:g.116569530G&gt;C</th>\n",
        "      <td>     NaN</td>\n",
        "      <td>    NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr2:g.220286142G&gt;A</th>\n",
        "      <td>     NaN</td>\n",
        "      <td>    NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr3:g.9027504T&gt;G</th>\n",
        "      <td>  SRGAP3</td>\n",
        "      <td>  5.757</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 14,
       "text": [
        "                    dbnsfp.genename  dbnsfp.cadd.phred\n",
        "_id                                                   \n",
        "chr2:g.75907351T>C           C2orf3              9.424\n",
        "chr1:g.160395003G>A             NaN                NaN\n",
        "chr1:g.202574784G>A             NaN                NaN\n",
        "chr1:g.204924020C>T           NFASC             16.530\n",
        "chr2:g.29225504T>C          FAM179A              1.584\n",
        "chr2:g.198950240G>A           PLCL1             20.200\n",
        "chr2:g.26399214G>A              NaN                NaN\n",
        "chr1:g.116569530G>C             NaN                NaN\n",
        "chr2:g.220286142G>A             NaN                NaN\n",
        "chr3:g.9027504T>G            SRGAP3              5.757"
       ]
      }
     ],
     "prompt_number": 14
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "fiSSEA_results.mv_annot['dbnsfp.genename'].value_counts().head(10)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 15,
       "text": [
        "PDE4DIP     45\n",
        "HRNR        27\n",
        "ANKRD36     26\n",
        "OBSCN       24\n",
        "IGFN1       23\n",
        "NBPF10      22\n",
        "FSIP2       20\n",
        "MST1P9      17\n",
        "DNAH14      15\n",
        "HNRNPCL1    14\n",
        "dtype: int64"
       ]
      }
     ],
     "prompt_number": 15
    },
    {
     "cell_type": "heading",
     "level": 4,
     "metadata": {},
     "source": [
      "All MyVariant.info Annotations for SNPs flattened into a dataframe"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "fiSSEA_results.mv_annot.head(1).T"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th>_id</th>\n",
        "      <th>chr2:g.75907351T&gt;C</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>_id</th>\n",
        "      <td>                             chr2:g.75907351T&gt;C</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>cosmic.allele1</th>\n",
        "      <td>                                            NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>cosmic.allele2</th>\n",
        "      <td>                                            NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>cosmic.chrom</th>\n",
        "      <td>                                            NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>cosmic.chromEnd</th>\n",
        "      <td>                                            NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>cosmic.chromStart</th>\n",
        "      <td>                                            NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>cosmic.mut_freq</th>\n",
        "      <td>                                            NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>cosmic.mut_nt</th>\n",
        "      <td>                                            NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>cosmic.tumor_site</th>\n",
        "      <td>                                            NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.1000gp1.ac</th>\n",
        "      <td>                                            419</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.1000gp1.af</th>\n",
        "      <td>                                      0.1918498</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.1000gp1.afr_ac</th>\n",
        "      <td>                                            244</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.1000gp1.afr_af</th>\n",
        "      <td>                                       0.495935</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.1000gp1.amr_ac</th>\n",
        "      <td>                                             31</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.1000gp1.amr_af</th>\n",
        "      <td>                                     0.08563536</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.1000gp1.asn_ac</th>\n",
        "      <td>                                             38</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.1000gp1.asn_af</th>\n",
        "      <td>                                     0.06643357</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.1000gp1.eur_ac</th>\n",
        "      <td>                                            106</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.1000gp1.eur_af</th>\n",
        "      <td>                                      0.1398417</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.aa</th>\n",
        "      <td>                                            NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.aa.aapos_fathmm</th>\n",
        "      <td> [ENSP00000318690:T594A, ENSP00000386552:T556A]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.aa.aapos_sift</th>\n",
        "      <td>                          ENSP00000318690:T594A</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.aa.alt</th>\n",
        "      <td>                                              A</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.aa.codonpos</th>\n",
        "      <td>                                              1</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.aa.pos</th>\n",
        "      <td>                                     [594, 556]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.aa.ref</th>\n",
        "      <td>                                              T</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.aa.refcodon</th>\n",
        "      <td>                                            ACT</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.allele1</th>\n",
        "      <td>                                              T</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.allele2</th>\n",
        "      <td>                                              C</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>dbnsfp.ancestral_allele</th>\n",
        "      <td>                                              T</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>...</th>\n",
        "      <td>...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>evs.hgvs.coding</th>\n",
        "      <td>                                            NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>evs.hgvs.protein</th>\n",
        "      <td>                                            NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>evs.ma_fin_percent.african_american</th>\n",
        "      <td>                                        47.0949</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>evs.ma_fin_percent.all</th>\n",
        "      <td>                                        24.8962</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>evs.ma_fin_percent.european_american</th>\n",
        "      <td>                                        13.5233</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>evs.on_illumina_human_exome_chip</th>\n",
        "      <td>                                            yes</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>evs.polyphen2</th>\n",
        "      <td>                                            NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>evs.polyphen2.class</th>\n",
        "      <td>                                              b</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>evs.polyphen2.score</th>\n",
        "      <td>                                              e</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>evs.ref_base_ncbi</th>\n",
        "      <td>                                              T</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>evs.rs_id</th>\n",
        "      <td>                                      rs6722682</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>gwassnps.allele1</th>\n",
        "      <td>                                            NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>gwassnps.allele2</th>\n",
        "      <td>                                            NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>gwassnps.chrom</th>\n",
        "      <td>                                            NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>gwassnps.chromEnd</th>\n",
        "      <td>                                            NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>gwassnps.chromStart</th>\n",
        "      <td>                                            NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>gwassnps.pubmedID</th>\n",
        "      <td>                                            NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>gwassnps.rsid</th>\n",
        "      <td>                                            NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>gwassnps.trait</th>\n",
        "      <td>                                            NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>mutdb.allele1</th>\n",
        "      <td>                                            NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>mutdb.allele2</th>\n",
        "      <td>                                            NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>mutdb.chrom</th>\n",
        "      <td>                                            NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>mutdb.chromEnd</th>\n",
        "      <td>                                            NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>mutdb.chromStart</th>\n",
        "      <td>                                            NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>mutdb.cosmic_id</th>\n",
        "      <td>                                            NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>mutdb.dbsnp_id</th>\n",
        "      <td>                                            NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>mutdb.mutpred_score</th>\n",
        "      <td>                                            NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>mutdb.strand</th>\n",
        "      <td>                                            NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>mutdb.uniprot_id</th>\n",
        "      <td>                                            NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>snpedia.text</th>\n",
        "      <td>                                            NaN</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "<p>242 rows \u00d7 1 columns</p>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 16,
       "text": [
        "_id                                                               chr2:g.75907351T>C\n",
        "_id                                                               chr2:g.75907351T>C\n",
        "cosmic.allele1                                                                   NaN\n",
        "cosmic.allele2                                                                   NaN\n",
        "cosmic.chrom                                                                     NaN\n",
        "cosmic.chromEnd                                                                  NaN\n",
        "cosmic.chromStart                                                                NaN\n",
        "cosmic.mut_freq                                                                  NaN\n",
        "cosmic.mut_nt                                                                    NaN\n",
        "cosmic.tumor_site                                                                NaN\n",
        "dbnsfp.1000gp1.ac                                                                419\n",
        "dbnsfp.1000gp1.af                                                          0.1918498\n",
        "dbnsfp.1000gp1.afr_ac                                                            244\n",
        "dbnsfp.1000gp1.afr_af                                                       0.495935\n",
        "dbnsfp.1000gp1.amr_ac                                                             31\n",
        "dbnsfp.1000gp1.amr_af                                                     0.08563536\n",
        "dbnsfp.1000gp1.asn_ac                                                             38\n",
        "dbnsfp.1000gp1.asn_af                                                     0.06643357\n",
        "dbnsfp.1000gp1.eur_ac                                                            106\n",
        "dbnsfp.1000gp1.eur_af                                                      0.1398417\n",
        "dbnsfp.aa                                                                        NaN\n",
        "dbnsfp.aa.aapos_fathmm                [ENSP00000318690:T594A, ENSP00000386552:T556A]\n",
        "dbnsfp.aa.aapos_sift                                           ENSP00000318690:T594A\n",
        "dbnsfp.aa.alt                                                                      A\n",
        "dbnsfp.aa.codonpos                                                                 1\n",
        "dbnsfp.aa.pos                                                             [594, 556]\n",
        "dbnsfp.aa.ref                                                                      T\n",
        "dbnsfp.aa.refcodon                                                               ACT\n",
        "dbnsfp.allele1                                                                     T\n",
        "dbnsfp.allele2                                                                     C\n",
        "dbnsfp.ancestral_allele                                                            T\n",
        "...                                                                              ...\n",
        "evs.hgvs.coding                                                                  NaN\n",
        "evs.hgvs.protein                                                                 NaN\n",
        "evs.ma_fin_percent.african_american                                          47.0949\n",
        "evs.ma_fin_percent.all                                                       24.8962\n",
        "evs.ma_fin_percent.european_american                                         13.5233\n",
        "evs.on_illumina_human_exome_chip                                                 yes\n",
        "evs.polyphen2                                                                    NaN\n",
        "evs.polyphen2.class                                                                b\n",
        "evs.polyphen2.score                                                                e\n",
        "evs.ref_base_ncbi                                                                  T\n",
        "evs.rs_id                                                                  rs6722682\n",
        "gwassnps.allele1                                                                 NaN\n",
        "gwassnps.allele2                                                                 NaN\n",
        "gwassnps.chrom                                                                   NaN\n",
        "gwassnps.chromEnd                                                                NaN\n",
        "gwassnps.chromStart                                                              NaN\n",
        "gwassnps.pubmedID                                                                NaN\n",
        "gwassnps.rsid                                                                    NaN\n",
        "gwassnps.trait                                                                   NaN\n",
        "mutdb.allele1                                                                    NaN\n",
        "mutdb.allele2                                                                    NaN\n",
        "mutdb.chrom                                                                      NaN\n",
        "mutdb.chromEnd                                                                   NaN\n",
        "mutdb.chromStart                                                                 NaN\n",
        "mutdb.cosmic_id                                                                  NaN\n",
        "mutdb.dbsnp_id                                                                   NaN\n",
        "mutdb.mutpred_score                                                              NaN\n",
        "mutdb.strand                                                                     NaN\n",
        "mutdb.uniprot_id                                                                 NaN\n",
        "snpedia.text                                                                     NaN\n",
        "\n",
        "[242 rows x 1 columns]"
       ]
      }
     ],
     "prompt_number": 16
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "<center><h1>fiSSEA Functions</h1><center>"
     ]
    },
    {
     "cell_type": "heading",
     "level": 3,
     "metadata": {},
     "source": [
      "MyVariant.info Query and Parsing"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "\n",
      "def myvariant_post(hgvs_list):\n",
      "    '''\n",
      "    Query and Parser for myvariant.info\n",
      "    Parses Raw Elastic Search results into\n",
      "    pandas dataframe\n",
      "    \n",
      "    Parameters\n",
      "    -------------\n",
      "    hgvs_list: list, required\n",
      "        \n",
      "        \n",
      "    Output\n",
      "    -------------\n",
      "    pandas df:\n",
      "        normalized json of myvariant results\n",
      "    '''\n",
      "    \n",
      "    def normalize(con):\n",
      "        '''\n",
      "        This function uses the json.loads and json_normalize\n",
      "        modules to flatten the myvariant.info results\n",
      "        '''\n",
      "        temp_df = json_normalize(json.loads(con), 'docs')\n",
      "        source_parsed = []\n",
      "        for source_json in temp_df['_source'].values:\n",
      "            try:\n",
      "                source_parsed.append(json_normalize(source_json))\n",
      "            except TypeError:  #occurs with empty results\n",
      "                pass\n",
      "        return pd.concat(source_parsed)\n",
      "    \n",
      "    \n",
      "    \n",
      "    if type(hgvs_list) == list:\n",
      "        hgvs_list = ','.join(hgvs_list)\n",
      "        \n",
      "    assert type(hgvs_list) == str\n",
      "    \n",
      "    h = httplib2.Http()\n",
      "    headers = {'content-type': 'application/x-www-form-urlencoded'}\n",
      "    params = 'ids=' + hgvs_list +\",size=1000\"  #can pass a dictionary\n",
      "    #print params\n",
      "    res, con = h.request('http://myvariant.info/api/variant/', 'POST', params, headers=headers)\n",
      "    \n",
      "    \n",
      "    mv_df = normalize(con)\n",
      "    mv_df.index = mv_df['_id']\n",
      "\n",
      "    return mv_df\n",
      "    \n",
      "    "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 3
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def get_myvariant_snp_annot(vcf_df_mv):\n",
      "    '''\n",
      "    Returns myvariant.info annotations for all snps for\n",
      "    input vcf pandas dataframe. Submits myvariant.info\n",
      "    POST request in snp batches of 1000.\n",
      "    \n",
      "    Parameters\n",
      "    -------------\n",
      "    vcf_df_mv: pandas df; required\n",
      "    \n",
      "        vcf_df_mv must contain the following columns:\n",
      "            vartype1, vartype2, CHROM, REF, POS, a1, a2\n",
      "            *please see pandasVCF docs for column definitions\n",
      "            \n",
      "\n",
      "    Output\n",
      "    -------------\n",
      "    myvariant.info results: pandas df\n",
      "    \n",
      "    '''\n",
      "    \n",
      "    def chunks(l, n):\n",
      "        \"\"\" \n",
      "        Yield successive n-sized chunks from l.\n",
      "        \"\"\"\n",
      "        l = list(l)\n",
      "        chunk_list = []\n",
      "        for i in xrange(0, len(l), n):\n",
      "            chunk_list.append( l[i:i+n] )\n",
      "        return chunk_list\n",
      "    \n",
      "    \n",
      " \n",
      "    if 'CHROM' not in vcf_df_mv.columns:\n",
      "        vcf_df_mv.reset_index(inplace=True)\n",
      "        \n",
      "    \n",
      "    vcf_df_mv = vcf_df_mv[(vcf_df_mv['vartype1'] == 'snp') | (vcf_df_mv['vartype2'] == 'snp')] #restrict to snps\n",
      "\n",
      "    non_ref_a1 = vcf_df_mv[vcf_df_mv['REF'] != vcf_df_mv['a1']][['CHROM', 'POS', 'REF', 'a1']]  #restrict a1 to non-reference alleles\n",
      "    hgvs_a1 = 'chr' + non_ref_a1['CHROM'].astype(str) + \":g.\" + non_ref_a1['POS'].astype(str)  \\\n",
      "                                 + non_ref_a1['REF'] + '>' +  non_ref_a1['a1']\n",
      "\n",
      "    non_ref_a2 = vcf_df_mv[vcf_df_mv['REF'] != vcf_df_mv['a2']][['CHROM', 'POS', 'REF', 'a2']]  #restrict a2 to non-reference alleles\n",
      "    hgvs_a2 = 'chr' + non_ref_a2['CHROM'].astype(str) + \":g.\" + non_ref_a2['POS'].astype(str) \\\n",
      "                                 + non_ref_a2['REF'] + '>' +  non_ref_a2['a2']\n",
      "\n",
      "\n",
      "    hgvs_ids = set(hgvs_a1.values) | set(hgvs_a2.values)  #set of non-reference a1 and a2 snps\n",
      "\n",
      "    hgvs_ids_chunks = chunks(hgvs_ids, 1000)  #batch queries of hgvs a1 and a2 variants, n=1000\n",
      "\n",
      "    myvariant_results = map(myvariant_post, hgvs_ids_chunks)\n",
      "        \n",
      "    return myvariant_results"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 4
    },
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "fiSSEA Function"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "class fiSSEA(object):\n",
      "\n",
      "    \n",
      "    def __init__(self, vcf_file_path, sample_id='', chunksize=100000, fiSSEA_score='dbnsfp.cadd.phred', gene_stat=np.mean):\n",
      "        \n",
      "        \n",
      "        #setting fiSSEA_score\n",
      "        self.fiSSEA_score = fiSSEA_score\n",
      "        \n",
      "        #setting gene statistic\n",
      "        self.gene_stat = gene_stat\n",
      "        \n",
      "        #open Vcf object\n",
      "        vcf_df_obj = Vcf(vcf_file_path, sample_id, ['#CHROM', 'POS', 'REF', 'ALT', 'ID', 'FILTER', 'FORMAT'], chunksize=chunksize)\n",
      "\n",
      "        #read in vcf chunk\n",
      "        vcf_df_obj.get_vcf_df_chunk()\n",
      "\n",
      "        #drop duplicate vcf lines, if necessary\n",
      "        vcf_df_obj.df = vcf_df_obj.df.drop_duplicates(subset=['CHROM', 'POS'])\n",
      "\n",
      "        #remove empty variant lines\n",
      "        vcf_df_obj.df = vcf_df_obj.df[vcf_df_obj.df.ALT!='.']\n",
      "\n",
      "        #replace chr if necessary, depends on format of annotations\n",
      "        vcf_df_obj.df['CHROM'] = vcf_df_obj.df['CHROM'].str.replace('chr', '')\n",
      "\n",
      "        #parse vcf lines\n",
      "        vcf_df_obj.add_variant_annotations( inplace=True, verbose=True)\n",
      "        \n",
      "        #settting vcf_df as class object\n",
      "        self.vcf_df = vcf_df_obj.df\n",
      "\n",
      "        #get myvariant.info annotations, create fiSSEA dataframe\n",
      "        self.fiSSEA_status = self.get_fi_scores()\n",
      "        \n",
      "        #setting sample_id\n",
      "        self.sample_id = sample_id\n",
      "        \n",
      "\n",
      "    \n",
      "    def get_multigene_var_records(self, df_line):\n",
      "        '''\n",
      "        Splits multigene variants into separate\n",
      "        rows\n",
      "        '''\n",
      "        genes = df_line['dbnsfp.genename']\n",
      "        gene_series = [] #list of gene variants\n",
      "        for gene in genes: #iterate through all genes intersecting this variant\n",
      "            df_line_copy = df_line.copy()  #avoids aliasing\n",
      "            del df_line_copy['dbnsfp.genename']  #removing genename to allow gene insertion\n",
      "            df_line_copy['dbnsfp.genename'] = gene\n",
      "            gene_series.append(df_line_copy)\n",
      "\n",
      "        return gene_series\n",
      "    \n",
      "    \n",
      "    \n",
      "    def get_fi_scores(self):\n",
      "        '''\n",
      "        Creates fi_scores series, index on gene symbol, value fi_score.\n",
      "        '''\n",
      "        \n",
      "        #Retrieving myvariant.info annotations for all snps\n",
      "        myvariant_results = get_myvariant_snp_annot(self.vcf_df) #query and parse myvariant.info using all snps in self.vcf_df\n",
      "        self.mv_annot = pd.concat(myvariant_results)  #setting dataframe of parsed myvariant.info results\n",
      "\n",
      "        #reducing size of dataframe for fiSSEA\n",
      "        fiSSEA_cols = ['dbnsfp.genename', self.fiSSEA_score]\n",
      "        mv_annot_fiSSEA = self.mv_annot[fiSSEA_cols] \n",
      "        \n",
      "        #dropping variants without a genename & filling fi_score NaN with 0\n",
      "        mv_annot_fiSSEA.dropna(subset=['dbnsfp.genename'], inplace=True)  \n",
      "        mv_annot_fiSSEA[self.fiSSEA_score].fillna(value=0)  \n",
      "\n",
      "        \n",
      "        #Splitting variants intersecting multiple genes into separate rows\n",
      "        mv_annot_fiSSEA['dbnsfp.genename_type'] = mv_annot_fiSSEA['dbnsfp.genename'].map(type)\n",
      "        mv_annot_fiSSEA_multigene = mv_annot_fiSSEA[mv_annot_fiSSEA['dbnsfp.genename_type']==list]\n",
      "        mv_annot_fiSSEA = mv_annot_fiSSEA[~mv_annot_fiSSEA.index.isin(mv_annot_fiSSEA_multigene.index)]\n",
      "        \n",
      "        if len(mv_annot_fiSSEA_multigene) > 0:\n",
      "            mv_annot_fiSSEA_multigene['index'] = mv_annot_fiSSEA_multigene.index\n",
      "            mv_annot_fiSSEA_multigene.drop_duplicates(subset='index', inplace=True)\n",
      "            mv_annot_fiSSEA_multigene = mv_annot_fiSSEA_multigene.apply(self.get_multigene_var_records, axis=1)\n",
      "            mv_annot_fiSSEA_multigene = [n for i in mv_annot_fiSSEA_multigene for n in i]\n",
      "            mv_annot_fiSSEA = mv_annot_fiSSEA.append(pd.DataFrame(mv_annot_fiSSEA_multigene))\n",
      "            del mv_annot_fiSSEA_multigene\n",
      "        \n",
      "        gc.collect()\n",
      "\n",
      "        \n",
      "        #Setting gene specific functional impact score\n",
      "        self.fi_scores = mv_annot_fiSSEA.groupby('dbnsfp.genename')[self.fiSSEA_score].aggregate(self.gene_stat)\n",
      "        \n",
      "        return 0\n",
      "\n",
      "\n",
      "    def write_rnk_file(self):\n",
      "            '''\n",
      "            Writes functional impact scores to .rnk file in\n",
      "            the fiSSEA/gsea/rnk/ directory.\n",
      "    \n",
      "            Returns path to rnk file\n",
      "            '''\n",
      "    \n",
      "            fiSSEA_results_write = self.fi_scores.copy()\n",
      "            fiSSEA_results_write.name = '#' + fiSSEA_results_write.name \n",
      "    \n",
      "    \n",
      "            rnk_path = self.rnk_write_dir + self.sample_id + self.fi_scores.name +'.rnk'\n",
      "            self.rnk_path = rnk_path\n",
      "            \n",
      "            if not os.path.exists(self.rnk_write_dir):\n",
      "                    os.makedirs(self.rnk_write_dir)\n",
      "            \n",
      "            fiSSEA_results_write.to_csv(self.rnk_path, sep='\\t')\n",
      "            return True\n",
      "        \n",
      "        \n",
      "        \n",
      "    def call_preranked_gsea(self):\n",
      "        '''\n",
      "        Calls GSEA PreRanked Analysis using several default parameters\n",
      "        '''\n",
      "    \n",
      "        gsea_cmd = ['java', '-cp', self.gsea_jar_path, '-Xmx512m', 'xtools.gsea.GseaPreranked', \\\n",
      "                '-gmx', self.tissue_geneset_path, '-collapse false', '-mode Max_probe', \\\n",
      "                 '-nperm 1000', '-rnk', self.rnk_path, '-scoring_scheme classic', \\\n",
      "                '-rpt_label my_analysis', '-include_only_symbols true', \\\n",
      "                '-rpt_label', self.sample_id,  '-chip gseaftp.broadinstitute.org://pub/gsea/annotations/GENE_SYMBOL.chip', \\\n",
      "                '-include_only_symbols true', '-make_sets true', '-plot_top_x 20', \\\n",
      "                '-rnd_seed timestamp', '-set_max 15000', '-set_min 15', '-zip_report false', \\\n",
      "                '-out', self.output_dir_path, '-gui false']\n",
      "    \n",
      "    \n",
      "        p = subprocess.Popen(gsea_cmd,\n",
      "                         stdout=subprocess.PIPE,\n",
      "                         stderr=subprocess.STDOUT)\n",
      "    \n",
      "        run_stdout = [\">>> \" + line.rstrip() for line in iter(p.stdout.readline, b'')]\n",
      "        return run_stdout\n",
      "    \n",
      "    \n",
      "    \n",
      "    def run_GSEA_preranked(self, gsea_jar_path, tissue_geneset_path, output_dir_path):\n",
      "        '''\n",
      "        Calls GSEA jar for a PreRanked analysis.\n",
      "        \n",
      "        Parameters\n",
      "        -------------\n",
      "        gsea_jar_path: str, required\n",
      "            Specifies the path to GSEA jar, e.g. /Users/bin/gsea/gsea2-2.1.0.jar\n",
      "                Can be downloaded from http://www.broadinstitute.org/gsea/downloads.jsp\n",
      "                \n",
      "        rnk_write_dir: str, required\n",
      "            Specifies the directory path where the rnk file should be written\n",
      "                e.g. /Users/data/gsea/rnk/\n",
      "        \n",
      "        tissue_geneset_path: str, required\n",
      "            Specifies the path to the tissue genesets (.gmt format)\n",
      "                e.g.  /Users/data/gsea/genesets/EnrichmentValues_Primary_U133A_ave_percent_over70percent_no_brain.gmt\n",
      "        \n",
      "        output_dir_path: str, required\n",
      "            Specifies the directory path where the GSEA output should be written\n",
      "                e.g. /Users/data/gsea/results/\n",
      "        \n",
      "        Returns\n",
      "        -------------\n",
      "        Writes GSEA PreRanked Results to output_dir\n",
      "        \n",
      "        '''\n",
      "        \n",
      "        #Setting GSEA parameter paths\n",
      "        self.gsea_jar_path = gsea_jar_path\n",
      "        self.rnk_write_dir = output_dir_path\n",
      "        self.tissue_geneset_path = tissue_geneset_path\n",
      "        self.output_dir_path = output_dir_path\n",
      "        \n",
      "        #Writing the rnk file to disk\n",
      "        self.write_rnk_file()\n",
      "        \n",
      "        #Run PreRanked Gene Set Enrichment Analysis, please see run_preranked_gsea function to change\n",
      "        #parameters\n",
      "        #Returns 0 if successful run\n",
      "        run_stdout = self.call_preranked_gsea()\n",
      "        self.run_stdout = run_stdout\n",
      "\n",
      "        \n",
      "        last_written_dir = os.popen('ls -lht ' + self.output_dir_path).readlines()\n",
      "        last_written_dir = last_written_dir[1].split()[-1]\n",
      "        self.last_written_dir = last_written_dir  #likely the path to the output files\n",
      "        \n",
      "        return 0\n",
      "\n",
      "    \n",
      "    def open_GSEA_html(self):\n",
      "        '''\n",
      "        Polls the user-specified output directory for the last written\n",
      "        file and opens the index.html file in that directory\n",
      "        '''\n",
      "        #Opening GSEA HTML Results Report\n",
      "        os.system('open ' + self.output_dir_path + self.last_written_dir + '/index.html')\n",
      "        return 0\n",
      "\n",
      "    "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 29
    }
   ],
   "metadata": {}
  }
 ]
}